{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "from __future__ import (absolute_import, division,\n",
    "                        print_function, unicode_literals)\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter('ignore')\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import json\n",
    "import time\n",
    "import re\n",
    "\n",
    "import psycopg2\n",
    "import pandas.io.sql as sqlio\n",
    "\n",
    "import scipy.stats as stats\n",
    "import statsmodels.stats as smstats\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "from multiprocessing import Process, Manager, Pool\n",
    "import multiprocessing\n",
    "from functools import partial\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "import seaborn as sns; sns.set()\n",
    "\n",
    "import matplotlib\n",
    "matplotlib.style.use('seaborn')\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "matplotlib.rcParams['backend'] = \"Qt5Agg\"\n",
    "import matplotlib.ticker as ticker\n",
    "from matplotlib.ticker import FuncFormatter\n",
    "\n",
    "from IPython.display import display, Image\n",
    "\n",
    "from adjustText import adjust_text\n",
    "\n",
    "%pylab inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# example of a configuration for connection to the \"gtex_ipsa\" PostgreSQL database. \n",
    "# Please update it according to your configuration settings\n",
    "\n",
    "conn = psycopg2.connect(database='gtex_ipsa', user='postgres', host='localhost', port='5439')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import external script with necessary helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib\n",
    "import helper_functions as aaf\n",
    "importlib.reload(aaf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check RBP expression in RBP perturbation experiments\n",
    "\n",
    "We check that RBP expression changed upon perturbation in the intended direction, i.e. decreased upon KD or KO and increased upon OE; otherwise, the experiment was excluded from the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [],
   "source": [
    "sql = \"\"\"SELECT * FROM experiment_metadata\"\"\"\n",
    "experiment_metadata = pd.read_sql_query(sql, conn)\n",
    "experiment_metadata['sample_number'] = experiment_metadata.apply(lambda x: len(x['list_of_ids'].split(' ')),1)\n",
    "experiment_metadata = experiment_metadata.loc[experiment_metadata['experiment']!='no treat']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_name(x):\n",
    "    if x['experiment'][0]==' ':\n",
    "        gene_name = x['experiment'].split(' ')[1]\n",
    "    else:\n",
    "        gene_name = x['experiment'].split(' ')[0]\n",
    "    return gene_name\n",
    "    \n",
    "def get_perturbation(x):\n",
    "    if x['experiment'][0]==' ':\n",
    "        perturbation = x['experiment'].split(' ')[2]\n",
    "    else:\n",
    "        perturbation = x['experiment'].split(' ')[1]\n",
    "    if perturbation.lower() in ['kd','sirna','shrna','ko','kd,']:\n",
    "        perturbation = 'down'\n",
    "    elif perturbation.lower()=='oe':\n",
    "        perturbation = 'up'\n",
    "    return perturbation\n",
    "    \n",
    "experiment_metadata['gene_name'] = experiment_metadata.apply(lambda x: get_gene_name(x),1)\n",
    "experiment_metadata['perturbation'] = experiment_metadata.apply(lambda x: get_perturbation(x),1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "sql = \"\"\"SELECT * FROM deseq\"\"\"\n",
    "deseq = pd.read_sql_query(sql, conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_metadata = pd.merge(experiment_metadata,deseq[['exper_id','gene_name','log2foldchange']],how='left',on=['exper_id','gene_name'])\n",
    "experiment_metadata['log2foldchange'] = experiment_metadata['log2foldchange'].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [],
   "source": [
    "def correct_direction(x):\n",
    "    if x['perturbation']=='down':\n",
    "        if x['log2foldchange']<0:\n",
    "            return 'y'\n",
    "        else:\n",
    "            return 'n'\n",
    "    elif x['perturbation']=='up':\n",
    "        if x['log2foldchange']>0:\n",
    "            return 'y'\n",
    "        else:\n",
    "            return 'n'\n",
    "    else:\n",
    "        return 'y'\n",
    "experiment_metadata['correct_direction'] = experiment_metadata.apply(lambda x: correct_direction(x),1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_metadata = experiment_metadata.loc[(experiment_metadata['correct_direction']=='y')|(\n",
    "    experiment_metadata['data_source']=='SRP041788')]\n",
    "# SRP041788 stores NMD inactivation experiment in which we did not run DESeq2 analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "425"
      ]
     },
     "execution_count": 347,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(experiment_metadata)\n",
    "# including 419 RBP perturbations and 6 NMD inactivation experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_metadata.to_csv('./experiment_metadata.tsv', \n",
    "                              sep=str('\\t'),index=None,header=True,encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regulation of validated RBP-USE pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_metadata = pd.read_csv('./experiment_metadata.tsv',delimiter=\"\\t\",index_col=None,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the table with validated USEs\n",
    "\n",
    "leus = pd.read_csv('./misc/Literature_examples_of_unproductive_splicing.txt',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)\n",
    "leus = leus.fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = ['Target','AS event position', 'rMATS AS type','NMD site upstream hg19','NMD site downstream hg19',\n",
    "       'NMD junction upstream hg19', 'NMD junction downstream hg19',\n",
    "       'coding site upstream hg19', 'coding site downstream hg19',\n",
    "       'coding junction upstream hg19', 'coding junction downstream hg19']\n",
    "leus_df = leus.groupby(cols)['Regulators'].apply(';\\n'.join).reset_index()\n",
    "tmp = leus.groupby(cols)['Regulation'].apply(';\\n'.join).reset_index()\n",
    "leus_df = pd.merge(leus_df,tmp,how='inner',on=cols)\n",
    "leus_df = leus_df.loc[leus_df['Target']!='']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_isoform_number(x):\n",
    "    if 'NMD' in x['isoform']:\n",
    "        return 1\n",
    "    else:\n",
    "        return 2\n",
    "\n",
    "def get_average_expression(x):\n",
    "    a = []\n",
    "    for field in ['ijc_sample_1','sjc_sample_1','ijc_sample_2','sjc_sample_2']:\n",
    "        a = a+x[field].split(',')\n",
    "    return np.median(pd.Series(a).astype('float'))\n",
    "\n",
    "def get_perturb_coef(x):\n",
    "    if x['perturbation'] == 'up':\n",
    "        perturb_coef = 1\n",
    "    elif x['perturbation'] == 'down':\n",
    "        perturb_coef = -1\n",
    "    else:\n",
    "        perturb_coef = 0\n",
    "    return perturb_coef\n",
    "\n",
    "def get_rmats_deseq(gene_name,input_df,AS_type):\n",
    "\n",
    "    rMATS_coordinates = aaf.get_rMATS_coordinates(input_df,AS_type)\n",
    "    if rMATS_coordinates is None:\n",
    "        return None\n",
    "    sql = aaf.get_sql_query_for_rmats(rMATS_coordinates)\n",
    "    rmats = pd.read_sql_query(sql, conn)\n",
    "    if len(rmats)==0:\n",
    "        return None\n",
    "    rmats = pd.merge(rmats,experiment_metadata[['exper_id','data_source','cell_lines','experiment','corr_coef','sample_number',\n",
    "                                                'gene_name','perturbation']].rename(columns = {'gene_name':'regulator'}),\n",
    "             how='inner',on='exper_id')\n",
    "    adj_coef = rMATS_coordinates.loc[rMATS_coordinates['param']=='adj_coef']['val'].iloc[0]\n",
    "    rmats['adj_coef'] = adj_coef\n",
    "    rmats['delta_psi'] = adj_coef*rmats['corr_coef']*rmats['incleveldifference']\n",
    "    rmats['median_local_expr'] = rmats.apply(lambda x: get_average_expression(x),1)\n",
    "\n",
    "    sql = \"SELECT * FROM deseq WHERE gene_name = '\"+gene_name+\"\"\"'\"\"\"\n",
    "    deseq = pd.read_sql_query(sql, conn)\n",
    "    deseq.rename(columns = {'pvalue':'pvalue_deseq'},inplace=True)\n",
    "\n",
    "    rmats_deseq = pd.merge(rmats.loc[(~rmats['experiment'].str.contains('|'.join(['CHX','UPF1'])))],deseq[['exper_id','log2foldchange','pvalue_deseq']],how='inner',on='exper_id')\n",
    "    rmats_deseq.rename(columns = {'pvalue':'pvalue_rMATS'},inplace=True)\n",
    "\n",
    "    # adjust the splicing and expression changes as if all the experiments had overexpressed RBPs\n",
    "    rmats_deseq['perturb_coef'] = rmats_deseq.apply(lambda x: get_perturb_coef(x),1)\n",
    "    rmats_deseq = rmats_deseq.loc[rmats_deseq['perturb_coef']!=0]\n",
    "    rmats_deseq['delta_psi'] = rmats_deseq['perturb_coef']*rmats_deseq['delta_psi']\n",
    "    rmats_deseq['log2foldchange'] = rmats_deseq['perturb_coef']*rmats_deseq['log2foldchange']\n",
    "\n",
    "    return rmats_deseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we only look at USEs in which there are reported regulators and the AS class can be quantified with rMATS\n",
    "test_leus_df = leus_df.loc[(leus_df['Regulators']!='')&(leus_df['rMATS AS type']!='complex')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done: CPSF7 exon 4\n",
      "done: DLG4 exon 17\n",
      "done: EZH2 exon 8a\n",
      "done: FUS exon 7\n",
      "done: GABBR1 exon 15\n",
      "done: HNRNPA2B1 3'UTR\n",
      "done: HNRNPD 3'UTR\n",
      "done: HNRNPDL 3'UTR\n",
      "done: HNRNPL exon 6a\n",
      "done: HNRNPLL exon 6a\n",
      "done: HPS1 exon 18\n",
      "done: LMNA exon 2a\n",
      "done: MDM4 exon 6\n",
      "done: PTBP1 exon 11\n",
      "done: PTBP2 exon 10\n",
      "done: PTBP3 exon 2\n",
      "no results: RBFOX2 exon 5a\n",
      "no results: RBFOX2 exon 6a\n",
      "done: RBM10 exon 12\n",
      "done: RBM10 exon 6\n",
      "done: RBM39 exon 3\n",
      "done: RBM5 exon 16\n",
      "done: RBM5 exon 6\n",
      "done: REST exon 4 ab\n",
      "done: RPL10A intron 3\n",
      "done: RPL3 exon 4\n",
      "done: RPS3 exon 2\n",
      "done: SF1 exon 3\n",
      "done: SMNDC1 exon 3a\n",
      "done: SNRNP70 exon 8a\n",
      "done: SNRNP70 exon 8b\n",
      "done: SNRNP70 exon 8c\n",
      "done: SNRPB exon 3a\n",
      "done: SRSF1 3'UTR\n",
      "done: SRSF2 3'UTR\n",
      "done: SRSF3 exon 4\n",
      "done: SRSF4 exon 3a\n",
      "done: SRSF5 exon 5a\n",
      "done: SRSF5 intron 5\n",
      "done: SRSF6 exon 3a\n",
      "done: SRSF7 exon 3a\n",
      "no results: SRSF8 3'UTR\n",
      "done: SRSF9 exon 3a\n",
      "done: TARDBP 3'UTR\n",
      "done: TIA1 exon 7a\n",
      "done: TRA2A exon 2a\n",
      "done: TRA2B exon 2\n",
      "done: U2AF1 exon 3b\n"
     ]
    }
   ],
   "source": [
    "cols = pd.Series(leus_df.columns)\n",
    "NMD_cols = list(cols.loc[cols.str.contains('NMD ')])\n",
    "can_cols = list(cols.loc[cols.str.contains('coding ')])\n",
    "          \n",
    "a = []\n",
    "r = []\n",
    "\n",
    "for index, row in test_leus_df.iterrows():\n",
    "    gene_name = row['Target']\n",
    "    AS_event_position = row['AS event position']\n",
    "    \n",
    "    Regulators = pd.DataFrame([\n",
    "        re.split(';\\n|; \\n',row['Regulators']),re.split(';\\n|; \\n',row['Regulation'])]).transpose()\n",
    "    Regulators.columns = ['regulator','regulation']\n",
    "    Regulators = Regulators.drop_duplicates('regulator')\n",
    "    Regulators['regulator'] = Regulators['regulator'].str.replace('\\n','')\n",
    "    Regulators['regulation'] = Regulators['regulation'].str.replace('\\n','')\n",
    "    Regulators['regulation'] = Regulators['regulation'].str.replace(';','')\n",
    "    Regulators['gene_name'] = gene_name\n",
    "    Regulators['AS_event_position'] = AS_event_position\n",
    "    r.append(Regulators)\n",
    "    \n",
    "    input_df = pd.DataFrame([list(row[NMD_cols+can_cols]),\n",
    "                          list(NMD_cols+can_cols)]).transpose()\n",
    "    input_df.columns=['long_id','isoform']\n",
    "    input_df['number'] = input_df.apply(lambda x:get_isoform_number(x),1)\n",
    "    input_df = input_df.loc[input_df['long_id']!='']\n",
    "    AS_type = row['rMATS AS type']\n",
    "    directory = './literature_events/'+gene_name+'_'+AS_event_position.replace(\"\"\"'\"\"\",'prime')\n",
    "    rmats_deseq = get_rmats_deseq(gene_name,input_df,AS_type)\n",
    "    if rmats_deseq is not None and len(rmats_deseq)>0:\n",
    "        rmats_deseq = rmats_deseq[['exper_id','data_source', 'cell_lines',\n",
    "       'experiment', 'corr_coef','adj_coef','perturb_coef', 'sample_number', 'regulator', 'perturbation','median_local_expr','inclevel1','inclevel2',\n",
    "       'delta_psi','pvalue_rMATS','log2foldchange', 'pvalue_deseq']]\n",
    "        rmats_deseq['gene_name'] = gene_name\n",
    "        rmats_deseq['AS_event_position'] = AS_event_position\n",
    "        a.append(rmats_deseq)\n",
    "        print('done: '+gene_name+' '+AS_event_position)\n",
    "    else:\n",
    "        print('no results: '+gene_name+' '+AS_event_position)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we obtained the combined rMATS output \n",
    "rmats_deseq_res = pd.concat(a)\n",
    "rmats_deseq_res = rmats_deseq_res.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_psi_L(x):\n",
    "    if x['corr_coef']*x['perturb_coef']==1:\n",
    "        t = pd.Series(x['inclevel2'].split(','))\n",
    "    else:\n",
    "        t = pd.Series(x['inclevel1'].split(','))\n",
    "    t = t.loc[t!='NA']\n",
    "    t = t.astype('float')\n",
    "    psi = np.mean(t)    \n",
    "    if x['adj_coef']==-1:\n",
    "        psi = 1-psi\n",
    "    return psi\n",
    "\n",
    "def get_psi_H(x):\n",
    "    if x['corr_coef']*x['perturb_coef']==1:\n",
    "        t = pd.Series(x['inclevel1'].split(','))\n",
    "    else:\n",
    "        t = pd.Series(x['inclevel2'].split(','))\n",
    "    t = t.loc[t!='NA']\n",
    "    t = t.astype('float')\n",
    "    psi = np.mean(t)\n",
    "    if x['adj_coef']==-1:\n",
    "        psi = 1-psi\n",
    "    return psi\n",
    "\n",
    "rmats_deseq_res['psi_L'] = rmats_deseq_res.apply(lambda x: get_psi_L(x),1) # the value in the condition of low RBP expression\n",
    "rmats_deseq_res['psi_H'] = rmats_deseq_res.apply(lambda x: get_psi_H(x),1) # the value in the condition of high RBP expression\n",
    "rmats_deseq_res['psi_H_L'] = rmats_deseq_res['psi_H']-rmats_deseq_res['psi_L']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-0.0005000000000000976, 0.0005000000000001115)"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we checked that psi_H_L and delta_psi values are equal.\n",
    "\n",
    "rmats_deseq_res['temp'] = rmats_deseq_res['psi_H_L']-rmats_deseq_res['delta_psi']\n",
    "rmats_deseq_res['temp'].min(),rmats_deseq_res['temp'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "rmats_deseq_res['delta_psi'] = rmats_deseq_res['psi_H_L']\n",
    "rmats_deseq_res = rmats_deseq_res.drop('psi_H_L',1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get psi in control conditions\n",
    "def get_psi_ctl(x):\n",
    "    if x['perturbation']=='down':\n",
    "        return x['psi_H']\n",
    "    else:\n",
    "        return x['psi_L']\n",
    "\n",
    "# get psi in experimental conditions\n",
    "def get_psi_exp(x):\n",
    "    if x['perturbation']=='down':\n",
    "        return x['psi_L']\n",
    "    else:\n",
    "        return x['psi_H']\n",
    "\n",
    "rmats_deseq_res['psi_ctl'] = rmats_deseq_res.apply(lambda x:get_psi_ctl(x),1)\n",
    "rmats_deseq_res['psi_exp'] = rmats_deseq_res.apply(lambda x:get_psi_exp(x),1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_category(x):\n",
    "    if x['regulator']==x['gene_name']:\n",
    "        category = 'auto'\n",
    "    elif x['regulation']=='NMD-inhibiting':\n",
    "        category = 'cross, NMD-inhibiting'\n",
    "    elif x['regulation']=='NMD-promoting':\n",
    "        category = 'cross, NMD-promoting'\n",
    "    else:\n",
    "        category = ''\n",
    "    return category\n",
    "\n",
    "Regulators_res = pd.concat(r)\n",
    "Regulators_res= Regulators_res.reset_index(drop=True)\n",
    "Regulators_res = Regulators_res.loc[Regulators_res['regulator']!='']\n",
    "Regulators_res['category'] = Regulators_res.apply(lambda x: get_category(x),1)\n",
    "\n",
    "experiment_metadata['t']=1\n",
    "tmp1 = experiment_metadata.loc[experiment_metadata['perturbation']=='up'].groupby(['gene_name']).agg({'t':sum}).reset_index().rename(columns ={'gene_name':'regulator','t':'t_up'})\n",
    "tmp1_1 = experiment_metadata.loc[experiment_metadata['perturbation']=='down'].groupby(['gene_name']).agg({'t':sum}).reset_index().rename(columns ={'gene_name':'regulator','t':'t_down'})\n",
    "\n",
    "tmp1 = pd.merge(tmp1,tmp1_1,how='outer',on='regulator')\n",
    "tmp1[['t_up','t_down']] = tmp1[['t_up','t_down']].fillna(0).astype('int')\n",
    "\n",
    "tmp2 = Regulators_res[['regulator']].drop_duplicates()\n",
    "tmp2 = pd.merge(tmp2,tmp1,how='left',on='regulator')\n",
    "tmp2[['t_up','t_down']] = tmp2[['t_up','t_down']].fillna(0).astype('int')\n",
    "\n",
    "Regulators_res = pd.merge(Regulators_res,tmp2,how='left',on='regulator')\n",
    "\n",
    "Regulators_res[['t_up','t_down']] = Regulators_res[['t_up','t_down']].fillna(0).astype('int')\n",
    "Regulators_res['n'] = ((Regulators_res['t_up']>0)|(Regulators_res['t_down']>0)).astype('int')\n",
    "Regulators_res['n_up'] = (Regulators_res['t_up']>0).astype('int')\n",
    "Regulators_res['n_down'] = (Regulators_res['t_down']>0).astype('int')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We convert the results of differential splicing (DS) and differential expression (DE) to scores (ranks).\n",
    "\n",
    "def get_rank_DS(x,abs_delta_psi_thr):    \n",
    "    r_DS = 0\n",
    "    thr = abs_delta_psi_thr\n",
    "    \n",
    "    if abs(x['delta_psi'])>thr:\n",
    "        r_DS = 1\n",
    "        if x['fdr_rMATS']<0.05:\n",
    "            r_DS = 1.5\n",
    "    if x['delta_psi']<0:\n",
    "        r_DS = r_DS*(-1)\n",
    "    return r_DS\n",
    "\n",
    "def get_rank_DE(x,abs_log2foldchange_thr):\n",
    "    r_DE = 0\n",
    "    thr = abs_log2foldchange_thr\n",
    "    if abs(x['log2foldchange'])>thr:\n",
    "        r_DE = 1\n",
    "        if x['fdr_deseq']<0.05:\n",
    "            r_DE = 1.5\n",
    "    if x['log2foldchange']<0:\n",
    "        r_DE = r_DE*(-1)\n",
    "    return r_DE\n",
    "\n",
    "rank_DS_thr = 0\n",
    "rank_DE_thr = 0\n",
    "\n",
    "thr_local_expression = 15\n",
    "abs_log2foldchange_thr = 0\n",
    "abs_delta_psi_thr = 0.1\n",
    "\n",
    "rmats_deseq_t = rmats_deseq_res.copy()\n",
    "targets = rmats_deseq_t[['gene_name','AS_event_position']].drop_duplicates()\n",
    "a = []\n",
    "for index,row in targets.iterrows():\n",
    "    tmp = rmats_deseq_t.loc[(rmats_deseq_t['gene_name']==row['gene_name'])&(\n",
    "        rmats_deseq_t['AS_event_position']==row['AS_event_position'])]\n",
    "    tmp['fdr_rMATS'] = multipletests(tmp['pvalue_rMATS'],method='fdr_bh')[1]\n",
    "    tmp['fdr_deseq'] = multipletests(tmp['pvalue_deseq'],method='fdr_bh')[1]\n",
    "    a.append(tmp)\n",
    "rmats_deseq_t = pd.concat(a)\n",
    "rmats_deseq_t['rank_DS'] = rmats_deseq_t.apply(lambda x: get_rank_DS(x,abs_delta_psi_thr),1)\n",
    "rmats_deseq_t['rank_DE'] = rmats_deseq_t.apply(lambda x: get_rank_DE(x,abs_log2foldchange_thr),1)\n",
    "rmats_deseq_t['rank_DS'] = rmats_deseq_t['rank_DS']*(rmats_deseq_t['median_local_expr']>thr_local_expression).astype('int')\n",
    "rmats_deseq_t['rank_DE'] = rmats_deseq_t['rank_DE']*(rmats_deseq_t['median_local_expr']>thr_local_expression).astype('int')\n",
    "\n",
    "rmats_deseq_t = pd.merge(Regulators_res.loc[Regulators_res['n']==1][['regulator','gene_name','regulation','AS_event_position','category']].drop_duplicates(),\n",
    "                         rmats_deseq_t, how='left', on = ['gene_name','AS_event_position','regulator'])\n",
    "\n",
    "res = rmats_deseq_t.groupby(['gene_name','AS_event_position','regulator','regulation','category']).agg({'rank_DS':mean,'rank_DE':mean,'psi_L':mean,'psi_H':mean,'psi_ctl':mean,'psi_exp':mean}).reset_index()\n",
    "\n",
    "# We test if the observed changes in splicing match the expectations from literature\n",
    "res['co_DS'] = ((res['regulation']=='NMD-promoting').astype('int')*2-1)*res['rank_DS']\n",
    "res['anti_DS'] = (res['co_DS']<-abs(rank_DS_thr)).astype('int')\n",
    "res['co_DS'] = (res['co_DS']>abs(rank_DS_thr)).astype('int')\n",
    "\n",
    "# We test if the observed changes in expression match the expectations from literature\n",
    "res['co_DE'] = ((res['regulation']=='NMD-promoting').astype('int')*2-1)*res['rank_DE']*(-1)\n",
    "res['anti_DE'] = (res['co_DE']<-abs(rank_DE_thr)).astype('int')\n",
    "res['co_DE'] = (res['co_DE']>abs(rank_DE_thr)).astype('int')\n",
    "\n",
    "def get_co(x):\n",
    "    if x['regulator']==x['gene_name']:\n",
    "        if x['co_DS']==1:\n",
    "            return 1\n",
    "        else:\n",
    "            return 0\n",
    "    else:\n",
    "        if x['co_DS']==1 and x['co_DE']==1:\n",
    "            return 1\n",
    "        else:\n",
    "            return 0\n",
    "\n",
    "def get_anti(x):\n",
    "    if x['regulator']==x['gene_name']:\n",
    "        if x['anti_DS']==1:\n",
    "            return 1\n",
    "        else:\n",
    "            return 0\n",
    "    else:\n",
    "        if x['anti_DS']==1 and x['anti_DE']==1:\n",
    "            return 1\n",
    "        else:\n",
    "            return 0\n",
    "        \n",
    "res['co'] = res.apply(lambda x:get_co(x),1) # both DS and DE changes match the expectations\n",
    "res['anti'] = res.apply(lambda x:get_anti(x),1) # both DS and DE changes are opposite to expected\n",
    "res['n']=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {},
   "outputs": [],
   "source": [
    "rmats_deseq_t.to_csv('./misc/perturbation_results_details.tsv', \n",
    "                              sep=str('\\t'),index=None,header=True,encoding='utf-8')\n",
    "res.to_csv('./misc/perturbation_results_grouped.tsv', \n",
    "                              sep=str('\\t'),index=None,header=True,encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identification of annotated USEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we extract \"exon\" and \"stop codon\" objects from the annotation using the following bash commands:\n",
    "# awk '$3 == \"exon\"' gencode.v35lift37.annotation.gtf | gzip > hg19_exons.gtf.gz\n",
    "# awk '$3 == \"stop_codon\"' gencode.v35lift37.annotation.gtf | gzip > hg19_stop_codons.gtf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extraction of exon borders and stop codon positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = pd.read_csv('./Annotation/hg19_exons.gtf.gz',delimiter=\"\\t\",index_col=None,header=None,compression='gzip')\n",
    "exons.columns = ['chr','tmp1','tmp0','start','end','tmp2','str','tmp3','info']\n",
    "exons.drop(['tmp1','tmp2','tmp3','tmp0'],1,inplace=True)\n",
    "exons['exon'] = exons['chr']+'_'+exons['start'].astype('str')+'_'+exons['end'].astype('str')+'_'+exons['str']\n",
    "exons.drop(['start','end','str'],1,inplace=True)\n",
    "\n",
    "stop_codons = pd.read_csv('./Annotation/hg19_stop_codons.gtf.gz',delimiter=\"\\t\",index_col=None,header=None,compression='gzip')\n",
    "stop_codons.columns = ['chr','tmp1','tmp0','start','end','tmp2','str','tmp3','info']\n",
    "stop_codons.drop(['tmp1','tmp2','tmp3','tmp0'],1,inplace=True)\n",
    "\n",
    "exons['transcript_id'] = exons['info'].str.split(';',expand=True)[1].str.split('\"',expand=True)[1]\n",
    "exons['gene_name'] = exons['info'].str.split(';',expand=True)[3].str.split('\"',expand=True)[1]\n",
    "exons['exon_number'] = exons['info'].str.split(';',expand=True)[6].str.split(' ',expand=True)[2]\n",
    "exons['transcript_type'] = exons['info'].str.split(';',expand=True)[4].str.split('\"',expand=True)[1]\n",
    "\n",
    "exons['ccds'] = exons['info'].str.contains('ccds').astype('int')\n",
    "\n",
    "stop_codons['transcript_id'] = stop_codons['info'].str.split(';',expand=True)[1].str.split('\"',expand=True)[1]\n",
    "\n",
    "# we do not consider transcripts withous stop codons\n",
    "exons = pd.merge(exons[['exon','gene_name','transcript_id','transcript_type','ccds','exon_number']],\n",
    "stop_codons.rename(columns = {'start':'stop_start','end':'stop_end'})[['transcript_id','stop_start','stop_end']],\n",
    "how='inner',on='transcript_id')\n",
    "\n",
    "exons['exon_start'] = exons['exon'].str.split('_',expand=True)[1].astype('int')\n",
    "exons['exon_end'] = exons['exon'].str.split('_',expand=True)[2].astype('int')\n",
    "\n",
    "exons = exons.loc[exons['exon_number'].str.isnumeric()]\n",
    "exons['exon_number'] = exons['exon_number'].astype('int')\n",
    "\n",
    "gr = exons.groupby('transcript_id').agg({'exon_number':max}).reset_index()\n",
    "\n",
    "exons = pd.merge(exons,gr.rename(columns={'exon_number':'max_exon_number'}),how='left',on='transcript_id')\n",
    "\n",
    "exons.sort_values(['transcript_id','exon_start'],ascending=[True,True],inplace=True)\n",
    "\n",
    "exons = exons.drop_duplicates(['exon','transcript_id']).reset_index(drop=True)\n",
    "\n",
    "transcript_id_pred = ''\n",
    "exon_end_pred = 0\n",
    "upst_junctions = []\n",
    "for a in exons[['transcript_id','exon_start','exon_end']].values:\n",
    "    if (exon_end_pred>0) and (a[0]==transcript_id_pred):\n",
    "        upst_junction = str(exon_end_pred)+'_'+str(a[1])\n",
    "    else:\n",
    "        upst_junction = ''\n",
    "    upst_junctions.append(upst_junction)\n",
    "    exon_end_pred = a[2]\n",
    "    transcript_id_pred = a[0]\n",
    "exons['upst_junction'] = upst_junctions\n",
    "\n",
    "exons.sort_values(['transcript_id','exon_start'],ascending=[True,False],inplace=True)\n",
    "\n",
    "transcript_id_pred = ''\n",
    "exon_start_pred = 0\n",
    "downst_junctions = []\n",
    "for a in exons[['transcript_id','exon_start','exon_end']].values:\n",
    "    if (exon_start_pred>0) and (a[0]==transcript_id_pred):\n",
    "        downst_junction = str(a[2])+'_'+str(exon_start_pred)\n",
    "    else:\n",
    "        downst_junction = ''\n",
    "    downst_junctions.append(downst_junction)\n",
    "    exon_start_pred = a[1]\n",
    "    transcript_id_pred = a[0]\n",
    "exons['downst_junction'] = downst_junctions\n",
    "\n",
    "exons = exons.sort_values(['transcript_id','exon_start'],ascending=[True,True]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Poison exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We identify exons that are included in at least one NMD transcript but skipped in all protein-coding transcripts\n",
    "# We check the following conditions: 1) the presence of NMD isoform with inclusion 2) the presence of coding isoform with skipping \n",
    "# 3) absence of coding isoforms with inclusion 4) absence of NMD isoforms with skipping\n",
    "CE_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(\n",
    "exons['max_exon_number']>exons['exon_number'])&(exons['exon_number']>1)]\n",
    "CE_in_nmd_trs['skip_junction'] = CE_in_nmd_trs['upst_junction'].str.split('_',expand=True)[0]+'_'+CE_in_nmd_trs['downst_junction'].str.split('_',expand=True)[1]\n",
    "\n",
    "exons_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')]\n",
    "exons_in_coding_trs = pd.concat([exons_in_coding_trs[['upst_junction','gene_name']].drop_duplicates().rename(columns = {'upst_junction':'skip_junction'}),\n",
    "exons_in_coding_trs[['downst_junction','gene_name']].drop_duplicates().rename(columns = {'downst_junction':'skip_junction'})]).drop_duplicates().reset_index(drop=True)\n",
    "exons_in_coding_trs = exons_in_coding_trs.loc[exons_in_coding_trs['skip_junction']!='']\n",
    "exons_in_coding_trs['skip_coding_exists'] = 1\n",
    "\n",
    "CE_in_nmd_trs = pd.merge(CE_in_nmd_trs,exons_in_coding_trs,how='left',on=['skip_junction','gene_name'])\n",
    "CE_in_nmd_trs['skip_coding_exists'] = CE_in_nmd_trs['skip_coding_exists'].fillna(0).astype('int')\n",
    "\n",
    "exons_in_coding_trs.rename(columns={'skip_coding_exists':'upst_coding_exists','skip_junction':'upst_junction'},inplace=True)\n",
    "CE_in_nmd_trs = pd.merge(CE_in_nmd_trs,exons_in_coding_trs,how='left',on=['upst_junction','gene_name'])\n",
    "CE_in_nmd_trs['upst_coding_exists'] = CE_in_nmd_trs['upst_coding_exists'].fillna(0).astype('int')\n",
    "\n",
    "exons_in_coding_trs.rename(columns={'upst_coding_exists':'downst_coding_exists','upst_junction':'downst_junction'},inplace=True)\n",
    "CE_in_nmd_trs = pd.merge(CE_in_nmd_trs,exons_in_coding_trs,how='left',on=['downst_junction','gene_name'])\n",
    "CE_in_nmd_trs['downst_coding_exists'] = CE_in_nmd_trs['downst_coding_exists'].fillna(0).astype('int')\n",
    "\n",
    "exons_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')]\n",
    "exons_in_nmd_trs = pd.concat([exons_in_nmd_trs[['upst_junction','gene_name']].drop_duplicates().rename(columns = {'upst_junction':'skip_junction'}),\n",
    "exons_in_nmd_trs[['downst_junction','gene_name']].drop_duplicates().rename(columns = {'downst_junction':'skip_junction'})]).drop_duplicates().reset_index(drop=True)\n",
    "exons_in_nmd_trs = exons_in_nmd_trs.loc[exons_in_nmd_trs['skip_junction']!='']\n",
    "exons_in_nmd_trs['skip_nmd_exists'] = 1\n",
    "CE_in_nmd_trs = pd.merge(CE_in_nmd_trs,exons_in_nmd_trs,how='left',on=['skip_junction','gene_name'])\n",
    "CE_in_nmd_trs['skip_nmd_exists'] = CE_in_nmd_trs['skip_nmd_exists'].fillna(0).astype('int')\n",
    "\n",
    "CE_in_nmd_trs['str']=CE_in_nmd_trs['exon'].str.split('_',expand=True)[3]\n",
    "\n",
    "def get_stop_rel_position(x):\n",
    "    stop_rel_position = ''\n",
    "    if (x['exon_start']<=x['stop_start'] and x['stop_start']<=x['exon_end']) or (x['exon_start']<=x['stop_end'] and x['stop_end']<=x['exon_end']):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>x['exon_end']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<x['exon_start']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>x['exon_end']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<x['exon_start']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "CE_in_nmd_trs['inc_specific_to_NMD_transcript'] = ((CE_in_nmd_trs['skip_coding_exists']==1)&(CE_in_nmd_trs['skip_nmd_exists']==0)&(CE_in_nmd_trs['upst_coding_exists']==0)&(CE_in_nmd_trs['downst_coding_exists']==0)).astype('int')\n",
    "CE_in_nmd_trs['stop_rel_position'] = CE_in_nmd_trs.apply(lambda x:get_stop_rel_position(x),1)\n",
    "poison_exons = CE_in_nmd_trs.loc[CE_in_nmd_trs['inc_specific_to_NMD_transcript']==1]\n",
    "poison_exons['chr'] = poison_exons['exon'].str.split('_',expand=True)[0]\n",
    "poison_exons['str'] = poison_exons['exon'].str.split('_',expand=True)[3]\n",
    "poison_exons = poison_exons[['chr','str','gene_name','transcript_id',\n",
    "'upst_junction','downst_junction','skip_junction','stop_rel_position']].drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'stop_within': 1753, 'stop_downst': 352, 'stop_upst': 207})"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(poison_exons['stop_rel_position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "CE_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(\n",
    "exons['max_exon_number']>exons['exon_number'])&(exons['exon_number']>1)]\n",
    "CE_in_coding_trs['skip_junction'] = CE_in_coding_trs['upst_junction'].str.split('_',expand=True)[0]+'_'+CE_in_coding_trs['downst_junction'].str.split('_',expand=True)[1]\n",
    "\n",
    "exons_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')]\n",
    "exons_in_coding_trs = pd.concat([exons_in_coding_trs[['upst_junction','gene_name']].drop_duplicates().rename(columns = {'upst_junction':'skip_junction'}),\n",
    "exons_in_coding_trs[['downst_junction','gene_name']].drop_duplicates().rename(columns = {'downst_junction':'skip_junction'})]).drop_duplicates().reset_index(drop=True)\n",
    "exons_in_coding_trs = exons_in_coding_trs.loc[exons_in_coding_trs['skip_junction']!='']\n",
    "exons_in_coding_trs['skip_coding_exists'] = 1\n",
    "\n",
    "CE_in_coding_trs = pd.merge(CE_in_coding_trs,exons_in_coding_trs,how='left',on=['skip_junction','gene_name'])\n",
    "CE_in_coding_trs['skip_coding_exists'] = CE_in_coding_trs['skip_coding_exists'].fillna(0).astype('int')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Essential exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We idenify all exons that are skipped in at least one NMD transcript but are included in all protein-coding transcripts\n",
    "CE_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(\n",
    "exons['max_exon_number']>exons['exon_number'])&(exons['exon_number']>1)]\n",
    "CE_in_coding_trs['skip_junction'] = CE_in_coding_trs['upst_junction'].str.split('_',expand=True)[0]+'_'+CE_in_coding_trs['downst_junction'].str.split('_',expand=True)[1]\n",
    "\n",
    "exons_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')]\n",
    "exons_in_coding_trs = pd.concat([exons_in_coding_trs[['upst_junction','gene_name']].drop_duplicates().rename(columns = {'upst_junction':'skip_junction'}),\n",
    "exons_in_coding_trs[['downst_junction','gene_name']].drop_duplicates().rename(columns = {'downst_junction':'skip_junction'})]).drop_duplicates().reset_index(drop=True)\n",
    "exons_in_coding_trs = exons_in_coding_trs.loc[exons_in_coding_trs['skip_junction']!='']\n",
    "exons_in_coding_trs['skip_coding_exists'] = 1\n",
    "\n",
    "CE_in_coding_trs = pd.merge(CE_in_coding_trs,exons_in_coding_trs,how='left',on=['skip_junction','gene_name'])\n",
    "CE_in_coding_trs['skip_coding_exists'] = CE_in_coding_trs['skip_coding_exists'].fillna(0).astype('int')\n",
    "\n",
    "CE_in_coding_trs = CE_in_coding_trs.loc[CE_in_coding_trs['skip_coding_exists']==0][\n",
    "    ['gene_name',\n",
    "     'upst_junction',\n",
    "     'downst_junction',\n",
    "     'skip_junction']].drop_duplicates()\n",
    "\n",
    "CE_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')&(exons['downst_junction']!='')]\n",
    "CE_in_nmd_trs = CE_in_nmd_trs[['gene_name','upst_junction','downst_junction']].drop_duplicates()\n",
    "CE_in_nmd_trs['inc_nmd_exists'] = 1\n",
    "CE_in_coding_trs = pd.merge(CE_in_coding_trs,CE_in_nmd_trs,how='left',on=['gene_name','upst_junction','downst_junction'])\n",
    "CE_in_coding_trs['inc_nmd_exists'] = CE_in_coding_trs['inc_nmd_exists'].fillna(0).astype('int')\n",
    "CE_in_coding_trs = CE_in_coding_trs.loc[CE_in_coding_trs['inc_nmd_exists']==0][\n",
    "    ['gene_name',\n",
    "     'upst_junction',\n",
    "     'downst_junction',\n",
    "     'skip_junction']].drop_duplicates()\n",
    "\n",
    "CE_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')]\n",
    "CE_in_nmd_trs.rename(columns = {'upst_junction':'skip_junction'},inplace=True)\n",
    "CE_in_nmd_trs['chr'] = CE_in_nmd_trs['exon'].str.split('_',expand=True)[0]\n",
    "CE_in_nmd_trs['str'] = CE_in_nmd_trs['exon'].str.split('_',expand=True)[3]\n",
    "CE_in_nmd_trs = CE_in_nmd_trs[['gene_name','chr','str','transcript_id','stop_start','stop_end','skip_junction','exon_start','exon_end']].drop_duplicates()\n",
    "CE_in_nmd_trs.rename(columns={'exon_start':'downst_exon_start','exon_end':'downst_exon_end'},inplace=True)\n",
    "\n",
    "temp_nmd = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['downst_junction']!='')]\n",
    "temp_nmd.rename(columns = {'downst_junction':'skip_junction'},inplace=True)\n",
    "temp_nmd = temp_nmd[['gene_name','transcript_id','skip_junction','exon_start','exon_end']].drop_duplicates()\n",
    "temp_nmd.rename(columns={'exon_start':'upst_exon_start','exon_end':'upst_exon_end'},inplace=True)\n",
    "\n",
    "CE_in_nmd_trs = pd.merge(CE_in_nmd_trs,temp_nmd,how='inner',on=['gene_name','transcript_id','skip_junction'])\n",
    "CE_in_nmd_trs = pd.merge(CE_in_nmd_trs,CE_in_coding_trs,how='inner',on=['gene_name','skip_junction'])\n",
    "CE_in_nmd_trs.rename(columns= {'upst_junction':'upst_junction_cod',\n",
    "                               'downst_junction':'downst_junction_cod',\n",
    "                              'upst_exon_end':'skip_junction_start',\n",
    "                              'downst_exon_start':'skip_junction_end'},inplace=True)\n",
    "CE_in_nmd_trs['stop_len'] = CE_in_nmd_trs['stop_end']-CE_in_nmd_trs['stop_start']+1\n",
    "CE_in_nmd_trs.reset_index(drop=True,inplace=True)\n",
    "\n",
    "def get_stop_rel_position_ess(x):\n",
    "    stop_rel_position = ''\n",
    "    if (x['stop_len']<3) and (x['stop_start']<=x['skip_junction_start'] and (x['stop_start']+2)>=x['skip_junction_start']):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>=x['skip_junction_end']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<=x['skip_junction_start']:\n",
    "            stop_rel_position = 'stop_upst'          \n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>=x['skip_junction_end']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<=x['skip_junction_start']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "CE_in_nmd_trs['stop_rel_position'] = CE_in_nmd_trs.apply(lambda x:get_stop_rel_position_ess(x),1)\n",
    "essential_exons = CE_in_nmd_trs.copy()\n",
    "essential_exons = essential_exons[['chr','str','gene_name','transcript_id',\n",
    "                                   'upst_junction_cod',\n",
    "                                   'downst_junction_cod',\n",
    "                                   'skip_junction',\n",
    "                                   'stop_rel_position']].drop_duplicates().reset_index(drop=True).rename(columns={'upst_junction_cod':'upst_junction','downst_junction_cod':'downst_junction'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'stop_upst': 256, 'stop_downst': 1953, 'stop_within': 33})"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(essential_exons['stop_rel_position'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Control CE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We identify exons that are included and skipped in different protein-coding transcripts\n",
    "# excluding essential and poison exons\n",
    "CE_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(\n",
    "exons['max_exon_number']>exons['exon_number'])&(exons['exon_number']>1)]\n",
    "CE_in_coding_trs['skip_junction'] = CE_in_coding_trs['upst_junction'].str.split('_',expand=True)[0]+'_'+CE_in_coding_trs['downst_junction'].str.split('_',expand=True)[1]\n",
    "\n",
    "exons_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')]\n",
    "exons_in_coding_trs = pd.concat([exons_in_coding_trs[['upst_junction','gene_name']].drop_duplicates().rename(columns = {'upst_junction':'skip_junction'}),\n",
    "exons_in_coding_trs[['downst_junction','gene_name']].drop_duplicates().rename(columns = {'downst_junction':'skip_junction'})]).drop_duplicates().reset_index(drop=True)\n",
    "exons_in_coding_trs = exons_in_coding_trs.loc[exons_in_coding_trs['skip_junction']!='']\n",
    "exons_in_coding_trs['skip_coding_exists'] = 1\n",
    "\n",
    "CE_in_coding_trs = pd.merge(CE_in_coding_trs,exons_in_coding_trs,how='left',on=['skip_junction','gene_name'])\n",
    "CE_in_coding_trs['skip_coding_exists'] = CE_in_coding_trs['skip_coding_exists'].fillna(0).astype('int')\n",
    "\n",
    "CE_in_coding_trs = CE_in_coding_trs.loc[CE_in_coding_trs['skip_coding_exists']==1]\n",
    "\n",
    "CE_in_coding_trs['chr'] = CE_in_coding_trs['exon'].str.split('_',expand=True)[0]\n",
    "CE_in_coding_trs['str'] = CE_in_coding_trs['exon'].str.split('_',expand=True)[3]\n",
    "\n",
    "essential_exons['essential_exon']=1\n",
    "CE_in_coding_trs = pd.merge(CE_in_coding_trs,\n",
    "                            essential_exons[['chr', 'str', 'gene_name', 'upst_junction', \n",
    "                                             'downst_junction','skip_junction','essential_exon']].drop_duplicates(),\n",
    "how='left',on=['chr', 'str', 'gene_name', 'upst_junction', 'downst_junction','skip_junction'])\n",
    "essential_exons.drop('essential_exon',1,inplace=True)\n",
    "\n",
    "poison_exons['poison_exon']=1\n",
    "CE_in_coding_trs = pd.merge(CE_in_coding_trs,\n",
    "                            poison_exons[['chr', 'str', 'gene_name', 'upst_junction', \n",
    "                                             'downst_junction','skip_junction','poison_exon']].drop_duplicates(),\n",
    "how='left',on=['chr', 'str', 'gene_name', 'upst_junction', 'downst_junction','skip_junction'])\n",
    "poison_exons.drop('poison_exon',1,inplace=True)\n",
    "\n",
    "CE_in_coding_trs[['poison_exon','essential_exon']]=CE_in_coding_trs[['poison_exon','essential_exon']].fillna(0).astype('int') \n",
    "CE_in_coding_trs = CE_in_coding_trs.loc[(CE_in_coding_trs['poison_exon']==0)&(CE_in_coding_trs['essential_exon']==0)]\n",
    "\n",
    "def get_stop_rel_position(x):\n",
    "    stop_rel_position = ''\n",
    "    if (x['exon_start']<=x['stop_start'] and x['stop_start']<=x['exon_end']) or (x['exon_start']<=x['stop_end'] and x['stop_end']<=x['exon_end']):\n",
    "        stop_rel_position = 'stop_within_CE'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>x['exon_end']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<x['exon_start']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>x['exon_end']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<x['exon_start']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "CE_in_coding_trs['stop_rel_position'] = CE_in_coding_trs.apply(lambda x:get_stop_rel_position(x),1)\n",
    "\n",
    "coding_ce = CE_in_coding_trs.loc[CE_in_coding_trs['stop_rel_position']=='stop_downst'][['chr','str','gene_name',\n",
    "                                   'upst_junction',\n",
    "                                   'downst_junction',\n",
    "                                   'skip_junction','stop_rel_position']].drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13292"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(coding_ce)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_ce.to_csv('./misc/hg19_cassette_exons_coding.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alternative 5' and 3' splice sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we consider alternative ends of junctions regardeless of the strand \n",
    "# (so, disregarding the classification as 5' or 3' at this step)\n",
    "sites_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['downst_junction']!='')]\n",
    "temp_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')]\n",
    "sites_in_nmd_trs = pd.merge(sites_in_nmd_trs,\n",
    "                            temp_nmd_trs[['gene_name','transcript_id',\n",
    "                                          'exon_end','upst_junction']].rename(columns={'upst_junction':'downst_junction',\n",
    "                                                                                                        'exon_end':'flank_exon_end'}),\n",
    "how='inner',on=['gene_name','transcript_id','downst_junction'])\n",
    "\n",
    "sites_in_nmd_trs = sites_in_nmd_trs[['exon','gene_name','transcript_id','stop_start',\n",
    "                  'stop_end','exon_start','exon_end',\n",
    "                  'downst_junction','flank_exon_end']].drop_duplicates().reset_index(drop=True)\n",
    "sites_in_nmd_trs['flank_exon_start'] = sites_in_nmd_trs['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "sites_in_nmd_trs['flank_exon_end']  = sites_in_nmd_trs['flank_exon_end'].astype('int')\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.drop(['downst_junction'],1).drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "sites_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['downst_junction']!='')]\n",
    "temp_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['upst_junction']!='')]\n",
    "sites_in_coding_trs = pd.merge(sites_in_coding_trs,\n",
    "                            temp_coding_trs[['gene_name','transcript_id',\n",
    "                                          'exon_end','upst_junction']].rename(columns={'upst_junction':'downst_junction',\n",
    "                                                                                    'exon_end':'flank_exon_end'}),\n",
    "how='inner',on=['gene_name','transcript_id','downst_junction'])\n",
    "\n",
    "sites_in_coding_trs = sites_in_coding_trs[['gene_name',\n",
    "                                           'exon_start','exon_end',\n",
    "                                          'downst_junction','flank_exon_end']].drop_duplicates().reset_index(drop=True)\n",
    "sites_in_coding_trs['flank_exon_start'] = sites_in_coding_trs['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "sites_in_coding_trs['flank_exon_end']  = sites_in_coding_trs['flank_exon_end'].astype('int')\n",
    "sites_in_coding_trs = sites_in_coding_trs.drop(['downst_junction'],1).drop_duplicates().reset_index(drop=True)\n",
    "sites_in_coding_trs['coding_site_exists']=1\n",
    "\n",
    "sites_in_nmd_trs = pd.merge(sites_in_nmd_trs,\n",
    "                            sites_in_coding_trs[['gene_name','exon_start','exon_end',\n",
    "                                      'flank_exon_start','flank_exon_end','coding_site_exists']].drop_duplicates(),\n",
    "how='left',\n",
    "on=['gene_name','exon_start','exon_end','flank_exon_start','flank_exon_end'])\n",
    "sites_in_nmd_trs['coding_site_exists'] = sites_in_nmd_trs['coding_site_exists'].fillna(0).astype('int')\n",
    "\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.loc[sites_in_nmd_trs['coding_site_exists']==0].drop('coding_site_exists',1)\n",
    "\n",
    "sites_in_nmd_trs = pd.merge(sites_in_nmd_trs,\n",
    "                            sites_in_coding_trs[['gene_name','exon_start','exon_end',\n",
    "                    'flank_exon_start','flank_exon_end']].drop_duplicates().rename(columns={'exon_end':'exon_end_coding'}),\n",
    "how='left',\n",
    "on=['gene_name','exon_start','flank_exon_start','flank_exon_end'])\n",
    "\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.loc[~sites_in_nmd_trs['exon_end_coding'].isna()]\n",
    "sites_in_nmd_trs['exon_end_coding'] = sites_in_nmd_trs['exon_end_coding'].astype('int')\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.loc[sites_in_nmd_trs['exon_end_coding']!=sites_in_nmd_trs['exon_end']]\n",
    "\n",
    "\n",
    "sites_in_nmd_trs_2 = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['downst_junction']!='')]\n",
    "temp_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')]\n",
    "sites_in_nmd_trs_2 = pd.merge(sites_in_nmd_trs_2,\n",
    "                            temp_nmd_trs[['gene_name','transcript_id',\n",
    "                                          'exon_end','upst_junction']].rename(columns={'upst_junction':'downst_junction',\n",
    "                                                                                                        'exon_end':'flank_exon_end'}),\n",
    "how='inner',on=['gene_name','transcript_id','downst_junction'])\n",
    "sites_in_nmd_trs_2 = sites_in_nmd_trs_2[['gene_name',\n",
    "                                         'exon_start','exon_end',\n",
    "                              'downst_junction','flank_exon_end']].drop_duplicates().reset_index(drop=True)\n",
    "sites_in_nmd_trs_2['flank_exon_start'] = sites_in_nmd_trs_2['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "sites_in_nmd_trs_2['flank_exon_end']  = sites_in_nmd_trs_2['flank_exon_end'].astype('int')\n",
    "sites_in_nmd_trs_2 = sites_in_nmd_trs_2.drop(['downst_junction'],1).drop_duplicates().reset_index(drop=True)\n",
    "sites_in_nmd_trs_2.rename(columns={'exon_end':'exon_end_coding'},inplace=True)\n",
    "sites_in_nmd_trs_2['alt_nmd_site_exists'] = 1\n",
    "\n",
    "sites_in_nmd_trs = pd.merge(sites_in_nmd_trs,sites_in_nmd_trs_2[['gene_name','exon_end_coding','flank_exon_start','alt_nmd_site_exists']].drop_duplicates(),\n",
    "how='left',on=['gene_name','exon_end_coding','flank_exon_start'])\n",
    "sites_in_nmd_trs['alt_nmd_site_exists'] = sites_in_nmd_trs['alt_nmd_site_exists'].fillna(0).astype('int')\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.loc[sites_in_nmd_trs['alt_nmd_site_exists']==0].drop('alt_nmd_site_exists',1)\n",
    "\n",
    "sites_in_nmd_trs[['stop_start', 'stop_end',\n",
    "       'exon_start', 'exon_end', 'flank_exon_end', 'flank_exon_start',\n",
    "       'exon_end_coding']] = sites_in_nmd_trs[['stop_start', 'stop_end',\n",
    "       'exon_start', 'exon_end', 'flank_exon_end', 'flank_exon_start',\n",
    "       'exon_end_coding']].astype('int')\n",
    "\n",
    "sites_in_nmd_trs['str'] = sites_in_nmd_trs['exon'].str.split('_',expand=True)[3]\n",
    "\n",
    "def get_stop_rel_position_site(x):\n",
    "    stop_rel_position = ''\n",
    "    min_end = min([x['exon_end_coding'],x['exon_end']])\n",
    "    max_end = max([x['exon_end_coding'],x['exon_end']])\n",
    "    if (x['exon_end_coding']<x['exon_end']) and ((x['exon_end_coding']<=x['stop_start'] and x['stop_start']<=x['exon_end']) or (x['exon_end_coding']<=x['stop_end'] and x['stop_end']<=x['exon_end'])):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>=x['flank_exon_start']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<=min_end:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>=x['flank_exon_start']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<=min_end:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "sites_in_nmd_trs['stop_rel_position'] = sites_in_nmd_trs.apply(lambda x:get_stop_rel_position_site(x),1)\n",
    "alt_ends = sites_in_nmd_trs.copy()\n",
    "alt_ends['chr'] = alt_ends['exon'].str.split('_',expand=True)[0]\n",
    "alt_ends = alt_ends[['chr','str','gene_name','transcript_id',\n",
    "                                   'exon_start',\n",
    "                                   'exon_end',\n",
    "                                   'exon_end_coding',\n",
    "                                   'flank_exon_start','flank_exon_end',\n",
    "                         'stop_rel_position']].drop_duplicates()\n",
    "alt_ends = alt_ends.sort_values('stop_rel_position').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'stop_downst': 561, 'stop_upst': 159, 'stop_within': 293})"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(alt_ends['stop_rel_position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we consider protein-coding alternative ends of junctions as a control\n",
    "sites_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['downst_junction']!='')]\n",
    "temp_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['upst_junction']!='')]\n",
    "sites_in_coding_trs = pd.merge(sites_in_coding_trs,\n",
    "                            temp_coding_trs[['gene_name','transcript_id',\n",
    "                                          'exon_end','upst_junction']].rename(columns={'upst_junction':'downst_junction',\n",
    "                                                                                    'exon_end':'flank_exon_end'}),\n",
    "how='inner',on=['gene_name','transcript_id','downst_junction'])\n",
    "\n",
    "sites_in_coding_trs = sites_in_coding_trs[['exon','gene_name','transcript_id','stop_start',\n",
    "                  'stop_end','exon_start','exon_end',\n",
    "                  'downst_junction','flank_exon_end']].drop_duplicates().reset_index(drop=True)\n",
    "sites_in_coding_trs['flank_exon_start'] = sites_in_coding_trs['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "sites_in_coding_trs['flank_exon_end']  = sites_in_coding_trs['flank_exon_end'].astype('int')\n",
    "sites_in_coding_trs = sites_in_coding_trs.drop(['downst_junction'],1).drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "sites_in_coding_trs = pd.merge(sites_in_coding_trs,\n",
    "                               sites_in_coding_trs[['gene_name','exon_start',\n",
    "                     'exon_end','flank_exon_end',\n",
    "                     'flank_exon_start']].drop_duplicates().rename(columns={'exon_end':'exon_end_alt'}),\n",
    "how='left',on=['gene_name','exon_start','flank_exon_end','flank_exon_start'])\n",
    "\n",
    "sites_in_coding_trs = sites_in_coding_trs.loc[sites_in_coding_trs['exon_end']!=sites_in_coding_trs['exon_end_alt']]\n",
    "\n",
    "temp = alt_ends[['gene_name','exon_start','exon_end','exon_end_coding',\n",
    "                 'flank_exon_start','flank_exon_end']].drop_duplicates().rename(columns={'exon_end_coding':'exon_end_alt'})\n",
    "temp['nmd_event']=1\n",
    "\n",
    "sites_in_coding_trs = pd.merge(sites_in_coding_trs,temp,\n",
    "how='left',on=['gene_name','exon_start','exon_end','exon_end_alt','flank_exon_start','flank_exon_end'])\n",
    "sites_in_coding_trs['nmd_event'] = sites_in_coding_trs['nmd_event'].fillna(0).astype('int')\n",
    "sites_in_coding_trs = sites_in_coding_trs.loc[sites_in_coding_trs['nmd_event']==0]\n",
    "\n",
    "sites_in_coding_trs['str'] = sites_in_coding_trs['exon'].str.split('_',expand=True)[3]\n",
    "sites_in_coding_trs['chr'] = sites_in_coding_trs['exon'].str.split('_',expand=True)[0]\n",
    "\n",
    "def get_stop_rel_position_site(x):\n",
    "    stop_rel_position = ''\n",
    "    min_end = min([x['exon_end_alt'],x['exon_end']])\n",
    "    max_end = max([x['exon_end_alt'],x['exon_end']])\n",
    "    if ((min_end<=x['stop_start'] and x['stop_start']<=max_end) or (min_end<=x['stop_end'] and x['stop_end']<=max_end)):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>=x['flank_exon_start']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<=min_end:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>=x['flank_exon_start']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<=min_end:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "sites_in_coding_trs['stop_rel_position'] = sites_in_coding_trs.apply(lambda x:get_stop_rel_position_site(x),1)\n",
    "alt_ends_coding = sites_in_coding_trs[['chr','str','gene_name','exon_start','exon_end','exon_end_alt',\n",
    "                     'flank_exon_start','flank_exon_end','stop_rel_position']].drop_duplicates()\n",
    "alt_ends_coding['min_exon_end'] = alt_ends_coding[['exon_end','exon_end_alt']].min(1)\n",
    "alt_ends_coding['max_exon_end'] = alt_ends_coding[['exon_end','exon_end_alt']].max(1)\n",
    "alt_ends_coding = alt_ends_coding[['chr','str','exon_start','gene_name','min_exon_end','max_exon_end',\n",
    "                                       'flank_exon_start','flank_exon_end','stop_rel_position']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2564"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(alt_ends_coding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_ends_coding.to_csv('./misc/hg19_alt_ends_coding.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We consider alternative starts of junctions regardeless of the strand\n",
    "# (so, disregarding the classification as 5' or 3' at this step)\n",
    "sites_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')]\n",
    "\n",
    "temp_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['downst_junction']!='')]\n",
    "sites_in_nmd_trs = pd.merge(sites_in_nmd_trs,\n",
    "                            temp_nmd_trs[['gene_name',\n",
    "                                          'transcript_id',\n",
    "                                          'exon_start',\n",
    "                                          'downst_junction']].rename(columns={'downst_junction':'upst_junction',\n",
    "                                                                            'exon_start':'flank_exon_start'}),\n",
    "how='inner',on=['gene_name','transcript_id','upst_junction'])\n",
    "\n",
    "sites_in_nmd_trs = sites_in_nmd_trs[['exon','gene_name','transcript_id','stop_start',\n",
    "                  'stop_end','exon_start','exon_end',\n",
    "                  'upst_junction','flank_exon_start']].drop_duplicates().reset_index(drop=True)\n",
    "sites_in_nmd_trs['flank_exon_end'] = sites_in_nmd_trs['upst_junction'].str.split('_',expand=True)[0].astype('int')\n",
    "sites_in_nmd_trs['flank_exon_start']  = sites_in_nmd_trs['flank_exon_start'].astype('int')\n",
    "\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.drop(['upst_junction'],1).drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "sites_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['upst_junction']!='')]\n",
    "temp_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['downst_junction']!='')]\n",
    "sites_in_coding_trs = pd.merge(sites_in_coding_trs,\n",
    "                            temp_coding_trs[['gene_name','transcript_id',\n",
    "                                          'exon_start','downst_junction']].rename(columns={'downst_junction':'upst_junction',\n",
    "                                                                                    'exon_start':'flank_exon_start'}),\n",
    "how='inner',on=['gene_name','transcript_id','upst_junction'])\n",
    "\n",
    "sites_in_coding_trs = sites_in_coding_trs[['exon','gene_name','transcript_id','stop_start',\n",
    "                  'stop_end','exon_start','exon_end',\n",
    "                  'upst_junction','flank_exon_start']].drop_duplicates().reset_index(drop=True)\n",
    "sites_in_coding_trs['flank_exon_end'] = sites_in_coding_trs['upst_junction'].str.split('_',expand=True)[0].astype('int')\n",
    "sites_in_coding_trs['flank_exon_start']  = sites_in_coding_trs['flank_exon_start'].astype('int')\n",
    "sites_in_coding_trs = sites_in_coding_trs.drop(['upst_junction'],1).drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "sites_in_coding_trs['coding_site_exists']=1\n",
    "sites_in_nmd_trs = pd.merge(sites_in_nmd_trs,\n",
    "                            sites_in_coding_trs[['gene_name','exon_start','exon_end',\n",
    "                                      'flank_exon_start','flank_exon_end','coding_site_exists']].drop_duplicates(),\n",
    "how='left',\n",
    "on=['gene_name','exon_start','exon_end','flank_exon_start','flank_exon_end'])\n",
    "sites_in_nmd_trs['coding_site_exists'] = sites_in_nmd_trs['coding_site_exists'].fillna(0).astype('int')\n",
    "\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.loc[sites_in_nmd_trs['coding_site_exists']==0].drop('coding_site_exists',1)\n",
    "\n",
    "sites_in_nmd_trs = pd.merge(sites_in_nmd_trs,\n",
    "                            sites_in_coding_trs[['gene_name','exon_start','exon_end',\n",
    "                    'flank_exon_start','flank_exon_end']].drop_duplicates().rename(columns={'exon_start':'exon_start_coding'}),\n",
    "how='left',\n",
    "on=['gene_name','exon_end','flank_exon_start','flank_exon_end'])\n",
    "\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.loc[~sites_in_nmd_trs['exon_start_coding'].isna()]\n",
    "sites_in_nmd_trs['exon_start_coding'] = sites_in_nmd_trs['exon_start_coding'].astype('int')\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.loc[sites_in_nmd_trs['exon_start_coding']!=sites_in_nmd_trs['exon_start']]\n",
    "\n",
    "\n",
    "sites_in_nmd_trs_2 = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')]\n",
    "temp_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['downst_junction']!='')]\n",
    "sites_in_nmd_trs_2 = pd.merge(sites_in_nmd_trs_2,\n",
    "                            temp_nmd_trs[['gene_name',\n",
    "                                          'transcript_id',\n",
    "                                          'exon_start',\n",
    "                                          'downst_junction']].rename(columns={'downst_junction':'upst_junction',\n",
    "                                                                            'exon_start':'flank_exon_start'}),\n",
    "how='inner',on=['gene_name','transcript_id','upst_junction'])\n",
    "\n",
    "sites_in_nmd_trs_2 = sites_in_nmd_trs_2[['gene_name',\n",
    "                                         'exon_start','exon_end',\n",
    "                                      'upst_junction','flank_exon_start']].drop_duplicates().reset_index(drop=True)\n",
    "sites_in_nmd_trs_2['flank_exon_end'] = sites_in_nmd_trs_2['upst_junction'].str.split('_',expand=True)[0].astype('int')\n",
    "sites_in_nmd_trs_2['flank_exon_start']  = sites_in_nmd_trs_2['flank_exon_start'].astype('int')\n",
    "sites_in_nmd_trs_2 = sites_in_nmd_trs_2.drop(['upst_junction'],1).drop_duplicates().reset_index(drop=True)\n",
    "sites_in_nmd_trs_2.rename(columns = {'exon_start':'exon_start_coding'},inplace=True)\n",
    "sites_in_nmd_trs_2['alt_nmd_site_exists'] = 1\n",
    "\n",
    "sites_in_nmd_trs = pd.merge(sites_in_nmd_trs,\n",
    "                            sites_in_nmd_trs_2[['gene_name','flank_exon_end','exon_start_coding','alt_nmd_site_exists']].drop_duplicates(),\n",
    "                           how='left',on=['gene_name','flank_exon_end','exon_start_coding'])\n",
    "sites_in_nmd_trs['alt_nmd_site_exists'] = sites_in_nmd_trs['alt_nmd_site_exists'].fillna(0).astype('int')\n",
    "sites_in_nmd_trs = sites_in_nmd_trs.loc[sites_in_nmd_trs['alt_nmd_site_exists']==0].drop('alt_nmd_site_exists',1)\n",
    "\n",
    "sites_in_nmd_trs[['stop_start', 'stop_end',\n",
    "       'exon_start', 'exon_end', 'flank_exon_end', 'flank_exon_start',\n",
    "       'exon_start_coding']] = sites_in_nmd_trs[['stop_start', 'stop_end',\n",
    "       'exon_start', 'exon_end', 'flank_exon_end', 'flank_exon_start',\n",
    "       'exon_start_coding']].astype('int')\n",
    "\n",
    "sites_in_nmd_trs['str'] = sites_in_nmd_trs['exon'].str.split('_',expand=True)[3]\n",
    "\n",
    "def get_stop_rel_position_site(x):\n",
    "    stop_rel_position = ''\n",
    "    min_start = min([x['exon_start_coding'],x['exon_start']])\n",
    "    max_start = max([x['exon_start_coding'],x['exon_start']])\n",
    "    if (x['exon_start_coding']>x['exon_start']) and ((x['exon_start']<=x['stop_start'] and x['stop_start']<=x['exon_start_coding']) or (x['exon_start']<=x['stop_end'] and x['stop_end']<=x['exon_start_coding'])):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>=max_start:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<=x['flank_exon_end']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>=max_start:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<=x['flank_exon_end']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "sites_in_nmd_trs['stop_rel_position'] = sites_in_nmd_trs.apply(lambda x:get_stop_rel_position_site(x),1)\n",
    "alt_starts = sites_in_nmd_trs.copy()\n",
    "alt_starts['chr'] = alt_starts['exon'].str.split('_',expand=True)[0]\n",
    "alt_starts = alt_starts[['chr','str','gene_name','transcript_id','exon_start','exon_start_coding',\n",
    "                    'exon_end','flank_exon_start','flank_exon_end','stop_rel_position']].drop_duplicates()\n",
    "alt_starts = alt_starts.sort_values('stop_rel_position').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'stop_downst': 639, 'stop_upst': 137, 'stop_within': 309})"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(alt_starts['stop_rel_position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we consider protein-coding alternative starts of junctions as a control\n",
    "sites_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['upst_junction']!='')]\n",
    "temp_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['downst_junction']!='')]\n",
    "sites_in_coding_trs = pd.merge(sites_in_coding_trs,\n",
    "                            temp_coding_trs[['gene_name','transcript_id',\n",
    "                                          'exon_start','downst_junction']].rename(columns={'downst_junction':'upst_junction',\n",
    "                                                                                    'exon_start':'flank_exon_start'}),\n",
    "how='inner',on=['gene_name','transcript_id','upst_junction'])\n",
    "\n",
    "sites_in_coding_trs = sites_in_coding_trs[['exon','gene_name','transcript_id','stop_start',\n",
    "                  'stop_end','exon_start','exon_end',\n",
    "                  'upst_junction','flank_exon_start']].drop_duplicates().reset_index(drop=True)\n",
    "sites_in_coding_trs['flank_exon_end'] = sites_in_coding_trs['upst_junction'].str.split('_',expand=True)[0].astype('int')\n",
    "sites_in_coding_trs['flank_exon_start']  = sites_in_coding_trs['flank_exon_start'].astype('int')\n",
    "sites_in_coding_trs = sites_in_coding_trs.drop(['upst_junction'],1).drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "sites_in_coding_trs = pd.merge(sites_in_coding_trs,\n",
    "                               sites_in_coding_trs[['gene_name','exon_start',\n",
    "                     'exon_end','flank_exon_end',\n",
    "                     'flank_exon_start']].drop_duplicates().rename(columns={'exon_start':'exon_start_alt'}),\n",
    "how='left',on=['gene_name','exon_end','flank_exon_end','flank_exon_start'])\n",
    "\n",
    "sites_in_coding_trs = sites_in_coding_trs.loc[sites_in_coding_trs['exon_start']!=sites_in_coding_trs['exon_start_alt']]\n",
    "\n",
    "temp = alt_starts[['gene_name','exon_start','exon_end','exon_start_coding',\n",
    "                 'flank_exon_start','flank_exon_end']].drop_duplicates().rename(columns={'exon_start_coding':'exon_start_alt'})\n",
    "temp['nmd_event']=1\n",
    "\n",
    "sites_in_coding_trs = pd.merge(sites_in_coding_trs,temp,\n",
    "how='left',on=['gene_name','exon_start','exon_end','exon_start_alt','flank_exon_start','flank_exon_end'])\n",
    "sites_in_coding_trs['nmd_event'] = sites_in_coding_trs['nmd_event'].fillna(0).astype('int')\n",
    "sites_in_coding_trs = sites_in_coding_trs.loc[sites_in_coding_trs['nmd_event']==0]\n",
    "\n",
    "sites_in_coding_trs['str'] = sites_in_coding_trs['exon'].str.split('_',expand=True)[3]\n",
    "sites_in_coding_trs['chr'] = sites_in_coding_trs['exon'].str.split('_',expand=True)[0]\n",
    "\n",
    "def get_stop_rel_position_site(x):\n",
    "    stop_rel_position = ''\n",
    "    min_start = min([x['exon_start_alt'],x['exon_start']])\n",
    "    max_start = max([x['exon_start_alt'],x['exon_start']])\n",
    "    if ((min_start<=x['stop_start'] and x['stop_start']<=max_start) or (min_start<=x['stop_end'] and x['stop_end']<=max_start)):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>=max_start:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<=x['flank_exon_end']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>=max_start:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<=x['flank_exon_end']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "sites_in_coding_trs['stop_rel_position'] = sites_in_coding_trs.apply(lambda x:get_stop_rel_position_site(x),1)\n",
    "alt_starts_coding = sites_in_coding_trs[['chr','str','gene_name','exon_start','exon_end','exon_start_alt',\n",
    "                     'flank_exon_start','flank_exon_end','stop_rel_position']].drop_duplicates()\n",
    "alt_starts_coding['min_exon_start'] = alt_starts_coding[['exon_start','exon_start_alt']].min(1)\n",
    "alt_starts_coding['max_exon_start'] = alt_starts_coding[['exon_start','exon_start_alt']].max(1)\n",
    "alt_starts_coding = alt_starts_coding[['chr','str','gene_name','min_exon_start','max_exon_start','exon_end','flank_exon_start','flank_exon_end','stop_rel_position']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2584"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(alt_starts_coding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_starts_coding.to_csv('./misc/hg19_alt_starts_coding.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intron retention (IR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we identify introns that are specifically retained in NMD transcripts\n",
    "exons_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')]\n",
    "\n",
    "exons_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['downst_junction']!='')]\n",
    "temp = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['upst_junction']!='')]\n",
    "exons_in_coding_trs = pd.merge(exons_in_coding_trs,\n",
    "temp[['gene_name',\n",
    "      'transcript_id',\n",
    "      'exon_end',\n",
    "      'upst_junction']].drop_duplicates().rename(columns={'upst_junction':'downst_junction',\n",
    "                                                          'exon_end':'downst_exon_end'}),\n",
    "how='left',on=['gene_name','transcript_id','downst_junction'])\n",
    "exons_in_coding_trs['exon_end_coding'] = exons_in_coding_trs['exon_end']\n",
    "exons_in_coding_trs['downst_exon_start_coding'] = exons_in_coding_trs['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "\n",
    "exons_in_coding_trs = exons_in_coding_trs[['gene_name','exon_start',\n",
    "                     'exon_end_coding','downst_exon_start_coding',\n",
    "                     'downst_exon_end']].drop_duplicates().reset_index(drop=True).rename(columns={'downst_exon_end':'exon_end'})\n",
    "\n",
    "temp = exons.loc[(exons['transcript_type']=='protein_coding')][['gene_name','exon_start','exon_end']].drop_duplicates()\n",
    "temp['IR_exists_in_coding']=1\n",
    "exons_in_coding_trs = pd.merge(exons_in_coding_trs,temp,how='left',on=['gene_name','exon_start','exon_end'])\n",
    "exons_in_coding_trs['IR_exists_in_coding'] = exons_in_coding_trs['IR_exists_in_coding'].fillna(0).astype('int')\n",
    "exons_in_coding_trs = exons_in_coding_trs.loc[exons_in_coding_trs['IR_exists_in_coding']==0].drop('IR_exists_in_coding',1)\n",
    "\n",
    "exons_in_nmd_trs = pd.merge(exons_in_nmd_trs,exons_in_coding_trs,how='left',on=['gene_name','exon_start','exon_end'])\n",
    "exons_in_nmd_trs = exons_in_nmd_trs.loc[(~exons_in_nmd_trs['exon_end_coding'].isna())&(~exons_in_nmd_trs['downst_exon_start_coding'].isna())]\n",
    "exons_in_nmd_trs[['exon_end_coding','downst_exon_start_coding']] = exons_in_nmd_trs[['exon_end_coding','downst_exon_start_coding']].astype('int')\n",
    "exons_in_nmd_trs['chr'] = exons_in_nmd_trs['exon'].str.split('_',expand=True)[0]\n",
    "exons_in_nmd_trs['str'] = exons_in_nmd_trs['exon'].str.split('_',expand=True)[3]\n",
    "\n",
    "\n",
    "exons_in_nmd_trs_2 = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['downst_junction']!='')]\n",
    "temp = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')]\n",
    "exons_in_nmd_trs_2 = pd.merge(exons_in_nmd_trs_2,\n",
    "temp[['gene_name',\n",
    "      'transcript_id',\n",
    "      'exon_end',\n",
    "      'upst_junction']].drop_duplicates().rename(columns={'upst_junction':'downst_junction',\n",
    "                                                          'exon_end':'downst_exon_end'}),\n",
    "how='left',on=['gene_name','transcript_id','downst_junction'])\n",
    "exons_in_nmd_trs_2['exon_end_coding'] = exons_in_nmd_trs_2['exon_end']\n",
    "exons_in_nmd_trs_2['downst_exon_start_coding'] = exons_in_nmd_trs_2['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "\n",
    "exons_in_nmd_trs_2 = exons_in_nmd_trs_2[['gene_name','exon_start',\n",
    "                     'exon_end_coding','downst_exon_start_coding',\n",
    "                     'downst_exon_end']].drop_duplicates().reset_index(drop=True).rename(columns={'downst_exon_end':'exon_end'})\n",
    "exons_in_nmd_trs_2['ID_nmd_exists'] = 1\n",
    "\n",
    "exons_in_nmd_trs = pd.merge(exons_in_nmd_trs,exons_in_nmd_trs_2,how='left',on=['gene_name','exon_start','exon_end_coding','downst_exon_start_coding','exon_end'])\n",
    "exons_in_nmd_trs['ID_nmd_exists'] = exons_in_nmd_trs['ID_nmd_exists'].fillna(0).astype('int')\n",
    "exons_in_nmd_trs = exons_in_nmd_trs.loc[exons_in_nmd_trs['ID_nmd_exists']==0].drop('ID_nmd_exists',1)\n",
    "\n",
    "def get_stop_rel_position_intron(x):\n",
    "    stop_rel_position = ''\n",
    "    if ((x['exon_end_coding']<x['stop_start'] and x['stop_start']<x['downst_exon_start_coding']) or (x['exon_end_coding']<x['stop_end'] and x['stop_end']<x['downst_exon_start_coding'])):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>=x['downst_exon_start_coding']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<=x['exon_end_coding']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>=x['downst_exon_start_coding']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<=x['exon_end_coding']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "exons_in_nmd_trs['stop_rel_position'] = exons_in_nmd_trs.apply(lambda x:get_stop_rel_position_intron(x),1)\n",
    "\n",
    "intron_retention = exons_in_nmd_trs[['chr','str','gene_name','transcript_id','exon_start',\n",
    "                  'exon_end_coding','downst_exon_start_coding',\n",
    "                  'exon_end','stop_rel_position']].drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'stop_upst': 127, 'stop_downst': 28, 'stop_within': 68})"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(intron_retention['stop_rel_position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we identify introns that are specifically detained (spliced) in NMD transcripts\n",
    "exons_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['downst_junction']!='')]\n",
    "temp = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')]\n",
    "\n",
    "exons_in_nmd_trs = pd.merge(exons_in_nmd_trs,\n",
    "    temp[['gene_name',\n",
    "          'transcript_id',\n",
    "          'exon_end',\n",
    "          'upst_junction']].drop_duplicates().rename(columns={'upst_junction':'downst_junction',\n",
    "                                                              'exon_end':'downst_exon_end'}),\n",
    "    how='left',on=['gene_name','transcript_id','downst_junction'])\n",
    "exons_in_nmd_trs['upst_exon_end'] = exons_in_nmd_trs['exon_end']\n",
    "exons_in_nmd_trs['upst_exon_start'] = exons_in_nmd_trs['exon_start']\n",
    "exons_in_nmd_trs['downst_exon_start'] = exons_in_nmd_trs['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "\n",
    "exons_in_nmd_trs['chr'] = exons_in_nmd_trs['exon'].str.split('_',expand=True)[0]\n",
    "exons_in_nmd_trs['str'] = exons_in_nmd_trs['exon'].str.split('_',expand=True)[3]\n",
    "\n",
    "exons_in_nmd_trs = exons_in_nmd_trs[['chr','str','gene_name','transcript_id',\n",
    "                                     'upst_exon_start','upst_exon_end',\n",
    "                                     'downst_exon_start','downst_exon_end',\n",
    "                     'stop_start','stop_end']].drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "exons_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['downst_junction']!='')]\n",
    "temp = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['upst_junction']!='')]\n",
    "exons_in_coding_trs = pd.merge(exons_in_coding_trs,\n",
    "temp[['gene_name',\n",
    "      'transcript_id',\n",
    "      'exon_end',\n",
    "      'upst_junction']].drop_duplicates().rename(columns={'upst_junction':'downst_junction',\n",
    "                                                          'exon_end':'downst_exon_end'}),\n",
    "how='left',on=['gene_name','transcript_id','downst_junction'])\n",
    "exons_in_coding_trs['upst_exon_end'] = exons_in_coding_trs['exon_end']\n",
    "exons_in_coding_trs['downst_exon_start'] = exons_in_coding_trs['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "exons_in_coding_trs['upst_exon_start'] = exons_in_coding_trs['exon_start']\n",
    "\n",
    "exons_in_coding_trs = exons_in_coding_trs[['gene_name',\n",
    "                                     'upst_exon_start','upst_exon_end',\n",
    "                                     'downst_exon_start','downst_exon_end']].drop_duplicates().reset_index(drop=True)\n",
    "exons_in_coding_trs['ID_exists_in_coding'] = 1\n",
    "exons_in_nmd_trs = pd.merge(exons_in_nmd_trs,exons_in_coding_trs,how='left',on=['gene_name','upst_exon_start','upst_exon_end',\n",
    "                                                    'downst_exon_start','downst_exon_end'])\n",
    "exons_in_nmd_trs['ID_exists_in_coding'] = exons_in_nmd_trs['ID_exists_in_coding'].fillna(0).astype('int')\n",
    "exons_in_nmd_trs = exons_in_nmd_trs.loc[exons_in_nmd_trs['ID_exists_in_coding']==0]\n",
    "\n",
    "temp = exons.loc[(exons['transcript_type']=='protein_coding')]\n",
    "temp = temp[['gene_name',\n",
    "             'exon_start',\n",
    "             'exon_end']].drop_duplicates().reset_index(drop=True).rename(columns={'exon_start':'upst_exon_start',\n",
    "                                                                                   'exon_end':'downst_exon_end'})\n",
    "temp['IR_exists_in_coding']=1\n",
    "exons_in_nmd_trs = pd.merge(exons_in_nmd_trs,temp,how='left',on=['gene_name','upst_exon_start','downst_exon_end'])\n",
    "exons_in_nmd_trs['IR_exists_in_coding'] = exons_in_nmd_trs['IR_exists_in_coding'].fillna(0).astype('int')\n",
    "exons_in_nmd_trs = exons_in_nmd_trs.loc[exons_in_nmd_trs['IR_exists_in_coding']==1]\n",
    "\n",
    "temp = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')]\n",
    "temp = temp[['gene_name',\n",
    "             'exon_start',\n",
    "             'exon_end']].drop_duplicates().reset_index(drop=True).rename(columns={'exon_start':'upst_exon_start',\n",
    "                                                                                   'exon_end':'downst_exon_end'})\n",
    "temp['IR_exists_in_nmd']=1\n",
    "exons_in_nmd_trs = pd.merge(exons_in_nmd_trs,temp,how='left',on=['gene_name','upst_exon_start','downst_exon_end'])\n",
    "exons_in_nmd_trs['IR_exists_in_nmd'] = exons_in_nmd_trs['IR_exists_in_nmd'].fillna(0).astype('int')\n",
    "exons_in_nmd_trs = exons_in_nmd_trs.loc[exons_in_nmd_trs['IR_exists_in_nmd']==0]\n",
    "\n",
    "exons_in_nmd_trs['stop_len'] = exons_in_nmd_trs['stop_end']-exons_in_nmd_trs['stop_start']+1\n",
    "\n",
    "def get_stop_rel_position_intron(x):\n",
    "    stop_rel_position = ''\n",
    "    if (x['stop_len']<3) and (x['stop_start']<=x['upst_exon_end'] and (x['stop_start']+2)>=x['upst_exon_end']):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>=x['downst_exon_start']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<=x['upst_exon_end']:\n",
    "            stop_rel_position = 'stop_upst'          \n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>=x['downst_exon_start']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<=x['upst_exon_end']:\n",
    "            stop_rel_position = 'stop_downst' \n",
    "    return stop_rel_position\n",
    "\n",
    "exons_in_nmd_trs['stop_rel_position'] = exons_in_nmd_trs.apply(lambda x:get_stop_rel_position_intron(x),1)\n",
    "\n",
    "intron_detention = exons_in_nmd_trs[['chr','str','gene_name',\n",
    "                                     'transcript_id','upst_exon_start',\n",
    "                  'upst_exon_end','downst_exon_start',\n",
    "                'downst_exon_end','stop_rel_position']].drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'stop_upst': 101, 'stop_downst': 70, 'stop_within': 1})"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(intron_detention['stop_rel_position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# contol set of protein-coding IR events\n",
    "exons_in_coding_trs = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['downst_junction']!='')]\n",
    "temp = exons.loc[(exons['transcript_type']=='protein_coding')&(exons['upst_junction']!='')]\n",
    "exons_in_coding_trs = pd.merge(exons_in_coding_trs,\n",
    "temp[['gene_name',\n",
    "      'transcript_id',\n",
    "      'exon_end',\n",
    "      'upst_junction']].drop_duplicates().rename(columns={'upst_junction':'downst_junction',\n",
    "                                                          'exon_end':'downst_exon_end'}),\n",
    "how='left',on=['gene_name','transcript_id','downst_junction'])\n",
    "exons_in_coding_trs['upst_exon_end'] = exons_in_coding_trs['exon_end']\n",
    "exons_in_coding_trs['downst_exon_start'] = exons_in_coding_trs['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "exons_in_coding_trs['upst_exon_start'] = exons_in_coding_trs['exon_start']\n",
    "\n",
    "exons_in_coding_trs['chr'] = exons_in_coding_trs['exon'].str.split('_',expand=True)[0]\n",
    "exons_in_coding_trs['str'] = exons_in_coding_trs['exon'].str.split('_',expand=True)[3]\n",
    "\n",
    "temp = exons.loc[(exons['transcript_type']=='protein_coding')][['gene_name','exon_start','exon_end']].drop_duplicates()\n",
    "temp.rename(columns={'exon_start':'upst_exon_start','exon_end':'downst_exon_end'},inplace=True)\n",
    "temp['IR_exists_in_coding']=1\n",
    "exons_in_coding_trs = pd.merge(exons_in_coding_trs,temp,how='left',on=['gene_name','upst_exon_start','downst_exon_end'])\n",
    "exons_in_coding_trs['IR_exists_in_coding'] = exons_in_coding_trs['IR_exists_in_coding'].fillna(0).astype('int')\n",
    "exons_in_coding_trs = exons_in_coding_trs.loc[exons_in_coding_trs['IR_exists_in_coding']==1].drop('IR_exists_in_coding',1)\n",
    "\n",
    "temp = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')][['gene_name','exon_start','exon_end']].drop_duplicates()\n",
    "temp.rename(columns={'exon_start':'upst_exon_start','exon_end':'downst_exon_end'},inplace=True)\n",
    "temp['IR_exists_in_nmd']=1\n",
    "exons_in_coding_trs = pd.merge(exons_in_coding_trs,temp,how='left',on=['gene_name','upst_exon_start','downst_exon_end'])\n",
    "exons_in_coding_trs['IR_exists_in_nmd'] = exons_in_coding_trs['IR_exists_in_nmd'].fillna(0).astype('int')\n",
    "exons_in_coding_trs = exons_in_coding_trs.loc[exons_in_coding_trs['IR_exists_in_nmd']==0].drop('IR_exists_in_nmd',1)\n",
    "\n",
    "\n",
    "exons_in_nmd_trs = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['downst_junction']!='')]\n",
    "temp = exons.loc[(exons['transcript_type']=='nonsense_mediated_decay')&(exons['upst_junction']!='')]\n",
    "exons_in_nmd_trs = pd.merge(exons_in_nmd_trs,\n",
    "temp[['gene_name',\n",
    "      'transcript_id',\n",
    "      'exon_end',\n",
    "      'upst_junction']].drop_duplicates().rename(columns={'upst_junction':'downst_junction',\n",
    "                                                          'exon_end':'downst_exon_end'}),\n",
    "how='left',on=['gene_name','transcript_id','downst_junction'])\n",
    "exons_in_nmd_trs['upst_exon_end'] = exons_in_nmd_trs['exon_end']\n",
    "exons_in_nmd_trs['downst_exon_start'] = exons_in_nmd_trs['downst_junction'].str.split('_',expand=True)[1].astype('int')\n",
    "exons_in_nmd_trs['upst_exon_start'] = exons_in_nmd_trs['exon_start']\n",
    "exons_in_nmd_trs['ID_exists_in_nmd'] = 1\n",
    "\n",
    "exons_in_coding_trs = pd.merge(exons_in_coding_trs,\n",
    "                               exons_in_nmd_trs[['gene_name','upst_exon_end',\n",
    "                                                 'downst_exon_start','ID_exists_in_nmd']].drop_duplicates(),\n",
    "                               how='left',on=['gene_name','upst_exon_end','downst_exon_start'])\n",
    "exons_in_coding_trs['ID_exists_in_nmd'] = exons_in_coding_trs['ID_exists_in_nmd'].fillna(0).astype('int')\n",
    "exons_in_coding_trs = exons_in_coding_trs.loc[exons_in_coding_trs['ID_exists_in_nmd']==0]\n",
    "\n",
    "exons_in_coding_trs = exons_in_coding_trs[['chr','str','gene_name','transcript_id','stop_start','stop_end',\n",
    "                                           'upst_exon_start','upst_exon_end',\n",
    "                     'downst_exon_start','downst_exon_end']].drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "def get_stop_rel_position_intron(x):\n",
    "    stop_rel_position = ''\n",
    "    if ((x['upst_exon_end']<x['stop_start'] and x['stop_start']<x['downst_exon_start']) or (x['upst_exon_end']<x['stop_end'] and x['stop_end']<x['downst_exon_start'])):\n",
    "        stop_rel_position = 'stop_within'\n",
    "    elif x['str']=='+':\n",
    "        if x['stop_start']>=x['downst_exon_start']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "        elif x['stop_end']<=x['upst_exon_end']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "    elif x['str']=='-':\n",
    "        if x['stop_start']>=x['downst_exon_start']:\n",
    "            stop_rel_position = 'stop_upst'\n",
    "        elif x['stop_end']<=x['upst_exon_end']:\n",
    "            stop_rel_position = 'stop_downst'\n",
    "    return stop_rel_position\n",
    "\n",
    "exons_in_coding_trs['stop_rel_position'] = exons_in_coding_trs.apply(lambda x:get_stop_rel_position_intron(x),1)\n",
    "intron_retention_coding = exons_in_coding_trs[['chr','str','gene_name','upst_exon_start',\n",
    "                                               'upst_exon_end','downst_exon_start','downst_exon_end','stop_rel_position']].drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "758"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(intron_retention_coding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "intron_retention_coding.to_csv('./misc/hg19_intron_retention_coding.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We group identified NMD-inducing events over transcripts  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We select for each transcript the most likely events that cause NMD, \n",
    "# from the point of their position relative to the stop codon\n",
    "\n",
    "poison_exons['event']='PE'\n",
    "essential_exons['event']='EE'\n",
    "alt_ends['event'] = 'alt_end'\n",
    "alt_starts['event'] = 'alt_start'\n",
    "intron_retention['event'] = 'IR'\n",
    "intron_detention['event'] = 'ID'\n",
    "\n",
    "summary = pd.concat([poison_exons[['transcript_id','stop_rel_position','event']].drop_duplicates(),\n",
    "essential_exons[['transcript_id','stop_rel_position','event']].drop_duplicates(),\n",
    "alt_ends[['transcript_id','stop_rel_position','event']].drop_duplicates(),\n",
    "alt_starts[['transcript_id','stop_rel_position','event']].drop_duplicates(),\n",
    "intron_retention[['transcript_id','stop_rel_position','event']].drop_duplicates(),\n",
    "intron_detention[['transcript_id','stop_rel_position','event']].drop_duplicates()\n",
    "                    ])\n",
    "\n",
    "summary['t']=1\n",
    "gr = summary.groupby(['transcript_id','stop_rel_position']).agg({'t':sum}).reset_index()\n",
    "def get_rank(x):\n",
    "    rank=4\n",
    "    if x['stop_rel_position']=='stop_within':\n",
    "        rank=1\n",
    "    elif x['stop_rel_position']=='stop_downst':\n",
    "        rank=2\n",
    "    elif x['stop_rel_position']=='stop_upst':\n",
    "        rank=3\n",
    "    return rank\n",
    "gr['rank'] = gr.apply(lambda x:get_rank(x),1)\n",
    "gr = gr.sort_values(['transcript_id','rank'],ascending=[True,True]).reset_index(drop=True)\n",
    "gr = gr.drop_duplicates(['transcript_id'],keep='first')\n",
    "\n",
    "summary = pd.merge(summary,gr[['transcript_id','stop_rel_position']],how='inner',on=['transcript_id','stop_rel_position'])\n",
    "\n",
    "poison_exons = pd.merge(poison_exons,summary.loc[summary['event']=='PE'][['transcript_id','stop_rel_position']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id','stop_rel_position'])\n",
    "\n",
    "essential_exons = pd.merge(essential_exons,summary.loc[summary['event']=='EE'][['transcript_id','stop_rel_position']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id','stop_rel_position'])\n",
    "\n",
    "alt_ends = pd.merge(alt_ends,summary.loc[summary['event']=='alt_end'][['transcript_id','stop_rel_position']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id','stop_rel_position'])\n",
    "\n",
    "alt_starts = pd.merge(alt_starts,summary.loc[summary['event']=='alt_start'][['transcript_id','stop_rel_position']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id','stop_rel_position'])\n",
    "\n",
    "intron_retention = pd.merge(intron_retention,summary.loc[summary['event']=='IR'][['transcript_id','stop_rel_position']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id','stop_rel_position'])\n",
    "\n",
    "intron_detention = pd.merge(intron_detention,summary.loc[summary['event']=='ID'][['transcript_id','stop_rel_position']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id','stop_rel_position'])\n",
    "\n",
    "# Now we retain only those transcripts in which there is only one event that potentially causes NMD\n",
    "\n",
    "summary = pd.concat([poison_exons[['transcript_id','event']],\n",
    "essential_exons.drop_duplicates(['gene_name','transcript_id','skip_junction'])[['transcript_id','event']],\n",
    "alt_ends.drop_duplicates(['gene_name','transcript_id',\n",
    "                          'exon_start','exon_end',\n",
    "                          'flank_exon_start','flank_exon_end'])[['transcript_id','event']],\n",
    "alt_starts.drop_duplicates(['gene_name','transcript_id',\n",
    "                          'exon_start','exon_end',\n",
    "                          'flank_exon_start','flank_exon_end'])[['transcript_id','event']],\n",
    "intron_retention.drop_duplicates(['gene_name','transcript_id','exon_start','exon_end'])[['transcript_id','event']],\n",
    "intron_detention[['transcript_id','event']]])\n",
    "summary['t']=1\n",
    "\n",
    "gr = summary.groupby(['transcript_id']).agg({'t':sum}).reset_index()\n",
    "summary = summary.loc[summary['transcript_id'].isin(list(gr.loc[gr['t']==1]['transcript_id'].unique()))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({1: 6029, 2: 155, 3: 12, 4: 1})"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(gr['t'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "poison_exons = pd.merge(poison_exons,summary.loc[summary['event']=='PE'][['transcript_id']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id'])\n",
    "\n",
    "essential_exons = pd.merge(essential_exons,summary.loc[summary['event']=='EE'][['transcript_id']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id'])\n",
    "\n",
    "alt_ends = pd.merge(alt_ends,summary.loc[summary['event']=='alt_end'][['transcript_id']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id'])\n",
    "\n",
    "alt_starts = pd.merge(alt_starts,summary.loc[summary['event']=='alt_start'][['transcript_id']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id'])\n",
    "\n",
    "intron_retention = pd.merge(intron_retention,summary.loc[summary['event']=='IR'][['transcript_id']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id'])\n",
    "\n",
    "intron_detention = pd.merge(intron_detention,summary.loc[summary['event']=='ID'][['transcript_id']].drop_duplicates(),\n",
    "how='inner',on=['transcript_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "poison_exons = poison_exons[['chr','str','gene_name','upst_junction','downst_junction','skip_junction','stop_rel_position']].drop_duplicates()\n",
    "essential_exons = essential_exons[['chr','str','gene_name','upst_junction','downst_junction','skip_junction','stop_rel_position']].drop_duplicates()\n",
    "alt_ends = alt_ends[['chr','str','gene_name','exon_start','exon_end','exon_end_coding','flank_exon_start','flank_exon_end','stop_rel_position']].drop_duplicates()\n",
    "alt_starts = alt_starts[['chr','str','gene_name','exon_start','exon_start_coding','exon_end','flank_exon_start','flank_exon_end','stop_rel_position']].drop_duplicates()\n",
    "intron_retention = intron_retention[['chr','str','gene_name','exon_start','exon_end_coding','downst_exon_start_coding','exon_end','stop_rel_position']].drop_duplicates()\n",
    "intron_detention = intron_detention[['chr','str','gene_name','upst_exon_start','upst_exon_end','downst_exon_start','downst_exon_end','stop_rel_position']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "essential_exons = essential_exons.loc[essential_exons['stop_rel_position']!='stop_upst']\n",
    "alt_ends = alt_ends.loc[alt_ends['stop_rel_position']!='stop_upst']\n",
    "alt_starts = alt_starts.loc[alt_starts['stop_rel_position']!='stop_upst']\n",
    "intron_retention = intron_retention.loc[intron_retention['stop_rel_position']!='stop_upst']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('mkdir -p ./novel_events_catalog/')\n",
    "poison_exons.to_csv('./novel_events_catalog/hg19_poison_exons.tsv', sep=str('\\t'),header=True,index=None)\n",
    "essential_exons.to_csv('./novel_events_catalog/hg19_essential_exons.tsv', sep=str('\\t'),header=True,index=None)\n",
    "alt_ends.to_csv('./novel_events_catalog/hg19_alt_ends.tsv', sep=str('\\t'),header=True,index=None)\n",
    "alt_starts.to_csv('./novel_events_catalog/hg19_alt_starts.tsv', sep=str('\\t'),header=True,index=None)\n",
    "intron_retention.to_csv('./novel_events_catalog/hg19_intron_retention.tsv', sep=str('\\t'),header=True,index=None)\n",
    "intron_detention.to_csv('./novel_events_catalog/hg19_intron_detention.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5871"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(poison_exons)+len(essential_exons)+len(alt_ends)+len(alt_starts)+len(intron_retention)+len(intron_detention)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we check the abundance of alternative 5' (donor) and 3' (acceptor) splice sites "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_ends['DA_type'] = (alt_ends['str']=='+').astype('int').astype('str').str.replace('1','D').replace('0','A')\n",
    "alt_starts['DA_type'] = (alt_starts['str']=='+').astype('int').astype('str').str.replace('1','A').replace('0','D')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DA_type</th>\n",
       "      <th>stop_rel_position</th>\n",
       "      <th>t1</th>\n",
       "      <th>t2</th>\n",
       "      <th>t</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>stop_downst</td>\n",
       "      <td>547</td>\n",
       "      <td>572</td>\n",
       "      <td>1119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>stop_within</td>\n",
       "      <td>284</td>\n",
       "      <td>310</td>\n",
       "      <td>594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D</td>\n",
       "      <td>stop_downst</td>\n",
       "      <td>319</td>\n",
       "      <td>369</td>\n",
       "      <td>688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D</td>\n",
       "      <td>stop_within</td>\n",
       "      <td>232</td>\n",
       "      <td>213</td>\n",
       "      <td>445</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DA_type stop_rel_position   t1   t2     t\n",
       "0       A       stop_downst  547  572  1119\n",
       "1       A       stop_within  284  310   594\n",
       "2       D       stop_downst  319  369   688\n",
       "3       D       stop_within  232  213   445"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_ends['t1']=1\n",
    "temp1 = alt_ends.groupby(['DA_type','stop_rel_position']).agg({'t1':sum}).reset_index()\n",
    "alt_starts['t2']=1\n",
    "temp2 = alt_starts.groupby(['DA_type','stop_rel_position']).agg({'t2':sum}).reset_index()\n",
    "temp1 = pd.merge(temp1,temp2,how='inner',on=['DA_type','stop_rel_position'])\n",
    "temp1['t']=temp1['t1']+temp1['t2']\n",
    "temp1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DA_type</th>\n",
       "      <th>t</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>1713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D</td>\n",
       "      <td>1133</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DA_type     t\n",
       "0       A  1713\n",
       "1       D  1133"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp1.groupby('DA_type').agg({'t':sum}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'stop_upst': 68, 'stop_downst': 77, 'stop_within': 3})"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(intron_detention['stop_rel_position'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We quantify $\\psi_H-\\psi_L$, $\\Delta e_g$ and $\\Delta e_l$ using GTEx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## validated USEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "leus = pd.read_csv('./misc/Literature_examples_of_unproductive_splicing.txt',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)\n",
    "leus = leus.fillna('')\n",
    "\n",
    "cols = ['Target','AS event position','AS type','PTC position','rMATS AS type','NMD site upstream hg19','NMD site downstream hg19',\n",
    "       'NMD junction upstream hg19', 'NMD junction downstream hg19',\n",
    "       'coding site upstream hg19', 'coding site downstream hg19',\n",
    "       'coding junction upstream hg19', 'coding junction downstream hg19']\n",
    "leus_df = leus[cols].drop_duplicates().reset_index(drop=True)\n",
    "leus_df = leus_df.loc[leus_df['Target']!='']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(58, 51)"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(leus_df),len(leus_df['Target'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done: 0 out of 58\n",
      "no expression for: CCN1 intron 3\n",
      "done: 20 out of 58\n",
      "done: 40 out of 58\n"
     ]
    }
   ],
   "source": [
    "cols = pd.Series(leus_df.columns)\n",
    "NMD_cols = list(cols.loc[cols.str.contains('NMD ')])\n",
    "can_cols = list(cols.loc[cols.str.contains('coding ')])\n",
    "          \n",
    "GTEX_summary_pooled = aaf.get_GTEX_summary_pooled(leus_df,NMD_cols,can_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [],
   "source": [
    "GTEX_summary_pooled = pd.DataFrame(GTEX_summary_pooled,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])\n",
    "GTEX_summary_pooled['psi_H_L'] = GTEX_summary_pooled['psi_H']-GTEX_summary_pooled['psi_L']\n",
    "GTEX_summary_pooled = pd.merge(GTEX_summary_pooled,leus_df[['Target','AS event position','AS type']].rename(columns={'Target':'gene_name'}),\n",
    "how='left',on=['gene_name','AS event position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we need this step because the test was performed in an opposite direction\n",
    "GTEX_summary_pooled['zscore_global'] = GTEX_summary_pooled['zscore_global']*(-1)\n",
    "GTEX_summary_pooled['zscore_local'] = GTEX_summary_pooled['zscore_local']*(-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein-coding AS events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein-coding skipped exons\n",
    "coding_ce = pd.read_csv('/home/magmir/TASS/NMD_regulation/novel_events_catalog/hg19_cassette_exons_coding.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)\n",
    "coding_sample = coding_ce.copy()\n",
    "coding_sample['AS event position'] = coding_sample['chr']+'_'+coding_sample['upst_junction']+'_'+coding_sample['downst_junction']+'_'+coding_sample['str']\n",
    "coding_sample['skip_junction'] = coding_sample['chr']+'_'+coding_sample['skip_junction']+'_'+coding_sample['str']\n",
    "coding_sample['upst_junction'] = coding_sample['chr']+'_'+coding_sample['upst_junction']+'_'+coding_sample['str']\n",
    "coding_sample['downst_junction'] = coding_sample['chr']+'_'+coding_sample['downst_junction']+'_'+coding_sample['str']\n",
    "coding_sample['NMD_skip_junction'] = coding_sample['skip_junction']\n",
    "coding_sample.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "coding_sample = coding_sample.drop_duplicates(['Target','AS event position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: AAAS chr12_53708924_53709119_53709210_53709511_-\n",
      "LQ or UQ are zero length: AASDH chr4_57244630_57248643_57248763_57250236_-\n",
      "LQ or UQ are zero length: AASDH chr4_57248763_57250236_57250507_57253528_-\n",
      "LQ or UQ are zero length: ABCB7 chrX_74296489_74318777_74318896_74332721_-\n",
      "done: 20 out of 1000\n",
      "done: 40 out of 1000\n",
      "done: 60 out of 1000\n",
      "done: 80 out of 1000\n",
      "done: 100 out of 1000\n",
      "done: 140 out of 1000\n",
      "no expression for: CCN4 chr8_134225386_134232824_134233084_134237633_+\n",
      "no expression for: CCN4 chr8_134225386_134237633_134237826_134239654_+\n",
      "done: 160 out of 1000\n",
      "no expression for: CEP20 chr16_15961373_15967348_15967484_15973661_-\n",
      "no expression for: CEP20 chr16_15973745_15977865_15978062_15982415_-\n",
      "done: 180 out of 1000\n",
      "done: 200 out of 1000\n",
      "done: 220 out of 1000\n",
      "done: 240 out of 1000\n",
      "no expression for: DGLUCY chr14_91666249_91671050_91671169_91681749_+\n",
      "done: 260 out of 1000\n",
      "no expression for: DMAC2 chr19_41938313_41939177_41939339_41939442_-\n",
      "no expression for: DMAC2 chr19_41939578_41942297_41942377_41944123_-\n",
      "no expression for: DOP1A chr6_83872638_83874379_83874438_83877581_+\n",
      "done: 280 out of 1000\n",
      "done: 300 out of 1000\n",
      "done: 320 out of 1000\n",
      "done: 380 out of 1000\n",
      "no expression for: H2AZ2 chr7_44875257_44880498_44880611_44882876_-\n",
      "no expression for: H2AZ2 chr7_44880611_44882876_44882953_44887568_-\n",
      "done: 420 out of 1000\n",
      "no expression for: JPT2 chr16_1735588_1741844_1741986_1747813_+\n",
      "no expression for: MARCHF2 chr19_8491688_8495542_8495751_8503272_+\n",
      "done: 520 out of 1000\n",
      "done: 560 out of 1000\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "no expression for: PGAP6 chr16_422283_422611_422727_424005_-\n",
      "done: 680 out of 1000\n",
      "done: 720 out of 1000\n",
      "done: 740 out of 1000\n",
      "no expression for: RELCH chr18_59936181_59936644_59936745_59941213_+\n",
      "done: 760 out of 1000\n",
      "no expression for: RTF2 chr20_55043822_55047382_55047471_55048357_+\n",
      "no expression for: RTF2 chr20_55088484_55091991_55092045_55092162_+\n",
      "done: 780 out of 1000\n",
      "done: 820 out of 1000\n",
      "done: 840 out of 1000\n",
      "done: 860 out of 1000\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "no expression for: TRIR chr19_12841882_12842158_12842235_12845127_-\n",
      "done: 920 out of 1000\n",
      "done: 940 out of 1000\n",
      "no expression for: VPS35L chr16_19584563_19586420_19586444_19590374_+\n",
      "no expression for: VPS35L chr16_19586444_19590374_19590450_19592847_+\n",
      "no expression for: VPS35L chr16_19592975_19603112_19603196_19612986_+\n",
      "no expression for: VPS35L chr16_19659204_19661704_19661781_19663298_+\n",
      "done: 980 out of 1000\n",
      "0:1000 done\n",
      "no expression for: AARS1 chr16_70303666_70304099_70304243_70305684_-\n",
      "LQ or UQ are zero length: ABCC9 chr12_22028660_22035700_22035807_22040760_-\n",
      "done: 20 out of 1000\n",
      "done: 40 out of 1000\n",
      "done: 60 out of 1000\n",
      "done: 100 out of 1000\n",
      "no expression for: BUD23 chr7_73107999_73108322_73108380_73111935_+\n",
      "done: 120 out of 1000\n",
      "no expression for: CERT1 chr5_74685512_74695135_74695212_74696030_-\n",
      "no expression for: CERT1 chr5_74698883_74706836_74706928_74712701_-\n",
      "done: 180 out of 1000\n",
      "no expression for: CIDEB chr14_24774971_24775153_24775343_24775580_-\n",
      "done: 220 out of 1000\n",
      "no expression for: CRACD chr4_57189733_57190270_57190432_57193810_+\n",
      "done: 240 out of 1000\n",
      "no expression for: DARS1 chr2_136669063_136670056_136670136_136672039_-\n",
      "done: 260 out of 1000\n",
      "done: 280 out of 1000\n",
      "done: 300 out of 1000\n",
      "done: 340 out of 1000\n",
      "no expression for: GUCY1B1 chr4_156680295_156680939_156681012_156696120_+\n",
      "done: 400 out of 1000\n",
      "done: 420 out of 1000\n",
      "done: 440 out of 1000\n",
      "no expression for: LARS1 chr5_145547828_145551473_145551553_145552250_-\n",
      "no expression for: LARS1 chr5_145552337_145557110_145557228_145562051_-\n",
      "done: 460 out of 1000\n",
      "done: 480 out of 1000\n",
      "no expression for: MARCHF6 chr5_10378944_10381912_10382055_10387106_+\n",
      "no expression for: MARCHF7 chr2_160599764_160602281_160602448_160604316_+\n",
      "no expression for: MARCHF7 chr2_160619504_160621137_160621185_160623795_+\n",
      "done: 500 out of 1000\n",
      "done: 520 out of 1000\n",
      "no expression for: MYG1 chr12_53694046_53696818_53696977_53699692_+\n",
      "no expression for: MYG1 chr12_53696977_53699692_53699844_53699982_+\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "done: 600 out of 1000\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "done: 660 out of 1000\n",
      "done: 680 out of 1000\n",
      "done: 700 out of 1000\n",
      "done: 720 out of 1000\n",
      "no expression for: REXO5 chr16_20851790_20855256_20855348_20855951_+\n",
      "no expression for: RIPOR2 chr6_24843782_24847770_24847919_24848253_-\n",
      "done: 740 out of 1000\n",
      "no expression for: RTRAF chr14_52468584_52470912_52470960_52471080_+\n",
      "done: 760 out of 1000\n",
      "no expression for: SEPTIN12 chr16_4833767_4833932_4834069_4835808_-\n",
      "done: 780 out of 1000\n",
      "done: 840 out of 1000\n",
      "done: 880 out of 1000\n",
      "no expression for: TARS1 chr5_33445510_33448647_33448837_33453394_+\n",
      "done: 920 out of 1000\n",
      "no expression for: TUT7 chr9_88924204_88924378_88924491_88924843_-\n",
      "no expression for: TUT7 chr9_88940429_88941286_88941406_88943255_-\n",
      "done: 940 out of 1000\n",
      "done: 960 out of 1000\n",
      "done: 980 out of 1000\n",
      "1000:2000 done\n",
      "LQ or UQ are zero length: A1CF chr10_52619745_52622649_52622741_52623793_-\n",
      "LQ or UQ are zero length: ACE chr17_61571831_61573755_61573877_61574159_+\n",
      "done: 40 out of 1000\n",
      "done: 60 out of 1000\n",
      "no expression for: ATP5MF chr7_99055978_99056770_99056886_99057709_-\n",
      "no expression for: ATP5PD chr17_73035366_73036187_73036258_73038276_-\n",
      "no expression for: ATPSCKMT chr5_10227759_10235323_10235373_10236590_-\n",
      "no expression for: ATPSCKMT chr5_10239468_10244546_10244612_10249970_-\n",
      "done: 80 out of 1000\n",
      "no expression for: BMERB1 chr16_15675188_15677013_15677095_15680574_+\n",
      "done: 100 out of 1000\n",
      "done: 140 out of 1000\n",
      "no expression for: CFAP20DC chr3_58870435_58899434_58899590_58923374_-\n",
      "no expression for: CFAP298 chr21_33974283_33974582_33974677_33975471_-\n",
      "no expression for: CIAO2A chr15_64367736_64373301_64373350_64380886_-\n",
      "done: 180 out of 1000\n",
      "no expression for: CYBC1 chr17_80404572_80405456_80405497_80407046_-\n",
      "no expression for: CYBC1 chr17_80405497_80407046_80407168_80408576_-\n",
      "done: 220 out of 1000\n",
      "done: 260 out of 1000\n",
      "no expression for: ENTR1 chr9_139302390_139304542_139304691_139304780_-\n",
      "no expression for: FCSK chr16_70502871_70503055_70503226_70504217_+\n",
      "no expression for: FLACC1 chr2_202153504_202154180_202154289_202154419_-\n",
      "done: 300 out of 1000\n",
      "no expression for: GATD3A chr21_45557178_45560132_45560224_45563087_+\n",
      "no expression for: GOLM2 chr15_44630105_44630435_44630515_44671888_+\n",
      "no expression for: GOLM2 chr15_44673174_44695085_44695252_44705534_+\n",
      "no expression for: GRAMD2B chr5_125801237_125805369_125805480_125807930_+\n",
      "no expression for: H2AZ2 chr7_44874161_44875128_44875257_44880498_-\n",
      "no expression for: HARS1 chr5_140054410_140054602_140054718_140056239_-\n",
      "no expression for: IHO1 chr3_49278822_49281837_49281885_49282079_+\n",
      "done: 380 out of 1000\n",
      "done: 400 out of 1000\n",
      "done: 420 out of 1000\n",
      "done: 440 out of 1000\n",
      "done: 480 out of 1000\n",
      "done: 500 out of 1000\n",
      "done: 520 out of 1000\n",
      "done: 540 out of 1000\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "no expression for: NTAQ1 chr8_124442273_124448693_124448841_124449450_+\n",
      "done: 600 out of 1000\n",
      "no expression for: ODR4 chr1_186352288_186355120_186355215_186357574_+\n",
      "no expression for: ODR4 chr1_186360923_186363079_186363147_186367445_+\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "done: 660 out of 1000\n",
      "done: 700 out of 1000\n",
      "no expression for: QARS1 chr3_49133512_49135425_49135550_49135625_-\n",
      "done: 740 out of 1000\n",
      "done: 780 out of 1000\n",
      "no expression for: SHLD2 chr10_88930731_88935646_88935852_88939832_+\n",
      "no expression for: SLC66A3 chr2_11304387_11312051_11312171_11315094_+\n",
      "no expression for: SLC66A3 chr2_11312171_11315094_11315135_11317863_+\n",
      "done: 820 out of 1000\n",
      "done: 840 out of 1000\n",
      "no expression for: TBXT chr6_166574454_166575935_166576108_166578093_-\n",
      "no expression for: TENT2 chr5_78944516_78944885_78945013_78952781_+\n",
      "no expression for: TLCD3A chr17_636421_641086_641287_643745_+\n",
      "no expression for: TLCD3A chr17_636421_641086_641287_644541_+\n",
      "no expression for: TLCD3A chr17_641287_643745_643840_644541_+\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "done: 920 out of 1000\n",
      "done: 940 out of 1000\n",
      "no expression for: VXN chr8_67417763_67422331_67422392_67425775_+\n",
      "no expression for: VXN chr8_67422392_67425775_67425872_67428128_+\n",
      "done: 960 out of 1000\n",
      "done: 980 out of 1000\n",
      "2000:3000 done\n",
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: ABCF1 chr6_30548296_30550171_30550284_30550843_+\n",
      "LQ or UQ are zero length: ADGRL2 chr1_82450395_82450943_82451039_82456075_+\n",
      "done: 20 out of 1000\n",
      "done: 40 out of 1000\n",
      "done: 60 out of 1000\n",
      "done: 80 out of 1000\n",
      "done: 100 out of 1000\n",
      "done: 120 out of 1000\n",
      "done: 160 out of 1000\n",
      "no expression for: DENND2B chr11_8747756_8751497_8752756_8772168_-\n",
      "done: 220 out of 1000\n",
      "no expression for: DMAC2L chr14_50789437_50790663_50790834_50792327_+\n",
      "no expression for: DNAAF4 chr15_55722977_55724695_55724800_55727103_-\n",
      "done: 240 out of 1000\n",
      "done: 260 out of 1000\n",
      "done: 300 out of 1000\n",
      "no expression for: GATD1 chr11_772521_773522_773629_774008_-\n",
      "done: 320 out of 1000\n",
      "done: 340 out of 1000\n",
      "no expression for: HROB chr17_42232779_42235201_42235309_42239239_+\n",
      "no expression for: ITPRID2 chr2_182784173_182785324_182785389_182786675_+\n",
      "no expression for: KARS1 chr16_75675621_75678181_75678360_75681476_-\n",
      "done: 440 out of 1000\n",
      "no expression for: MARCHF10 chr17_60779119_60782900_60782942_60788592_-\n",
      "done: 460 out of 1000\n",
      "done: 480 out of 1000\n",
      "no expression for: MICOS10 chr1_19923603_19948594_19948641_19949968_+\n",
      "done: 520 out of 1000\n",
      "done: 540 out of 1000\n",
      "no expression for: NIPSNAP2 chr7_56045958_56046043_56046088_56049166_+\n",
      "no expression for: NIPSNAP2 chr7_56049260_56049903_56049973_56051421_+\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "no expression for: PALM2AKAP2 chr9_112687362_112694207_112694308_112704966_+\n",
      "done: 640 out of 1000\n",
      "no expression for: POGLUT3 chr11_108348438_108350028_108350222_108351907_-\n",
      "done: 660 out of 1000\n",
      "done: 700 out of 1000\n",
      "no expression for: RBIS chr8_86129731_86131552_86131592_86132535_-\n",
      "done: 720 out of 1000\n",
      "done: 740 out of 1000\n",
      "no expression for: SEPTIN4 chr17_56603674_56604043_56604339_56606476_-\n",
      "no expression for: SEPTIN4 chr17_56603674_56604043_56604339_56609322_-\n",
      "done: 760 out of 1000\n",
      "no expression for: SHOC1 chr9_114476903_114480464_114480568_114484689_-\n",
      "done: 840 out of 1000\n",
      "done: 880 out of 1000\n",
      "no expression for: VPS26C chr21_38604752_38605663_38605743_38610761_-\n",
      "no expression for: VPS26C chr21_38610910_38612797_38612940_38639539_-\n",
      "done: 980 out of 1000\n",
      "no expression for: ZNF875 chr19_37826159_37835512_37835676_37838092_+\n",
      "3000:4000 done\n",
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: ABHD12 chr20_25282982_25284186_25284264_25287469_-\n",
      "LQ or UQ are zero length: ABI1 chr10_27112234_27128662_27128712_27149676_-\n",
      "LQ or UQ are zero length: ABL2 chr1_179100616_179102447_179102509_179112068_-\n",
      "no expression for: AC093512.2 chr16_30064820_30066105_30066248_30066493_+\n",
      "no expression for: AC093512.2 chr16_30066248_30066493_30066648_30075050_+\n",
      "no expression for: ACP3 chr3_132051188_132056300_132056398_132061396_+\n",
      "done: 20 out of 1000\n",
      "done: 40 out of 1000\n",
      "no expression for: ATP5MD chr10_105152223_105153956_105154151_105156166_-\n",
      "no expression for: BABAM2 chr2_28113709_28115250_28115315_28117400_+\n",
      "done: 80 out of 1000\n",
      "done: 100 out of 1000\n",
      "done: 120 out of 1000\n",
      "done: 140 out of 1000\n",
      "no expression for: CIDEB chr14_24778036_24779613_24779754_24779855_-\n",
      "done: 180 out of 1000\n",
      "done: 200 out of 1000\n",
      "done: 220 out of 1000\n",
      "done: 240 out of 1000\n",
      "done: 260 out of 1000\n",
      "no expression for: EEF1AKMT3 chr12_58166911_58168412_58168550_58174038_+\n",
      "done: 280 out of 1000\n",
      "done: 300 out of 1000\n",
      "done: 320 out of 1000\n",
      "done: 340 out of 1000\n",
      "no expression for: GSDME chr7_24784373_24789183_24789412_24797326_-\n",
      "done: 360 out of 1000\n",
      "no expression for: HJV chr1_145413426_145414693_145414878_145415279_+\n",
      "done: 380 out of 1000\n",
      "done: 400 out of 1000\n",
      "done: 420 out of 1000\n",
      "done: 440 out of 1000\n",
      "no expression for: LRATD1 chr2_14773061_14773710_14773952_14774068_+\n",
      "done: 460 out of 1000\n",
      "no expression for: MARCHF11 chr5_16067902_16090998_16091190_16177835_-\n",
      "done: 480 out of 1000\n",
      "done: 500 out of 1000\n",
      "done: 520 out of 1000\n",
      "no expression for: NIBAN3 chr19_17638204_17641564_17641694_17643072_+\n",
      "no expression for: NIBAN3 chr19_17654242_17654333_17654440_17657495_+\n",
      "no expression for: NIBAN3 chr19_17657606_17660253_17660336_17662595_+\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "done: 600 out of 1000\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "done: 660 out of 1000\n",
      "no expression for: PWWP3B chrX_105415132_105421217_105421352_105447978_+\n",
      "no expression for: RAB5IF chr20_35236221_35238004_35238133_35240413_+\n",
      "done: 680 out of 1000\n",
      "no expression for: RHEX chr1_206240195_206241533_206241676_206243150_-\n",
      "done: 700 out of 1000\n",
      "done: 720 out of 1000\n",
      "done: 740 out of 1000\n",
      "no expression for: SINHCAF chr12_31448267_31451011_31451158_31478958_-\n",
      "done: 780 out of 1000\n",
      "done: 800 out of 1000\n",
      "no expression for: STING1 chr5_138856039_138856914_138857100_138857855_-\n",
      "no expression for: STING1 chr5_138860927_138861063_138861289_138861795_-\n",
      "done: 820 out of 1000\n",
      "done: 840 out of 1000\n",
      "done: 860 out of 1000\n",
      "done: 900 out of 1000\n",
      "done: 920 out of 1000\n",
      "done: 980 out of 1000\n",
      "4000:5000 done\n",
      "done: 0 out of 1000\n",
      "done: 20 out of 1000\n",
      "done: 40 out of 1000\n",
      "no expression for: ANKRD20A4P chr9_69382295_69385605_69385719_69385882_+\n",
      "done: 60 out of 1000\n",
      "done: 80 out of 1000\n",
      "no expression for: ATP5MC1 chr17_46970818_46971734_46971811_46972518_+\n",
      "done: 180 out of 1000\n",
      "done: 200 out of 1000\n",
      "no expression for: CEP43 chr6_167417889_167424301_167424381_167435897_+\n",
      "no expression for: CILK1 chr6_52906206_52925734_52925844_52926394_-\n",
      "done: 240 out of 1000\n",
      "done: 280 out of 1000\n",
      "no expression for: ENTR1 chr9_139302390_139303451_139303519_139304542_-\n",
      "done: 320 out of 1000\n",
      "done: 340 out of 1000\n",
      "done: 380 out of 1000\n",
      "done: 400 out of 1000\n",
      "no expression for: HJV chr1_145413426_145415279_145415838_145416313_+\n",
      "no expression for: IRAG2 chr12_25205890_25215754_25215834_25216654_+\n",
      "no expression for: IRAG2 chr12_25216750_25219299_25219446_25222284_+\n",
      "done: 440 out of 1000\n",
      "done: 460 out of 1000\n",
      "done: 480 out of 1000\n",
      "done: 520 out of 1000\n",
      "no expression for: MRTFA chr22_40817104_40819540_40819689_40820210_-\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "done: 600 out of 1000\n",
      "done: 620 out of 1000\n",
      "done: 680 out of 1000\n",
      "done: 700 out of 1000\n",
      "done: 720 out of 1000\n",
      "no expression for: PTGES3L-AARSD1 chr17_41122359_41123624_41123713_41131187_-\n",
      "done: 740 out of 1000\n",
      "done: 760 out of 1000\n",
      "done: 780 out of 1000\n",
      "no expression for: SEPTIN10 chr2_110301901_110303037_110303119_110303627_-\n",
      "no expression for: SEPTIN10 chr2_110343415_110350628_110350696_110371375_-\n",
      "no expression for: SEPTIN2 chr2_242255025_242256914_242257014_242263631_+\n",
      "no expression for: SEPTIN2 chr2_242283312_242285581_242285664_242287549_+\n",
      "no expression for: SEPTIN7 chr7_35840880_35872403_35872510_35903162_+\n",
      "done: 820 out of 1000\n",
      "no expression for: SLC35E2A chr1_1671143_1675690_1676107_1677163_-\n",
      "no expression for: SLC66A2 chr18_77664183_77679184_77679400_77703329_-\n",
      "done: 860 out of 1000\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "done: 920 out of 1000\n",
      "done: 940 out of 1000\n",
      "no expression for: WARS1 chr14_100828258_100835424_100835595_100842597_-\n",
      "5000:6000 done\n",
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: ABCA4 chr1_94520871_94522157_94522378_94526093_-\n",
      "LQ or UQ are zero length: ACADM chr1_76198607_76198736_76198834_76199213_+\n",
      "LQ or UQ are zero length: ACTR1A chr10_104240929_104241594_104241655_104241758_-\n",
      "no expression for: ATP5MD chr10_105152223_105153956_105154151_105155503_-\n",
      "no expression for: ATP5MD chr10_105152223_105155503_105155789_105156166_-\n",
      "no expression for: ATP5MD chr10_105154151_105155503_105155789_105156166_-\n",
      "no expression for: CALHM6 chr6_116782592_116783035_116783617_116784446_+\n",
      "no expression for: CATSPERE chr1_244624796_244627210_244627258_244640843_+\n",
      "no expression for: CATSPERE chr1_244643086_244681894_244682000_244715624_+\n",
      "no expression for: CATSPERE chr1_244773628_244780831_244780988_244798791_+\n",
      "no expression for: CEP43 chr6_167424381_167426996_167427055_167435897_+\n",
      "done: 140 out of 1000\n",
      "no expression for: COP1 chr1_176104222_176105624_176105683_176118142_-\n",
      "done: 160 out of 1000\n",
      "done: 180 out of 1000\n",
      "done: 200 out of 1000\n",
      "no expression for: DIPK1A chr1_93309752_93312740_93312916_93316405_-\n",
      "no expression for: DIPK1A chr1_93316512_93341853_93341987_93426933_-\n",
      "done: 220 out of 1000\n",
      "no expression for: EEF1AKMT2 chr10_126449072_126450869_126451127_126453961_-\n",
      "done: 240 out of 1000\n",
      "no expression for: ELP1 chr9_111631462_111637175_111637250_111640275_-\n",
      "no expression for: ELP1 chr9_111692201_111693277_111693481_111696144_-\n",
      "done: 300 out of 1000\n",
      "done: 320 out of 1000\n",
      "done: 360 out of 1000\n",
      "done: 380 out of 1000\n",
      "no expression for: HYI chr1_43917990_43918574_43918675_43919084_-\n",
      "no expression for: ILRUN chr6_34574681_34614378_34614575_34622402_-\n",
      "no expression for: KIFBP chr10_70749014_70760180_70760278_70764802_+\n",
      "no expression for: KIFBP chr10_70768681_70770656_70770771_70775297_+\n",
      "done: 460 out of 1000\n",
      "done: 480 out of 1000\n",
      "done: 500 out of 1000\n",
      "done: 520 out of 1000\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "no expression for: OGA chr10_103559212_103559999_103560157_103563492_-\n",
      "no expression for: PALM2AKAP2 chr9_112918777_112924374_112924412_112930679_+\n",
      "done: 620 out of 1000\n",
      "no expression for: PGAP4 chr9_104239451_104242953_104243007_104249231_-\n",
      "done: 640 out of 1000\n",
      "no expression for: PMF1-BGLAP chr1_156203519_156206140_156206274_156212344_+\n",
      "no expression for: PRKN chr6_162394449_162475123_162475206_162622163_-\n",
      "no expression for: PRR5-ARHGAP8 chr22_45198044_45204187_45204318_45210552_+\n",
      "done: 680 out of 1000\n",
      "no expression for: RAB5IF chr20_35236221_35238004_35238161_35240413_+\n",
      "done: 700 out of 1000\n",
      "done: 720 out of 1000\n",
      "no expression for: SARS1 chr1_109779170_109779427_109779492_109779866_+\n",
      "done: 740 out of 1000\n",
      "done: 760 out of 1000\n",
      "done: 780 out of 1000\n",
      "done: 800 out of 1000\n",
      "done: 820 out of 1000\n",
      "done: 840 out of 1000\n",
      "no expression for: TBXT chr6_166574454_166575935_166576111_166578093_-\n",
      "no expression for: TMEM273 chr10_50369667_50373250_50373328_50373811_-\n",
      "no expression for: TMEM273 chr10_50369667_50373250_50373406_50373811_-\n",
      "done: 900 out of 1000\n",
      "done: 960 out of 1000\n",
      "no expression for: YARS1 chr1_33276367_33276512_33276658_33282789_-\n",
      "no expression for: ZUP1 chr6_116973355_116977847_116978015_116979966_-\n",
      "6000:7000 done\n",
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: ABHD18 chr4_128905578_128922871_128922915_128930075_+\n",
      "LQ or UQ are zero length: ABI1 chr10_27059274_27065994_27066170_27112067_-\n",
      "no expression for: AL662899.2 chr6_31636949_31637096_31637285_31638933_+\n",
      "no expression for: ANKRD20A2P chr9_42368617_42371925_42372039_42372202_+\n",
      "done: 60 out of 1000\n",
      "no expression for: ARSL chrX_2873578_2876315_2876476_2878419_-\n",
      "done: 80 out of 1000\n",
      "done: 100 out of 1000\n",
      "done: 140 out of 1000\n",
      "no expression for: CARS1 chr11_3063486_3068983_3069231_3078573_-\n",
      "done: 160 out of 1000\n",
      "no expression for: CEMIP2 chr9_74347436_74349722_74349910_74354979_-\n",
      "done: 200 out of 1000\n",
      "done: 220 out of 1000\n",
      "done: 240 out of 1000\n",
      "done: 260 out of 1000\n",
      "no expression for: DUS4L-BCAP29 chr7_107253877_107254099_107254177_107258773_+\n",
      "done: 320 out of 1000\n",
      "done: 340 out of 1000\n",
      "no expression for: GARS1 chr7_30634759_30638412_30638513_30639563_+\n",
      "no expression for: GCOM1 chr15_57967266_58004182_58004386_58006734_+\n",
      "done: 360 out of 1000\n",
      "done: 380 out of 1000\n",
      "done: 400 out of 1000\n",
      "no expression for: JMJD7-PLA2G4B chr15_42139033_42139535_42139721_42139847_+\n",
      "done: 440 out of 1000\n",
      "done: 460 out of 1000\n",
      "no expression for: LARS1 chr5_145544034_145547691_145547828_145551473_-\n",
      "done: 480 out of 1000\n",
      "done: 500 out of 1000\n",
      "done: 520 out of 1000\n",
      "done: 540 out of 1000\n",
      "no expression for: NME1-NME2 chr17_49245704_49247298_49247410_49248848_+\n",
      "done: 580 out of 1000\n",
      "done: 600 out of 1000\n",
      "done: 620 out of 1000\n",
      "no expression for: PJVK chr2_179320878_179323237_179323354_179325075_+\n",
      "done: 640 out of 1000\n",
      "no expression for: PRXL2B chr1_2519892_2519989_2520064_2520393_+\n",
      "no expression for: PRXL2C chr9_99404168_99408175_99408306_99413672_-\n",
      "done: 680 out of 1000\n",
      "done: 700 out of 1000\n",
      "done: 720 out of 1000\n",
      "done: 760 out of 1000\n",
      "no expression for: SEPTIN2 chr2_242255397_242263631_242263656_242265408_+\n",
      "no expression for: SLC66A1 chr1_19639383_19644094_19644335_19651127_+\n",
      "done: 820 out of 1000\n",
      "done: 840 out of 1000\n",
      "no expression for: TEDC1 chr14_105958646_105958916_105959071_105960172_+\n",
      "no expression for: TEDC1 chr14_105959071_105960172_105960270_105963694_+\n",
      "no expression for: TEDC1 chr14_105960270_105962217_105962423_105963694_+\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "no expression for: TSBP1 chr6_32270406_32284171_32284191_32290263_-\n",
      "no expression for: TSBP1 chr6_32291391_32292433_32292456_32298366_-\n",
      "no expression for: TSBP1 chr6_32307415_32311015_32311035_32317517_-\n",
      "no expression for: TSBP1 chr6_32323446_32333944_32333964_32334050_-\n",
      "done: 940 out of 1000\n",
      "no expression for: WARS1 chr14_100828258_100835424_100835595_100841620_-\n",
      "done: 960 out of 1000\n",
      "done: 980 out of 1000\n",
      "no expression for: ZNF875 chr19_37825684_37826071_37826159_37838092_+\n",
      "7000:8000 done\n",
      "done: 0 out of 1000\n",
      "no expression for: AC093512.2 chr16_30066648_30070262_30070374_30075050_+\n",
      "LQ or UQ are zero length: ACY1 chr3_52019287_52019377_52019481_52019868_+\n",
      "LQ or UQ are zero length: ACY1 chr3_52021640_52022540_52022619_52022782_+\n",
      "no expression for: ADA2 chr22_17688180_17690246_17690613_17700246_-\n",
      "done: 20 out of 1000\n",
      "no expression for: AL662899.2 chr6_31636949_31638750_31638785_31638933_+\n",
      "done: 60 out of 1000\n",
      "no expression for: ATP5MPL chr14_104381526_104387277_104387545_104387807_-\n",
      "no expression for: ATP5PF chr21_27102112_27106644_27106686_27107164_-\n",
      "done: 80 out of 1000\n",
      "no expression for: CALHM4 chr6_116865035_116875354_116875514_116878988_+\n",
      "done: 120 out of 1000\n",
      "no expression for: CASTOR1 chr22_30682087_30682252_30682365_30682814_-\n",
      "no expression for: CCN6 chr6_112386200_112388172_112388235_112389408_+\n",
      "no expression for: CENPS-CORT chr1_10494747_10500404_10500470_10511434_+\n",
      "done: 160 out of 1000\n",
      "no expression for: COA8 chr14_104038157_104040444_104040507_104053611_+\n",
      "done: 180 out of 1000\n",
      "no expression for: CRPPA chr7_16341091_16348148_16348252_16415717_-\n",
      "no expression for: CRPPA chr7_16348252_16415717_16415866_16445686_-\n",
      "no expression for: CYRIA chr2_16769397_16804207_16804405_16805132_-\n",
      "no expression for: CYRIB chr8_130867993_130874475_130874580_130883621_-\n",
      "no expression for: CYRIB chr8_130891717_130908995_130909100_130915558_-\n",
      "no expression for: CYRIB chr8_130915596_130916745_130916831_130983189_-\n",
      "done: 240 out of 1000\n",
      "done: 260 out of 1000\n",
      "no expression for: EIPR1 chr2_3261226_3278110_3278190_3341788_-\n",
      "done: 280 out of 1000\n",
      "no expression for: FAM166C chr2_26785554_26794984_26795019_26798770_+\n",
      "done: 300 out of 1000\n",
      "no expression for: FBH1 chr10_5966471_5966714_5966764_5967330_+\n",
      "done: 320 out of 1000\n",
      "no expression for: GATD1 chr11_767373_770314_770398_771333_-\n",
      "done: 340 out of 1000\n",
      "done: 360 out of 1000\n",
      "no expression for: HARS1 chr5_140057338_140057493_140057600_140062685_-\n",
      "no expression for: HARS1 chr5_140057600_140062685_140062804_140070440_-\n",
      "done: 380 out of 1000\n",
      "done: 400 out of 1000\n",
      "done: 420 out of 1000\n",
      "no expression for: JPT1 chr17_73132345_73142761_73142779_73143651_-\n",
      "no expression for: MARF1 chr16_15719030_15719229_15719657_15724189_-\n",
      "done: 520 out of 1000\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "done: 600 out of 1000\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "done: 660 out of 1000\n",
      "done: 680 out of 1000\n",
      "no expression for: PWWP3A chr19_1371077_1371233_1371389_1373071_+\n",
      "done: 720 out of 1000\n",
      "no expression for: RBAK-RBAKDN chr7_5097410_5103326_5103386_5111993_+\n",
      "done: 740 out of 1000\n",
      "no expression for: RO60 chr1_193028906_193038164_193038764_193044950_+\n",
      "no expression for: RPL36A-HNRNPH2 chrX_100646810_100650323_100650445_100666924_+\n",
      "no expression for: SEPTIN2 chr2_242255397_242259618_242259702_242263631_+\n",
      "no expression for: SEPTIN2 chr2_242257014_242257652_242257759_242263631_+\n",
      "no expression for: SEPTIN2 chr2_242274627_242274737_242274766_242275390_+\n",
      "no expression for: SEPTIN3 chr22_42373078_42377649_42377840_42381988_+\n",
      "no expression for: SEPTIN7 chr7_35840880_35872408_35872510_35903162_+\n",
      "done: 800 out of 1000\n",
      "no expression for: SHFL chr19_10197976_10199787_10200054_10200333_+\n",
      "done: 820 out of 1000\n",
      "no expression for: SLC66A2 chr18_77679400_77693969_77694022_77703329_-\n",
      "done: 840 out of 1000\n",
      "done: 860 out of 1000\n",
      "done: 900 out of 1000\n",
      "done: 920 out of 1000\n",
      "done: 960 out of 1000\n",
      "done: 980 out of 1000\n",
      "8000:9000 done\n",
      "LQ or UQ are zero length: ABCA13 chr7_48494883_48506553_48506642_48511127_+\n",
      "LQ or UQ are zero length: ABCA8 chr17_66914326_66914568_66914687_66915442_-\n",
      "no expression for: AC037459.1 chr8_22449181_22451246_22451345_22458395_+\n",
      "LQ or UQ are zero length: ACKR2 chr3_42851024_42861123_42861203_42905958_+\n",
      "done: 20 out of 1000\n",
      "no expression for: AL645922.1 chr6_31907097_31910736_31910876_31911002_+\n",
      "done: 40 out of 1000\n",
      "done: 60 out of 1000\n",
      "no expression for: ARMH1 chr1_45140545_45153868_45153991_45155348_+\n",
      "done: 100 out of 1000\n",
      "done: 120 out of 1000\n",
      "no expression for: BUD23 chr7_73108380_73111796_73111846_73111935_+\n",
      "no expression for: C2CD6 chr2_202352625_202355892_202359482_202360618_-\n",
      "no expression for: CCDC198 chr14_57938308_57942300_57942422_57947313_-\n",
      "no expression for: CFAP300 chr11_101929686_101937216_101937382_101946604_+\n",
      "done: 220 out of 1000\n",
      "no expression for: CYBC1 chr17_80404572_80407046_80407168_80408576_-\n",
      "no expression for: CYBC1 chr17_80407168_80407307_80407356_80408356_-\n",
      "done: 240 out of 1000\n",
      "done: 260 out of 1000\n",
      "done: 280 out of 1000\n",
      "no expression for: ELAPOR2 chr7_86574362_86577043_86577238_86594220_-\n",
      "done: 300 out of 1000\n",
      "no expression for: FAM166C chr2_26800496_26801645_26801721_26802162_+\n",
      "done: 320 out of 1000\n",
      "no expression for: GATD1 chr11_770398_770993_771104_771333_-\n",
      "done: 360 out of 1000\n",
      "no expression for: HARS1 chr5_140057338_140057493_140057600_140058587_-\n",
      "no expression for: HARS1 chr5_140059468_140062685_140062804_140070440_-\n",
      "done: 380 out of 1000\n",
      "no expression for: IRAG1 chr11_10645403_10647513_10648130_10649263_-\n",
      "no expression for: IRAG1 chr11_10649307_10649520_10649572_10650298_-\n",
      "no expression for: IRAG1 chr11_10653608_10655515_10655618_10673572_-\n",
      "no expression for: IRAG1 chr11_10655618_10673572_10673729_10715083_-\n",
      "done: 420 out of 1000\n",
      "done: 460 out of 1000\n",
      "done: 480 out of 1000\n",
      "no expression for: MARCHF8 chr10_45956859_45958599_45959444_45959687_-\n",
      "done: 500 out of 1000\n",
      "done: 520 out of 1000\n",
      "done: 540 out of 1000\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "no expression for: ODAPH chr4_76481359_76481751_76481794_76489324_+\n",
      "done: 600 out of 1000\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "done: 660 out of 1000\n",
      "done: 680 out of 1000\n",
      "done: 700 out of 1000\n",
      "done: 720 out of 1000\n",
      "done: 740 out of 1000\n",
      "done: 760 out of 1000\n",
      "no expression for: SEPTIN10 chr2_110350696_110362412_110362510_110371375_-\n",
      "no expression for: SEPTIN4 chr17_56604339_56616473_56616569_56618027_-\n",
      "no expression for: SEPTIN9 chr17_75277636_75303223_75303279_75398141_+\n",
      "done: 780 out of 1000\n",
      "done: 800 out of 1000\n",
      "done: 820 out of 1000\n",
      "no expression for: TARS1 chr5_33448837_33452446_33452544_33453394_+\n",
      "no expression for: TASOR chr3_56667871_56672734_56672793_56674014_-\n",
      "no expression for: TENT2 chr5_78908898_78910898_78910986_78915435_+\n",
      "no expression for: TEX51 chr2_127657188_127657412_127657429_127658786_+\n",
      "no expression for: TEX51 chr2_127657188_127657512_127657595_127658786_+\n",
      "done: 860 out of 1000\n",
      "done: 880 out of 1000\n",
      "no expression for: TSBP1 chr6_32270406_32283549_32283569_32284171_-\n",
      "done: 920 out of 1000\n",
      "9000:10000 done\n",
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: AAMDC chr11_77588095_77589748_77589802_77611617_+\n",
      "LQ or UQ are zero length: ABCC4 chr13_95815901_95816632_95816772_95818412_-\n",
      "LQ or UQ are zero length: ABL2 chr1_179100616_179102447_179102509_179198376_-\n",
      "done: 20 out of 1000\n",
      "no expression for: AL353579.1 chr6_37180772_37182974_37183058_37225344_-\n",
      "done: 40 out of 1000\n",
      "done: 60 out of 1000\n",
      "done: 80 out of 1000\n",
      "done: 120 out of 1000\n",
      "done: 140 out of 1000\n",
      "no expression for: CARS1 chr11_3063486_3076107_3076269_3078573_-\n",
      "no expression for: CAVIN3 chr11_6340794_6340887_6340982_6341323_-\n",
      "no expression for: CCN4 chr8_134203456_134225107_134225386_134239654_+\n",
      "no expression for: CFAP20DC chr3_58817615_58834758_58835135_58849284_-\n",
      "no expression for: CFAP20DC chr3_58923488_58978234_58978341_59025283_-\n",
      "no expression for: CFAP299 chr4_81504337_81529487_81529537_81791147_+\n",
      "no expression for: CIBAR1 chr8_94722103_94730903_94731016_94738622_+\n",
      "done: 200 out of 1000\n",
      "no expression for: COA8 chr14_104053701_104054571_104054620_104056518_+\n",
      "done: 220 out of 1000\n",
      "no expression for: CPLANE1 chr5_37154095_37157415_37157522_37157772_-\n",
      "done: 240 out of 1000\n",
      "no expression for: CYRIB chr8_130883742_130891635_130891717_130915558_-\n",
      "no expression for: CYRIB chr8_130891717_130892622_130892704_130908918_-\n",
      "no expression for: CYRIB chr8_130891717_130908918_130909100_130915558_-\n",
      "no expression for: CYRIB chr8_130915596_130916745_130916831_130951854_-\n",
      "done: 260 out of 1000\n",
      "no expression for: DGLUCY chr14_91611657_91614125_91614202_91623983_+\n",
      "no expression for: DGLUCY chr14_91614202_91614918_91615196_91623983_+\n",
      "no expression for: DNAAF6 chrX_106449920_106450378_106450652_106456103_+\n",
      "done: 280 out of 1000\n",
      "done: 300 out of 1000\n",
      "no expression for: FYB1 chr5_39124380_39126100_39126237_39127843_-\n",
      "no expression for: GRAMD2B chr5_125801237_125801986_125802028_125807930_+\n",
      "no expression for: GUCY1A1 chr4_156588571_156589242_156589335_156617908_+\n",
      "no expression for: GUCY1B1 chr4_156680295_156680508_156680693_156680939_+\n",
      "no expression for: GUCY1B1 chr4_156681012_156693754_156693905_156696120_+\n",
      "no expression for: GUCY1B1 chr4_156698794_156706416_156706481_156710866_+\n",
      "no expression for: HJV chr1_145413426_145415279_145415315_145416313_+\n",
      "no expression for: IFTAP chr11_36616465_36619194_36619280_36631631_+\n",
      "no expression for: IGLL5 chr22_23230439_23235877_23235998_23237555_+\n",
      "no expression for: IGLL5 chr22_23230439_23235880_23235998_23237555_+\n",
      "no expression for: INSYN2A chr10_128936284_128952116_128952187_128973476_-\n",
      "no expression for: IRAG1 chr11_10653608_10655515_10655621_10673572_-\n",
      "no expression for: IRAG1 chr11_10655618_10711812_10711916_10715083_-\n",
      "done: 460 out of 1000\n",
      "done: 480 out of 1000\n",
      "no expression for: LY75-CD302 chr2_160639980_160663484_160663651_160664960_-\n",
      "no expression for: LYNX1-SLURP2 chr8_143846558_143856759_143856781_143857011_-\n",
      "no expression for: MARCHF1 chr4_164507081_164533423_164534190_164534466_-\n",
      "no expression for: MARCHF1 chr4_164775321_164909653_164909861_165032740_-\n",
      "done: 520 out of 1000\n",
      "done: 540 out of 1000\n",
      "done: 560 out of 1000\n",
      "no expression for: NDUFC2-KCTD14 chr11_77728316_77784044_77784187_77790625_-\n",
      "done: 580 out of 1000\n",
      "done: 600 out of 1000\n",
      "no expression for: PACC1 chr1_212583863_212587240_212587422_212588064_-\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "no expression for: PMF1-BGLAP chr1_156203519_156206079_156206274_156212344_+\n",
      "done: 660 out of 1000\n",
      "no expression for: PRORP chr14_35593437_35595941_35595988_35596685_+\n",
      "done: 680 out of 1000\n",
      "done: 700 out of 1000\n",
      "done: 780 out of 1000\n",
      "done: 800 out of 1000\n",
      "no expression for: STIMATE-MUSTN1 chr3_52867765_52868956_52869019_52874516_-\n",
      "done: 820 out of 1000\n",
      "done: 860 out of 1000\n",
      "no expression for: TMEM273 chr10_50369667_50373811_50373841_50374914_-\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "done: 940 out of 1000\n",
      "done: 960 out of 1000\n",
      "10000:11000 done\n",
      "no expression for: AARS1 chr16_70287941_70288294_70288386_70288524_-\n",
      "LQ or UQ are zero length: ABCA13 chr7_48494883_48506550_48506642_48511127_+\n",
      "no expression for: AC008770.2 chr19_12087979_12088186_12088246_12088931_+\n",
      "no expression for: AC093512.2 chr16_30064820_30066105_30066248_30075050_+\n",
      "no expression for: AC093512.2 chr16_30064820_30066493_30066648_30075050_+\n",
      "done: 20 out of 1000\n",
      "done: 40 out of 1000\n",
      "no expression for: ANTKMT chr16_771941_772084_772134_772309_+\n",
      "done: 60 out of 1000\n",
      "no expression for: ARSL chrX_2876476_2877644_2877929_2878419_-\n",
      "no expression for: ASDURF chr2_190526336_190528627_190528680_190530148_+\n",
      "done: 80 out of 1000\n",
      "no expression for: ATP5MC2 chr12_54066429_54067480_54067596_54069960_-\n",
      "no expression for: ATP5MPL chr14_104380321_104380816_104380970_104381403_-\n",
      "no expression for: ATP5MPL chr14_104381526_104384342_104384461_104387807_-\n",
      "done: 120 out of 1000\n",
      "no expression for: C2orf92 chr2_98286297_98292308_98292391_98315489_+\n",
      "no expression for: C2orf92 chr2_98292391_98305358_98305456_98306719_+\n",
      "no expression for: C2orf92 chr2_98315599_98317617_98317767_98319132_+\n",
      "no expression for: CCNQ chrX_152853912_152857964_152858191_152860005_-\n",
      "no expression for: CENPS-CORT chr1_10491458_10494714_10494747_10500404_+\n",
      "no expression for: CEP20 chr16_15961373_15973661_15973745_15977865_-\n",
      "no expression for: CEP20 chr16_15973745_15976806_15976877_15977865_-\n",
      "no expression for: CFAP298-TCP10L chr21_33951141_33954510_33954725_33975471_-\n",
      "no expression for: CIAO2B chr16_66966203_66966384_66966453_66966683_-\n",
      "done: 180 out of 1000\n",
      "no expression for: COA8 chr14_104040507_104045708_104045828_104053611_+\n",
      "no expression for: COA8 chr14_104040507_104047906_104047993_104053611_+\n",
      "no expression for: CRACD chr4_57164515_57167392_57167413_57173701_+\n",
      "no expression for: CYBC1 chr17_80403836_80404499_80404568_80408576_-\n",
      "no expression for: CYBC1 chr17_80403836_80404499_80404572_80408576_-\n",
      "no expression for: CYBC1 chr17_80404572_80405456_80405497_80408576_-\n",
      "no expression for: CYBC1 chr17_80407168_80407307_80407356_80408576_-\n",
      "no expression for: CYBC1 chr17_80407168_80408075_80408130_80408576_-\n",
      "done: 280 out of 1000\n",
      "no expression for: FCGR2C chr1_161562032_161565343_161565380_161569420_+\n",
      "no expression for: GCOM1 chr15_57967266_58000912_58001556_58004182_+\n",
      "done: 320 out of 1000\n",
      "no expression for: GRAMD2B chr5_125801237_125801986_125802028_125805369_+\n",
      "no expression for: GRAMD2B chr5_125802028_125805369_125805480_125807930_+\n",
      "done: 360 out of 1000\n",
      "done: 400 out of 1000\n",
      "no expression for: JPT2 chr16_1728357_1733510_1733593_1735440_+\n",
      "no expression for: MARCHF10 chr17_60821889_60824226_60824339_60837196_-\n",
      "no expression for: MARCHF10 chr17_60879113_60882261_60882408_60885438_-\n",
      "no expression for: MARCHF2 chr19_8478304_8483421_8483691_8486673_+\n",
      "no expression for: MARF1 chr16_15719030_15719229_15719654_15724189_-\n",
      "no expression for: MYG1 chr12_53696977_53699982_53700104_53700464_+\n",
      "done: 520 out of 1000\n",
      "done: 540 out of 1000\n",
      "no expression for: NIBAN3 chr19_17638204_17641564_17641625_17643072_+\n",
      "no expression for: NME1-NME2 chr17_49231023_49231586_49231805_49233012_+\n",
      "no expression for: NPY4R2 chr10_47083737_47085534_47085689_47086725_+\n",
      "done: 580 out of 1000\n",
      "no expression for: PIMREG chr17_6348724_6349415_6349507_6350783_+\n",
      "done: 640 out of 1000\n",
      "no expression for: PRORP chr14_35592116_35595941_35595988_35596685_+\n",
      "no expression for: PSME3IP1 chr16_57207781_57209451_57209502_57219733_-\n",
      "done: 660 out of 1000\n",
      "no expression for: RBIS chr8_86129731_86131465_86131592_86132535_-\n",
      "no expression for: RIPOR1 chr16_67562817_67563540_67563762_67572327_+\n",
      "done: 700 out of 1000\n",
      "no expression for: RPL17-C18orf32 chr18_47017297_47017775_47017815_47017902_-\n",
      "no expression for: RTRAF chr14_52468584_52470958_52470960_52471080_+\n",
      "done: 720 out of 1000\n",
      "done: 760 out of 1000\n",
      "done: 780 out of 1000\n",
      "done: 800 out of 1000\n",
      "done: 840 out of 1000\n",
      "done: 860 out of 1000\n",
      "done: 900 out of 1000\n",
      "no expression for: WARS1 chr14_100835595_100840473_100840602_100841620_-\n",
      "done: 960 out of 1000\n",
      "no expression for: ZNF875 chr19_37825684_37826102_37826159_37838092_+\n",
      "no expression for: ZNF875 chr19_37825684_37837325_37837394_37838092_+\n",
      "no expression for: ZNF875 chr19_37838774_37848864_37848941_37849443_+\n",
      "no expression for: ZNF875 chr19_37838774_37849443_37849563_37853011_+\n",
      "no expression for: ZNF875 chr19_37838774_37850328_37850448_37853011_+\n",
      "11000:12000 done\n",
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: ABCB11 chr2_169783872_169785802_169785861_169787175_-\n",
      "LQ or UQ are zero length: ABCB7 chrX_74332807_74334456_74334467_74334589_-\n",
      "LQ or UQ are zero length: ABCC4 chr13_95863035_95886864_95887088_95899229_-\n",
      "LQ or UQ are zero length: ABCG2 chr4_89061166_89132818_89132955_89152146_-\n",
      "LQ or UQ are zero length: ABLIM1 chr10_116233715_116241238_116241285_116247717_-\n",
      "no expression for: AC068580.4 chr11_1756659_1768897_1769349_1775033_-\n",
      "no expression for: AC104389.6 chr11_5269717_5274573_5274635_5275522_-\n",
      "no expression for: AC116366.2 chr5_131785381_131821884_131822352_131894976_+\n",
      "no expression for: AC116366.2 chr5_131785381_131892024_131892165_131894976_+\n",
      "no expression for: AC118553.2 chr1_100534142_100535170_100535241_100542721_+\n",
      "no expression for: AC244197.3 chrX_148582568_148608475_148608607_148610904_-\n",
      "no expression for: ADA2 chr22_17688180_17690246_17690488_17700246_-\n",
      "done: 20 out of 1000\n",
      "no expression for: AL590132.1 chr1_211034536_211092982_211093411_211192206_-\n",
      "no expression for: AL590132.1 chr1_211256240_211263904_211264032_211276838_-\n",
      "no expression for: ANKRD40CL chr17_48840918_48841526_48841609_48844235_-\n",
      "done: 40 out of 1000\n",
      "no expression for: AP002512.3 chr11_56230906_56233583_56233724_56237830_-\n",
      "done: 60 out of 1000\n",
      "no expression for: C20orf204 chr20_62669555_62669649_62669801_62669975_+\n",
      "no expression for: C2orf92 chr2_98286297_98287794_98287962_98290919_+\n",
      "no expression for: C2orf92 chr2_98286297_98290919_98291020_98292308_+\n",
      "no expression for: C2orf92 chr2_98292391_98306719_98306790_98315489_+\n",
      "done: 120 out of 1000\n",
      "no expression for: CALHM4 chr6_116865035_116875448_116875514_116878988_+\n",
      "done: 140 out of 1000\n",
      "no expression for: CERT1 chr5_74652837_74675244_74675298_74676897_-\n",
      "no expression for: CERT1 chr5_74681859_74685417_74685512_74696030_-\n",
      "no expression for: CFAP97D2 chr13_114965918_114977387_114977576_114987973_+\n",
      "no expression for: CKLF-CMTM1 chr16_66586696_66592093_66592251_66611007_+\n",
      "no expression for: CKLF-CMTM1 chr16_66586696_66597025_66597120_66611007_+\n",
      "done: 260 out of 1000\n",
      "no expression for: DENND10 chr10_120867676_120869705_120869826_120871361_+\n",
      "done: 280 out of 1000\n",
      "done: 300 out of 1000\n",
      "no expression for: FO681492.1 chr10_46968723_46969249_46969453_46970442_-\n",
      "no expression for: FP565260.3 chr21_45655445_45656750_45657100_45658331_-\n",
      "no expression for: FXYD6-FXYD2 chr11_117691610_117693121_117693195_117693394_-\n",
      "done: 340 out of 1000\n",
      "no expression for: GATD3B chr21_45557178_45560132_45560224_45563087_+\n",
      "no expression for: GET1 chr21_40751696_40762624_40762789_40763695_+\n",
      "done: 360 out of 1000\n",
      "no expression for: GOLM2 chr15_44581554_44611605_44611751_44615163_+\n",
      "done: 380 out of 1000\n",
      "no expression for: GSDME chr7_24784373_24789183_24789412_24797015_-\n",
      "no expression for: HDGFL2 chr19_4499716_4501203_4501329_4501923_+\n",
      "done: 400 out of 1000\n",
      "done: 420 out of 1000\n",
      "no expression for: IRAG2 chr12_25205890_25214526_25214587_25215754_+\n",
      "done: 440 out of 1000\n",
      "no expression for: KIFBP chr10_70749014_70754960_70755034_70760180_+\n",
      "no expression for: LMLN2 chr14_23570288_23570501_23570773_23570878_-\n",
      "no expression for: MTRES1 chr6_107350001_107351229_107351311_107360953_+\n",
      "done: 580 out of 1000\n",
      "no expression for: NDUFC2-KCTD14 chr11_77749804_77784044_77784187_77790625_-\n",
      "done: 600 out of 1000\n",
      "done: 640 out of 1000\n",
      "no expression for: PJVK chr2_179318347_179318832_179318932_179319059_+\n",
      "done: 680 out of 1000\n",
      "done: 720 out of 1000\n",
      "no expression for: RARS1 chr5_167913548_167915158_167915275_167915607_+\n",
      "no expression for: RBIS chr8_86126860_86127150_86127266_86129615_-\n",
      "no expression for: RBIS chr8_86129731_86131465_86131589_86132535_-\n",
      "done: 740 out of 1000\n",
      "no expression for: SEPTIN2 chr2_242255397_242255837_242255970_242263631_+\n",
      "done: 820 out of 1000\n",
      "done: 840 out of 1000\n",
      "no expression for: SPART chr13_36909969_36943011_36943127_36944226_-\n",
      "no expression for: TASOR2 chr10_5754881_5756098_5756170_5759622_+\n",
      "no expression for: TEX51 chr2_127657188_127657412_127657429_127657512_+\n",
      "no expression for: TEX51 chr2_127657429_127657512_127657595_127658786_+\n",
      "no expression for: TEX51 chr2_127658858_127658941_127658980_127659445_+\n",
      "no expression for: TMEM272 chr13_52391249_52396191_52396273_52400702_-\n",
      "no expression for: TMEM272 chr13_52391249_52400702_52400761_52412609_-\n",
      "done: 920 out of 1000\n",
      "no expression for: VSIG10L2 chr11_125821968_125823295_125823585_125823982_+\n",
      "12000:13000 done\n",
      "LQ or UQ are zero length: ABHD12 chr20_25281520_25281885_25281930_25282855_-\n",
      "LQ or UQ are zero length: ABHD12 chr20_25319987_25322582_25322732_25370562_-\n",
      "LQ or UQ are zero length: ACAT1 chr11_107992405_107998613_107998706_108002634_+\n",
      "LQ or UQ are zero length: ACAT1 chr11_107992405_108002634_108002681_108009625_+\n",
      "LQ or UQ are zero length: ACSL4 chrX_108926604_108934232_108934360_108939373_-\n",
      "LQ or UQ are zero length: ADGRL2 chr1_82266753_82277528_82277608_82302570_+\n",
      "LQ or UQ are zero length: ADK chr10_75984349_76002496_76002791_76074425_+\n",
      "no expression for: AL445524.2 chr1_231503382_231506308_231506444_231509726_-\n",
      "no expression for: AL445524.2 chr1_231598420_231663084_231663161_231664166_-\n",
      "no expression for: ARSL chrX_2878461_2882265_2882381_2886190_-\n",
      "no expression for: DMAC2L chr14_50779778_50781309_50781344_50788201_+\n",
      "no expression for: DUS4L-BCAP29 chr7_107204654_107205033_107205121_107207495_+\n",
      "no expression for: DUS4L-BCAP29 chr7_107211711_107214149_107214266_107215633_+\n",
      "no expression for: ELP1 chr9_111631462_111632734_111632833_111637175_-\n",
      "no expression for: ELP1 chr9_111692201_111693277_111693336_111696144_-\n",
      "no expression for: ELP1 chr9_111692201_111693277_111693479_111696144_-\n",
      "no expression for: GARS1 chr7_30634568_30634724_30634759_30638412_+\n",
      "no expression for: GUCY1B1 chr4_156681012_156693754_156693849_156696120_+\n",
      "no expression for: LARS1 chr5_145544034_145547691_145547828_145552250_-\n",
      "no expression for: LARS1 chr5_145544034_145551473_145551553_145552250_-\n",
      "no expression for: LARS1 chr5_145557228_145560575_145560750_145562051_-\n",
      "done: 180 out of 292\n",
      "no expression for: NFILZ chr19_8745200_8747947_8748110_8784386_+\n",
      "no expression for: SEPTIN7 chr7_35840880_35871102_35871106_35872408_+\n",
      "done: 260 out of 292\n",
      "done: 280 out of 292\n",
      "no expression for: ZNF875 chr19_37815838_37816056_37816176_37838092_+\n",
      "no expression for: ZNF875 chr19_37838774_37849449_37849563_37850328_+\n",
      "13000:14000 done\n"
     ]
    }
   ],
   "source": [
    "NMD_cols = ['NMD_skip_junction']\n",
    "can_cols = ['upst_junction','downst_junction']\n",
    "\n",
    "a = []\n",
    "s = 0\n",
    "while s<len(coding_sample):\n",
    "    e=s+1000\n",
    "    temp_df = coding_sample.iloc[s:min(e,len(coding_sample))]\n",
    "    temp_summary = aaf.get_GTEX_summary_pooled(temp_df,NMD_cols,can_cols)\n",
    "    a = a+temp_summary\n",
    "    print(str(s)+':'+str(e)+' done')\n",
    "    s=s+1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df = pd.DataFrame(a,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df.to_csv('./misc/hg19_coding_CE_skip_summary_2.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {},
   "outputs": [],
   "source": [
    "#   \n",
    "hg19_alt_starts_coding = pd.read_csv('./misc/hg19_alt_starts_coding.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)\n",
    "hg19_alt_ends_coding = pd.read_csv('./misc/hg19_alt_ends_coding.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)\n",
    "\n",
    "hg19_alt_starts_coding.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "hg19_alt_ends_coding.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "\n",
    "hg19_alt_starts_coding['AS event position'] = hg19_alt_starts_coding['chr']+'_'+hg19_alt_starts_coding['flank_exon_start'].astype('str')+\\\n",
    "'_'+hg19_alt_starts_coding['flank_exon_end'].astype('str')+'_'+hg19_alt_starts_coding['min_exon_start'].astype('str')+\\\n",
    "'_'+hg19_alt_starts_coding['max_exon_start'].astype('str')+'_'+hg19_alt_starts_coding['exon_end'].astype('str')+'_'+hg19_alt_starts_coding['str']\n",
    "\n",
    "hg19_alt_ends_coding['AS event position'] = hg19_alt_ends_coding['chr']+'_'+hg19_alt_ends_coding['exon_start'].astype('str')+\\\n",
    "'_'+hg19_alt_ends_coding['min_exon_end'].astype('str')+'_'+hg19_alt_ends_coding['max_exon_end'].astype('str')+\\\n",
    "'_'+hg19_alt_ends_coding['flank_exon_start'].astype('str')+'_'+hg19_alt_ends_coding['flank_exon_end'].astype('str')+'_'+hg19_alt_ends_coding['str']\n",
    "\n",
    "hg19_alt_starts_coding = hg19_alt_starts_coding.drop_duplicates(['Target','AS event position'])\n",
    "hg19_alt_ends_coding = hg19_alt_ends_coding.drop_duplicates(['Target','AS event position'])\n",
    "\n",
    "hg19_alt_starts_coding['DA_type'] = (hg19_alt_starts_coding['str']=='+').astype('int').astype('str').str.replace('1','A').replace('0','D')\n",
    "hg19_alt_ends_coding['DA_type'] = (hg19_alt_ends_coding['str']=='+').astype('int').astype('str').str.replace('1','D').replace('0','A')\n",
    "\n",
    "acc1 = hg19_alt_starts_coding.loc[hg19_alt_starts_coding['DA_type']=='A'].reset_index(drop=True)\n",
    "acc2 = hg19_alt_ends_coding.loc[hg19_alt_ends_coding['DA_type']=='A'].reset_index(drop=True)\n",
    "\n",
    "don1 = hg19_alt_starts_coding.loc[hg19_alt_starts_coding['DA_type']=='D'].reset_index(drop=True)\n",
    "don2 = hg19_alt_ends_coding.loc[hg19_alt_ends_coding['DA_type']=='D'].reset_index(drop=True)\n",
    "\n",
    "acc1['NMD_junction'] = acc1['chr']+'_'+acc1['flank_exon_end'].astype('str')+'_'+acc1['min_exon_start'].astype('str')+'_'+acc1['str']\n",
    "acc1['coding_junction'] = acc1['chr']+'_'+acc1['flank_exon_end'].astype('str')+'_'+acc1['max_exon_start'].astype('str')+'_'+acc1['str']\n",
    "\n",
    "acc2['NMD_junction'] = acc2['chr']+'_'+acc2['max_exon_end'].astype('str')+'_'+acc2['flank_exon_start'].astype('str')+'_'+acc2['str']\n",
    "acc2['coding_junction'] = acc2['chr']+'_'+acc2['min_exon_end'].astype('str')+'_'+acc2['flank_exon_start'].astype('str')+'_'+acc2['str']\n",
    "\n",
    "don1['NMD_junction'] = don1['chr']+'_'+don1['flank_exon_end'].astype('str')+'_'+don1['min_exon_start'].astype('str')+'_'+don1['str']\n",
    "don1['coding_junction'] = don1['chr']+'_'+don1['flank_exon_end'].astype('str')+'_'+don1['max_exon_start'].astype('str')+'_'+don1['str']\n",
    "\n",
    "don2['NMD_junction'] = don2['chr']+'_'+don2['max_exon_end'].astype('str')+'_'+don2['flank_exon_start'].astype('str')+'_'+don2['str']\n",
    "don2['coding_junction'] = don2['chr']+'_'+don2['min_exon_end'].astype('str')+'_'+don2['flank_exon_start'].astype('str')+'_'+don2['str']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LQ or UQ are zero length: ABCA13 chr7_48494634_48494883_48506550_48506553_48506642_+\n",
      "LQ or UQ are zero length: ABHD4 chr14_23070572_23070660_23072295_23072493_23072667_+\n",
      "LQ or UQ are zero length: ABI2 chr2_204260379_204260503_204261508_204261607_204261690_+\n",
      "no expression for: AC011479.1 chr19_38795015_38795130_38795205_38795264_38795326_+\n",
      "done: 40 out of 1649\n",
      "no expression for: ANKHD1-EIF4EBP3 chr5_139914947_139915123_139916923_139916974_139917174_+\n",
      "done: 60 out of 1649\n",
      "done: 80 out of 1649\n",
      "no expression for: ASDURF chr2_190528627_190528680_190530102_190530148_190530177_+\n",
      "no expression for: ATP5F1D chr19_1244096_1244184_1244314_1244318_1244439_+\n",
      "no expression for: ATP5MC1 chr17_46971734_46971811_46972518_46972545_46972696_+\n",
      "done: 120 out of 1649\n",
      "done: 140 out of 1649\n",
      "done: 160 out of 1649\n",
      "done: 180 out of 1649\n",
      "no expression for: CALHM4 chr6_116864928_116865035_116875354_116875448_116875514_+\n",
      "done: 200 out of 1649\n",
      "done: 240 out of 1649\n",
      "done: 280 out of 1649\n",
      "no expression for: CKLF-CMTM1 chr16_66611007_66611105_66612734_66612750_66613038_+\n",
      "done: 320 out of 1649\n",
      "done: 360 out of 1649\n",
      "done: 380 out of 1649\n",
      "no expression for: DOP1A chr6_83828607_83828706_83829467_83829494_83829576_+\n",
      "done: 400 out of 1649\n",
      "done: 420 out of 1649\n",
      "no expression for: EOLA1 chrX_148626914_148627046_148627148_148627314_148627429_+\n",
      "done: 440 out of 1649\n",
      "no expression for: FCSK chr16_70497526_70497677_70498933_70499029_70499079_+\n",
      "no expression for: FCSK chr16_70504874_70504975_70505085_70505101_70505255_+\n",
      "no expression for: FCSK chr16_70506886_70507256_70508032_70508038_70508242_+\n",
      "done: 480 out of 1649\n",
      "done: 500 out of 1649\n",
      "done: 520 out of 1649\n",
      "no expression for: GATD3A chr21_45560132_45560224_45560824_45561046_45561299_+\n",
      "done: 540 out of 1649\n",
      "done: 560 out of 1649\n",
      "no expression for: GRAMD2B chr5_125819155_125819266_125820091_125820094_125820216_+\n",
      "no expression for: GUCY1A1 chr4_156588498_156588571_156617908_156617958_156618274_+\n",
      "no expression for: GUCY1A1 chr4_156588498_156588571_156617908_156618087_156618274_+\n",
      "no expression for: GUCY1A1 chr4_156588498_156588571_156617958_156618087_156618274_+\n",
      "no expression for: GUCY1B1 chr4_156721029_156721226_156723365_156723494_156723731_+\n",
      "no expression for: GUCY1B1 chr4_156721029_156721226_156723365_156723593_156723731_+\n",
      "no expression for: GUCY1B1 chr4_156721029_156721226_156723494_156723593_156723731_+\n",
      "no expression for: H2BW2 chrX_103295395_103295453_103295655_103296679_103297021_+\n",
      "no expression for: HDGFL2 chr19_4501203_4501329_4501923_4501926_4502219_+\n",
      "no expression for: HROB chr17_42230005_42230145_42231921_42231924_42232032_+\n",
      "done: 620 out of 1649\n",
      "done: 640 out of 1649\n",
      "no expression for: IRAG2 chr12_25215754_25215834_25216654_25216657_25216750_+\n",
      "done: 660 out of 1649\n",
      "no expression for: KASH5 chr19_49912453_49912540_49913010_49913016_49913132_+\n",
      "no expression for: KIFBP chr10_70760180_70760278_70764802_70764853_70764881_+\n",
      "no expression for: KIRREL1 chr1_158058117_158058244_158059244_158059292_158059418_+\n",
      "done: 700 out of 1649\n",
      "done: 720 out of 1649\n",
      "done: 760 out of 1649\n",
      "done: 780 out of 1649\n",
      "done: 820 out of 1649\n",
      "done: 840 out of 1649\n",
      "done: 880 out of 1649\n",
      "no expression for: NFILZ chr19_8784386_8784435_8786194_8786200_8786291_+\n",
      "no expression for: NKAPD1 chr11_111952706_111952759_111953189_111953192_111955874_+\n",
      "no expression for: NOTCH2NLC chr1_145209380_145209436_145248813_145248821_145248894_+\n",
      "done: 940 out of 1649\n",
      "done: 980 out of 1649\n",
      "done: 1000 out of 1649\n",
      "no expression for: PMF1-BGLAP chr1_156203419_156203519_156206079_156206140_156206274_+\n",
      "done: 1040 out of 1649\n",
      "done: 1060 out of 1649\n",
      "no expression for: PRR5-ARHGAP8 chr22_45244812_45244940_45255560_45255611_45255714_+\n",
      "done: 1080 out of 1649\n",
      "no expression for: PRXL2B chr1_2519180_2519231_2519775_2519829_2519892_+\n",
      "no expression for: PRXL2B chr1_2519829_2519892_2520393_2520396_2520479_+\n",
      "no expression for: PRXL2B chr1_2519989_2520064_2520361_2520393_2520479_+\n",
      "done: 1100 out of 1649\n",
      "done: 1120 out of 1649\n",
      "done: 1140 out of 1649\n",
      "done: 1160 out of 1649\n",
      "no expression for: REX1BD chr19_18699805_18699887_18700223_18700289_18700493_+\n",
      "no expression for: REX1BD chr19_18701664_18701743_18702899_18702918_18703146_+\n",
      "no expression for: RIPOR1 chr16_67572327_67572453_67572563_67572575_67572727_+\n",
      "no expression for: RIPOR1 chr16_67575770_67577167_67577251_67577254_67577372_+\n",
      "no expression for: RMC1 chr18_21110193_21110261_21110351_21110475_21110576_+\n",
      "no expression for: RTRAF chr14_52468516_52468584_52470912_52470958_52470960_+\n",
      "done: 1240 out of 1649\n",
      "no expression for: SEPTIN11 chr4_77951936_77952123_77957923_77957926_77957990_+\n",
      "no expression for: SEPTIN5 chr22_19709163_19709259_19709345_19709356_19709480_+\n",
      "done: 1260 out of 1649\n",
      "no expression for: SHFL chr19_10200333_10200371_10200574_10200605_10200723_+\n",
      "no expression for: SHFL chr19_10201926_10202029_10202152_10202260_10202306_+\n",
      "done: 1280 out of 1649\n",
      "done: 1300 out of 1649\n",
      "done: 1320 out of 1649\n",
      "no expression for: SPEM2 chr17_7329312_7329369_7329507_7329845_7330887_+\n",
      "no expression for: TAFA3 chr1_113265638_113265787_113266567_113266635_113266759_+\n",
      "no expression for: TENT2 chr5_78936928_78937019_78938655_78938667_78938733_+\n",
      "no expression for: TEX51 chr2_127658941_127658980_127659445_127659448_127659673_+\n",
      "no expression for: TMEM131L chr4_154504716_154504824_154505987_154505990_154506083_+\n",
      "done: 1460 out of 1649\n",
      "done: 1480 out of 1649\n",
      "done: 1500 out of 1649\n",
      "done: 1520 out of 1649\n",
      "done: 1540 out of 1649\n",
      "done: 1560 out of 1649\n",
      "done: 1600 out of 1649\n",
      "done: 1620 out of 1649\n",
      "no expression for: ZNF875 chr19_37838679_37838774_37849443_37849449_37849563_+\n",
      "done: 0 out of 1602\n",
      "no expression for: AARS1 chr16_70291936_70292120_70292147_70292883_70293089_-\n",
      "no expression for: AARS1 chr16_70302174_70302282_70302298_70303521_70303666_-\n",
      "no expression for: AARS1 chr16_70305684_70305791_70305875_70310389_70310534_-\n",
      "LQ or UQ are zero length: ABCA2 chr9_139902868_139903047_139903074_139903183_139903320_-\n",
      "LQ or UQ are zero length: ABCA8 chr17_66878740_66878845_66878860_66879932_66880023_-\n",
      "LQ or UQ are zero length: ABCB7 chrX_74280058_74280103_74280165_74282163_74282266_-\n",
      "LQ or UQ are zero length: ABCB9 chr12_123413547_123414718_123414727_123416739_123416875_-\n",
      "done: 20 out of 1602\n",
      "done: 40 out of 1602\n",
      "done: 60 out of 1602\n",
      "done: 80 out of 1602\n",
      "done: 100 out of 1602\n",
      "done: 140 out of 1602\n",
      "no expression for: ATP5MC2 chr12_54063655_54063732_54063898_54066360_54066429_-\n",
      "no expression for: ATP5MF chr7_99057709_99057816_99057846_99063752_99063790_-\n",
      "done: 180 out of 1602\n",
      "done: 220 out of 1602\n",
      "no expression for: CCDC198 chr14_57948277_57948375_57948378_57949783_57949869_-\n",
      "done: 240 out of 1602\n",
      "done: 260 out of 1602\n",
      "done: 280 out of 1602\n",
      "no expression for: CERT1 chr5_74664311_74664483_74670016_74675162_74675298_-\n",
      "no expression for: CERT1 chr5_74696030_74696116_74696122_74698797_74698883_-\n",
      "no expression for: CFAP20DC chr3_58853543_58853623_58853662_58855054_58855216_-\n",
      "no expression for: CFAP410 chr21_45750706_45750799_45750802_45751726_45751897_-\n",
      "done: 320 out of 1602\n",
      "no expression for: CPLANE1 chr5_37157772_37157916_37157970_37158326_37158447_-\n",
      "done: 360 out of 1602\n",
      "done: 380 out of 1602\n",
      "no expression for: DENND11 chr7_141366087_141366186_141366225_141373867_141374020_-\n",
      "done: 400 out of 1602\n",
      "done: 420 out of 1602\n",
      "no expression for: ELAPOR2 chr7_86542257_86542381_86542520_86544039_86544176_-\n",
      "done: 440 out of 1602\n",
      "no expression for: ELP1 chr9_111693277_111693479_111693481_111696144_111696393_-\n",
      "done: 460 out of 1602\n",
      "done: 480 out of 1602\n",
      "done: 500 out of 1602\n",
      "no expression for: FLACC1 chr2_202154419_202154508_202154577_202163961_202164023_-\n",
      "done: 520 out of 1602\n",
      "no expression for: GATD1 chr11_771333_771426_771467_772427_772521_-\n",
      "done: 560 out of 1602\n",
      "no expression for: GSDME chr7_24746898_24747041_24747044_24747746_24747873_-\n",
      "done: 580 out of 1602\n",
      "done: 600 out of 1602\n",
      "done: 620 out of 1602\n",
      "done: 640 out of 1602\n",
      "no expression for: IQCN chr19_18375734_18378336_18378375_18380276_18380397_-\n",
      "no expression for: IRAG1 chr11_10649263_10649307_10649310_10649520_10649572_-\n",
      "no expression for: LARS1 chr5_145508235_145508342_145508345_145508541_145508681_-\n",
      "done: 720 out of 1602\n",
      "no expression for: LRRC24 chr8_145749494_145749653_145749662_145749825_145750103_-\n",
      "no expression for: MACROH2A1 chr5_134696187_134696294_134696297_134705096_134705293_-\n",
      "done: 740 out of 1602\n",
      "no expression for: MARCHF10 chr17_60821737_60821886_60821889_60837196_60837367_-\n",
      "no expression for: MARF1 chr16_15716791_15716914_15716980_15718626_15718784_-\n",
      "no expression for: MARF1 chr16_15716791_15716914_15716983_15718626_15718784_-\n",
      "no expression for: MARF1 chr16_15716791_15716980_15716983_15718626_15718784_-\n",
      "no expression for: MARF1 chr16_15719229_15719654_15719657_15724189_15724361_-\n",
      "no expression for: MARF1 chr16_15727474_15727691_15727700_15728614_15728788_-\n",
      "done: 800 out of 1602\n",
      "done: 840 out of 1602\n",
      "done: 860 out of 1602\n",
      "no expression for: OGA chr10_103552596_103552658_103552700_103553670_103553755_-\n",
      "done: 920 out of 1602\n",
      "done: 940 out of 1602\n",
      "done: 960 out of 1602\n",
      "done: 980 out of 1602\n",
      "done: 1020 out of 1602\n",
      "done: 1060 out of 1602\n",
      "done: 1080 out of 1602\n",
      "no expression for: PYCR3 chr8_144688637_144688789_144688849_144689123_144689302_-\n",
      "done: 1100 out of 1602\n",
      "done: 1120 out of 1602\n",
      "no expression for: RIPOR2 chr6_24839587_24839827_24841003_24843090_24843782_-\n",
      "no expression for: RUSF1 chr16_31504924_31505102_31505108_31505187_31505257_-\n",
      "done: 1160 out of 1602\n",
      "no expression for: SEPTIN4 chr17_56598104_56598150_56598203_56598359_56598521_-\n",
      "no expression for: SEPTIN8 chr5_132099398_132099578_132099584_132099916_132100111_-\n",
      "done: 1200 out of 1602\n",
      "done: 1260 out of 1602\n",
      "no expression for: SLC49A3 chr4_678272_678388_678391_678508_678645_-\n",
      "done: 1280 out of 1602\n",
      "done: 1300 out of 1602\n",
      "done: 1320 out of 1602\n",
      "done: 1340 out of 1602\n",
      "no expression for: TASOR chr3_56667072_56667871_56667994_56672734_56672793_-\n",
      "no expression for: TBXT chr6_166575935_166576108_166576111_166578093_166578154_-\n",
      "no expression for: TEX48 chr9_117433663_117433739_117433785_117434000_117434107_-\n",
      "done: 1400 out of 1602\n",
      "no expression for: TMEM273 chr10_50373250_50373328_50373406_50373811_50373841_-\n",
      "done: 1440 out of 1602\n",
      "no expression for: TSBP1 chr6_32303689_32303709_32303711_32304392_32304412_-\n",
      "no expression for: TSBP1 chr6_32317517_32317537_32317606_32322901_32322921_-\n",
      "done: 1480 out of 1602\n",
      "no expression for: TUT4 chr1_52902427_52902647_52902650_52903892_52904007_-\n",
      "done: 1520 out of 1602\n",
      "done: 1540 out of 1602\n",
      "LQ or UQ are zero length: AADAT chr4_171009547_171009715_171010763_171010775_171010887_-\n",
      "no expression for: AARS1 chr16_70287171_70287284_70287569_70287617_70287703_-\n",
      "no expression for: AARS1 chr16_70294947_70295060_70296249_70296270_70296427_-\n",
      "no expression for: AARS1 chr16_70294947_70295060_70296249_70296276_70296427_-\n",
      "no expression for: AARS1 chr16_70294947_70295060_70296270_70296276_70296427_-\n",
      "no expression for: AC099489.1 chr16_11484136_11484320_11485523_11485641_11485718_-\n",
      "done: 20 out of 824\n",
      "done: 40 out of 824\n",
      "no expression for: ATP5F1A chr18_43669532_43669698_43669789_43669855_43669962_-\n",
      "no expression for: ATP5MF chr7_99057709_99057816_99063734_99063752_99063790_-\n",
      "no expression for: ATPSCKMT chr5_10236590_10236727_10239154_10239179_10239468_-\n",
      "no expression for: BCLAF3 chrX_19953927_19954016_19955536_19955623_19955650_-\n",
      "done: 80 out of 824\n",
      "no expression for: CARS1 chr11_3022159_3022478_3023192_3023200_3023283_-\n",
      "no expression for: CCDC32 chr15_40854971_40855226_40856990_40857175_40857256_-\n",
      "no expression for: CFAP410 chr21_45749774_45750209_45750346_45750706_45750799_-\n",
      "done: 140 out of 824\n",
      "no expression for: CILK1 chr6_52876835_52877025_52878439_52878460_52878780_-\n",
      "done: 160 out of 824\n",
      "no expression for: COP1 chr1_176132005_176132124_176132951_176132963_176133027_-\n",
      "done: 180 out of 824\n",
      "no expression for: CYREN chr7_134852484_134852559_134853497_134853501_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853497_134853538_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853497_134853683_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853497_134853721_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853501_134853538_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853501_134853683_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853501_134853721_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853538_134853683_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853538_134853721_134853812_-\n",
      "no expression for: CYREN chr7_134852484_134852559_134853683_134853721_134853812_-\n",
      "no expression for: CYRIB chr8_130891635_130891717_130908918_130908995_130909100_-\n",
      "done: 220 out of 824\n",
      "no expression for: ELP1 chr9_111642332_111642445_111643943_111643985_111644045_-\n",
      "done: 240 out of 824\n",
      "no expression for: ENTR1 chr9_139299083_139299145_139299543_139299570_139299654_-\n",
      "no expression for: EOLA2 chrX_149102223_149102355_149105632_149105651_149105698_-\n",
      "done: 260 out of 824\n",
      "no expression for: GAB4 chr22_17449188_17449273_17450503_17450833_17451083_-\n",
      "done: 300 out of 824\n",
      "no expression for: H2AZ1 chr4_100869245_100869680_100869968_100869971_100870097_-\n",
      "no expression for: HARS1 chr5_140057493_140057600_140058587_140058647_140058712_-\n",
      "no expression for: HYI chr1_43916674_43916982_43917081_43917099_43917215_-\n",
      "no expression for: HYI chr1_43916825_43916982_43917081_43917099_43917215_-\n",
      "no expression for: HYI chr1_43916829_43916982_43917081_43917099_43917215_-\n",
      "no expression for: HYI chr1_43917319_43917388_43917469_43917499_43917518_-\n",
      "no expression for: HYI chr1_43917876_43917990_43919057_43919084_43919168_-\n",
      "done: 340 out of 824\n",
      "no expression for: IQCN chr19_18367908_18368916_18375173_18375734_18378336_-\n",
      "no expression for: JPT1 chr17_73142761_73142779_73143615_73143651_73143748_-\n",
      "no expression for: LARS1 chr5_145524187_145524272_145529185_145529221_145529262_-\n",
      "done: 400 out of 824\n",
      "done: 420 out of 824\n",
      "no expression for: MRTFA chr22_40806286_40807911_40812980_40812990_40813050_-\n",
      "done: 460 out of 824\n",
      "done: 480 out of 824\n",
      "done: 500 out of 824\n",
      "done: 520 out of 824\n",
      "no expression for: QARS1 chr3_49141296_49141405_49141757_49141790_49141904_-\n",
      "no expression for: RBIS chr8_86129615_86129731_86131465_86131552_86131592_-\n",
      "no expression for: SMIM29 chr6_34215244_34215269_34215458_34215518_34215641_-\n",
      "done: 680 out of 824\n",
      "done: 740 out of 824\n",
      "done: 760 out of 824\n",
      "done: 800 out of 824\n",
      "LQ or UQ are zero length: ABCF3 chr3_183904295_183904424_183904442_183904584_183904663_+\n",
      "no expression for: AC013477.2 chr4_154243800_154243968_154243980_154245230_154245300_+\n",
      "no expression for: AC022137.3 chr19_53948254_53948376_53948472_53949480_53949590_+\n",
      "no expression for: AC093512.2 chr16_30075050_30075392_30075569_30075745_30075826_+\n",
      "no expression for: AC093512.2 chr16_30080139_30080299_30080311_30080627_30080710_+\n",
      "LQ or UQ are zero length: ACO1 chr9_32386369_32386436_32386440_32405483_32405601_+\n",
      "LQ or UQ are zero length: ACO2 chr22_41911771_41911921_41911996_41913531_41913635_+\n",
      "done: 20 out of 850\n",
      "no expression for: ANKHD1-EIF4EBP3 chr5_139917671_139917819_139917846_139918504_139918668_+\n",
      "done: 80 out of 850\n",
      "no expression for: C2orf92 chr2_98290919_98291020_98291024_98292308_98292391_+\n",
      "done: 120 out of 850\n",
      "no expression for: CCN6 chr6_112375278_112375364_112375373_112375538_112375608_+\n",
      "no expression for: CFAP251 chr12_122356498_122356601_122356614_122359192_122359589_+\n",
      "done: 200 out of 850\n",
      "no expression for: DGLUCY chr14_91636347_91636530_91636545_91639633_91639783_+\n",
      "no expression for: DGLUCY chr14_91639633_91639747_91639783_91642278_91642400_+\n",
      "no expression for: DNAAF6 chrX_106450378_106450652_106450656_106456103_106456258_+\n",
      "no expression for: DUS4L-BCAP29 chr7_107204416_107204654_107204692_107205033_107205121_+\n",
      "no expression for: DUS4L-BCAP29 chr7_107204416_107204654_107205121_107207495_107207631_+\n",
      "no expression for: EOLA1 chrX_148622746_148622837_148622888_148623564_148623630_+\n",
      "no expression for: EOLA1 chrX_148623564_148623611_148623630_148626914_148627046_+\n",
      "no expression for: FAM47E-STBD1 chr4_77189813_77189921_77189966_77192721_77192921_+\n",
      "done: 260 out of 850\n",
      "no expression for: FCSK chr16_70509253_70509318_70509396_70509551_70509670_+\n",
      "done: 280 out of 850\n",
      "no expression for: GARS1 chr7_30671064_30671157_30671175_30671863_30672053_+\n",
      "done: 300 out of 850\n",
      "no expression for: GUCY1B1 chr4_156693754_156693849_156693905_156696120_156696220_+\n",
      "done: 360 out of 850\n",
      "no expression for: ITPRID1 chr7_31617477_31617604_31618048_31622748_31622805_+\n",
      "done: 380 out of 850\n",
      "done: 400 out of 850\n",
      "no expression for: MRTFB chr16_14340330_14341214_14341364_14342783_14342938_+\n",
      "no expression for: MTARC1 chr1_220971216_220971356_220971407_220978394_220978455_+\n",
      "done: 460 out of 850\n",
      "no expression for: NFILZ chr19_8784386_8784435_8784439_8786200_8786291_+\n",
      "no expression for: NIBAN3 chr19_17641564_17641625_17641694_17643072_17643197_+\n",
      "no expression for: NKAPD1 chr11_111946291_111946367_111946408_111948939_111949039_+\n",
      "no expression for: NOP53 chr19_48255769_48255801_48255816_48258731_48258780_+\n",
      "no expression for: NTAQ1 chr8_124429014_124429096_124429172_124440164_124440263_+\n",
      "done: 500 out of 850\n",
      "done: 540 out of 850\n",
      "done: 560 out of 850\n",
      "done: 580 out of 850\n",
      "done: 600 out of 850\n",
      "no expression for: RAB5IF chr20_35238004_35238133_35238161_35240413_35240960_+\n",
      "done: 620 out of 850\n",
      "no expression for: RO60 chr1_193053709_193053787_193053828_193060781_193060907_+\n",
      "no expression for: SEPTIN9 chr17_75283973_75284292_75284296_75303223_75303279_+\n",
      "done: 660 out of 850\n",
      "done: 680 out of 850\n",
      "no expression for: SPTY2D1OS chr11_18622639_18622711_18622727_18624705_18624742_+\n",
      "done: 700 out of 850\n",
      "no expression for: TAMALIN chr12_52404823_52404925_52404991_52407471_52407564_+\n",
      "done: 720 out of 850\n",
      "no expression for: TEDC1 chr14_105957318_105957463_105957469_105957959_105958037_+\n",
      "no expression for: TEX51 chr2_127658786_127658854_127658858_127658941_127658980_+\n",
      "no expression for: TEX51 chr2_127658786_127658854_127658858_127659445_127659673_+\n",
      "done: 760 out of 850\n",
      "done: 800 out of 850\n",
      "no expression for: VPS35L chr16_19592847_19592975_19592976_19612986_19613045_+\n",
      "done: 820 out of 850\n",
      "done: 840 out of 850\n",
      "no expression for: ZNRD2 chr11_65338075_65338210_65338226_65338357_65338441_+\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "NMD_cols = ['NMD_junction']\n",
    "can_cols = ['coding_junction']\n",
    "           \n",
    "acc1_summary = aaf.get_GTEX_summary_pooled(acc1,NMD_cols,can_cols)\n",
    "acc2_summary = aaf.get_GTEX_summary_pooled(acc2,NMD_cols,can_cols)\n",
    "don1_summary = aaf.get_GTEX_summary_pooled(don1,NMD_cols,can_cols)\n",
    "don2_summary = aaf.get_GTEX_summary_pooled(don2,NMD_cols,can_cols)\n",
    "time_took = time.time()-start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc1_summary = pd.DataFrame(acc1_summary,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])\n",
    "acc2_summary = pd.DataFrame(acc2_summary,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])\n",
    "don1_summary = pd.DataFrame(don1_summary,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])\n",
    "don2_summary = pd.DataFrame(don2_summary,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])\n",
    "don_summary = pd.concat([don1_summary,don2_summary]).reset_index(drop=True)\n",
    "acc_summary = pd.concat([acc1_summary,acc2_summary]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [],
   "source": [
    "don_summary.to_csv('./misc/hg19_coding_donor_summary_2.tsv', sep=str('\\t'),header=True,index=None)\n",
    "acc_summary.to_csv('./misc/hg19_coding_acceptor_summary_2.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein-coding IR\n",
    "hg19_intron_retention_coding = pd.read_csv('./misc/hg19_intron_retention_coding.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)\n",
    "hg19_intron_retention_coding.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "hg19_intron_retention_coding['AS event position'] = hg19_intron_retention_coding['chr']+'_'+hg19_intron_retention_coding['upst_exon_start'].astype('str')+'_'+\\\n",
    "hg19_intron_retention_coding['upst_exon_end'].astype('str')+'_'+hg19_intron_retention_coding['downst_exon_start'].astype('str')+'_'+hg19_intron_retention_coding['downst_exon_end'].astype('str')+'_'+\\\n",
    "hg19_intron_retention_coding['str']\n",
    "hg19_intron_retention_coding = hg19_intron_retention_coding.drop_duplicates(['Target','AS event position'])\n",
    "hg19_intron_retention_coding['NMD_site_upst'] = hg19_intron_retention_coding['chr']+'_'+hg19_intron_retention_coding['upst_exon_end'].astype('str')+'_'+\\\n",
    "hg19_intron_retention_coding['str']\n",
    "hg19_intron_retention_coding['NMD_site_downst'] = hg19_intron_retention_coding['chr']+'_'+hg19_intron_retention_coding['downst_exon_start'].astype('str')+'_'+\\\n",
    "hg19_intron_retention_coding['str']\n",
    "hg19_intron_retention_coding['coding_junction'] = hg19_intron_retention_coding['chr']+'_'+hg19_intron_retention_coding['upst_exon_end'].astype('str')+'_'+\\\n",
    "hg19_intron_retention_coding['downst_exon_start'].astype('str')+'_'+hg19_intron_retention_coding['str']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LQ or UQ are zero length: ABCA8 chr17_66917523_66917591_66917593_66917633_-\n",
      "LQ or UQ are zero length: ABCA8 chr17_66925190_66925251_66925319_66925375_-\n",
      "LQ or UQ are zero length: ABHD14B chr3_52005476_52005714_52005828_52005908_-\n",
      "LQ or UQ are zero length: ACAN chr15_89398083_89398752_89398810_89402648_+\n",
      "LQ or UQ are zero length: ACRV1 chr11_125547692_125547913_125548124_125548192_-\n",
      "LQ or UQ are zero length: ACRV1 chr11_125547692_125547958_125548124_125548192_-\n",
      "LQ or UQ are zero length: ADAR chr1_154573517_154574724_154574861_154575102_-\n",
      "no expression for: AL391650.1 chr1_26498108_26498180_26498261_26498312_-\n",
      "done: 40 out of 1006\n",
      "no expression for: ATP5F1D chr19_1244314_1244439_1244736_1244824_+\n",
      "done: 120 out of 1006\n",
      "no expression for: CCNP chr19_40728115_40728379_40729059_40729210_-\n",
      "no expression for: CFAP298 chr21_33973977_33974283_33975471_33975602_-\n",
      "no expression for: CIAO2B chr16_66965968_66966104_66966175_66966203_-\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "NMD_cols = ['NMD_site_upst','NMD_site_downst']\n",
    "can_cols = ['coding_junction']\n",
    "           \n",
    "ir_summary = aaf.get_GTEX_summary_pooled(hg19_intron_retention_coding,NMD_cols,can_cols)\n",
    "time_took = time.time()-start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [],
   "source": [
    "ir_summary = pd.DataFrame(ir_summary,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [],
   "source": [
    "ir_summary.to_csv('./misc/hg19_coding_ir_summary.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotated USEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [],
   "source": [
    "# poison exons\n",
    "nmd_sample = pd.read_csv('./novel_events_catalog/hg19_poison_exons.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_sample['AS event position'] = nmd_sample['chr']+'_'+nmd_sample['upst_junction']+'_'+nmd_sample['downst_junction']+'_'+nmd_sample['str']\n",
    "nmd_sample['coding_skip_junction'] = nmd_sample['chr']+'_'+nmd_sample['skip_junction']+'_'+nmd_sample['str']\n",
    "nmd_sample['NMD_upst_junction'] = nmd_sample['chr']+'_'+nmd_sample['upst_junction']+'_'+nmd_sample['str']\n",
    "nmd_sample['NMD_downst_junction'] = nmd_sample['chr']+'_'+nmd_sample['downst_junction']+'_'+nmd_sample['str']\n",
    "nmd_sample.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "nmd_sample = nmd_sample.drop_duplicates(['Target','AS event position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: ABCC5 chr3_183655848_183660221_183660303_183660515_-\n",
      "no expression for: ABHD14A-ACY1 chr3_52012390_52018063_52018174_52019223_+\n",
      "LQ or UQ are zero length: ABI2 chr2_204193354_204206920_204206953_204231600_+\n",
      "no expression for: AC004832.3 chr22_30806668_30812223_30812392_30824449_+\n",
      "no expression for: AC004922.1 chr7_98931040_98933076_98933107_98935804_+\n",
      "no expression for: AC004997.1 chr22_30683549_30684695_30684765_30688762_-\n",
      "no expression for: AC006059.2 chr3_42799843_42827520_42827654_42835649_-\n",
      "done: 40 out of 1000\n",
      "done: 80 out of 1000\n",
      "no expression for: ATP5MF-PTCD1 chr7_99032891_99039013_99039110_99057709_-\n",
      "no expression for: ATP5PO chr21_35276325_35279030_35279125_35279645_-\n",
      "no expression for: BCLAF3 chrX_19935452_19941276_19941525_19947903_-\n",
      "no expression for: BORCS8-MEF2B chr19_19261573_19291495_19291570_19293382_-\n",
      "no expression for: C2orf92 chr2_98286297_98287654_98287769_98290919_+\n",
      "done: 140 out of 1000\n",
      "no expression for: CASTOR1 chr22_30683549_30683779_30683851_30684695_-\n",
      "no expression for: CCDC196 chr14_66958912_66963545_66963585_66964827_+\n",
      "no expression for: CCDC197 chr14_94464581_94465683_94465806_94465929_+\n",
      "no expression for: CFAP20DC chr3_58703355_58720199_58720285_58739496_-\n",
      "no expression for: CIBAR1 chr8_94738741_94739359_94739472_94740433_+\n",
      "done: 200 out of 1000\n",
      "done: 220 out of 1000\n",
      "done: 240 out of 1000\n",
      "no expression for: EIPR1 chr2_3341920_3356977_3357094_3358321_-\n",
      "no expression for: ELAPOR2 chr7_86539317_86540938_86541011_86541388_-\n",
      "no expression for: ELAPOR2 chr7_86569431_86570668_86570701_86571334_-\n",
      "done: 280 out of 1000\n",
      "no expression for: GARS1 chr7_30665935_30666673_30666806_30668176_+\n",
      "done: 380 out of 1000\n",
      "no expression for: IARS1 chr9_94991408_95002925_95002973_95003138_-\n",
      "no expression for: JPT1 chr17_73144766_73145186_73145280_73150437_-\n",
      "done: 440 out of 1000\n",
      "done: 460 out of 1000\n",
      "no expression for: MIDEAS chr14_74196688_74196773_74196882_74203701_-\n",
      "no expression for: MOSMO chr16_22019728_22075659_22075759_22086808_+\n",
      "no expression for: MTARC2 chr1_220922145_220927710_220927774_220928289_+\n",
      "done: 520 out of 1000\n",
      "no expression for: NIPSNAP2 chr7_56046088_56047453_56047579_56049166_+\n",
      "no expression for: NIPSNAP2 chr7_56052571_56059521_56059577_56062587_+\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "no expression for: PJVK chr2_179323354_179323920_179323996_179325075_+\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "no expression for: PYCR3 chr8_144690296_144690934_144691046_144691628_-\n",
      "no expression for: RBM14-RBM4 chr11_66384528_66407171_66407594_66410921_+\n",
      "done: 720 out of 1000\n",
      "no expression for: SEPTIN2 chr2_242263656_242263824_242264077_242265408_+\n",
      "no expression for: SEPTIN6 chrX_118750705_118752676_118752749_118759298_-\n",
      "done: 740 out of 1000\n",
      "done: 760 out of 1000\n",
      "no expression for: SLC66A2 chr18_77703462_77709687_77709838_77710724_-\n",
      "no expression for: SLC66A3 chr2_11312171_11312877_11312950_11315094_+\n",
      "no expression for: SMIM27 chr9_32552477_32565395_32565626_32566389_+\n",
      "done: 780 out of 1000\n",
      "done: 800 out of 1000\n",
      "no expression for: STIMATE chr3_52889460_52929894_52929971_52931307_-\n",
      "done: 840 out of 1000\n",
      "done: 860 out of 1000\n",
      "no expression for: TMEM270 chr7_73275596_73275799_73275921_73279323_+\n",
      "done: 880 out of 1000\n",
      "done: 940 out of 1000\n",
      "no expression for: VXN chr8_67408727_67411060_67411153_67417610_+\n",
      "done: 960 out of 1000\n",
      "0:1000 done\n",
      "LQ or UQ are zero length: AAMDC chr11_77588095_77595826_77595962_77611617_+\n",
      "LQ or UQ are zero length: ABCB8 chr7_150725697_150729931_150730148_150731360_+\n",
      "LQ or UQ are zero length: ABCC3 chr17_48712342_48720910_48721022_48733193_+\n",
      "LQ or UQ are zero length: ABCE1 chr4_146042543_146042756_146042965_146044173_+\n",
      "LQ or UQ are zero length: ABHD6 chr3_58242432_58244130_58244351_58252916_+\n",
      "LQ or UQ are zero length: ABI3BP chr3_100645325_100664424_100664520_100712150_-\n",
      "LQ or UQ are zero length: ABLIM3 chr5_148577965_148578607_148578685_148579879_+\n",
      "no expression for: ABRAXAS1 chr4_84403397_84405145_84405319_84406139_-\n",
      "done: 20 out of 1000\n",
      "no expression for: AL049629.2 chr11_33720060_33724214_33724283_33731752_-\n",
      "done: 40 out of 1000\n",
      "no expression for: AOPEP chr9_97845001_97848212_97848401_97848964_+\n",
      "no expression for: ATP5MG chr11_118272432_118276372_118276424_118277652_+\n",
      "no expression for: ATPSCKMT chr5_10236727_10236950_10237152_10239179_-\n",
      "no expression for: BPNT2 chr8_57878911_57879669_57879777_57890609_-\n",
      "done: 120 out of 1000\n",
      "no expression for: CARS1 chr11_3061161_3061678_3061811_3062126_-\n",
      "done: 140 out of 1000\n",
      "no expression for: CFAP20DC chr3_58739590_58779907_58780057_58792121_-\n",
      "no expression for: CFAP299 chr4_81284038_81341368_81341477_81504247_+\n",
      "no expression for: CFAP299 chr4_81504337_81560872_81560989_81791147_+\n",
      "done: 180 out of 1000\n",
      "no expression for: CIBAR1 chr8_94713686_94714393_94714611_94715848_+\n",
      "no expression for: CIBAR1 chr8_94731016_94736738_94736823_94738622_+\n",
      "no expression for: COP1 chr1_176133027_176137801_176137884_176145046_-\n",
      "done: 220 out of 1000\n",
      "done: 240 out of 1000\n",
      "no expression for: DENND2B chr11_8720937_8721404_8721559_8724119_-\n",
      "no expression for: DGLUCY chr14_91626741_91631591_91631681_91633569_+\n",
      "no expression for: DGLUCY chr14_91681900_91690023_91690139_91691028_+\n",
      "no expression for: DNAAF4 chr15_55783454_55787326_55787400_55789910_-\n",
      "done: 260 out of 1000\n",
      "no expression for: ELAPOR1 chr1_109743522_109743941_109744039_109745566_+\n",
      "done: 320 out of 1000\n",
      "no expression for: FCSK chr16_70509670_70510090_70510237_70512175_+\n",
      "done: 340 out of 1000\n",
      "done: 360 out of 1000\n",
      "no expression for: HARS1 chr5_140062804_140063131_140063252_140070440_-\n",
      "done: 380 out of 1000\n",
      "done: 400 out of 1000\n",
      "no expression for: IQCM chr4_150607620_150612890_150612943_150654396_-\n",
      "no expression for: IRAG1 chr11_10655618_10709056_10709324_10711812_-\n",
      "no expression for: IRAG1 chr11_10673729_10709056_10709324_10711812_-\n",
      "done: 420 out of 1000\n",
      "no expression for: MARCHF1 chr4_164466895_164477001_164477125_164506901_-\n",
      "done: 500 out of 1000\n",
      "done: 520 out of 1000\n",
      "done: 560 out of 1000\n",
      "no expression for: NSG2 chr5_173491318_173493207_173493282_173531231_+\n",
      "no expression for: NTAQ1 chr8_124442273_124447682_124447782_124448693_+\n",
      "no expression for: NTAQ1 chr8_124453589_124460287_124460464_124479319_+\n",
      "no expression for: ODAPH chr4_76481359_76488420_76488555_76489324_+\n",
      "done: 600 out of 1000\n",
      "no expression for: POGLUT3 chr11_108348438_108348751_108348867_108350028_-\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "no expression for: QARS1 chr3_49133512_49134543_49134720_49135425_-\n",
      "no expression for: RAB5IF chr20_35236221_35236293_35236403_35238004_+\n",
      "no expression for: RAB5IF chr20_35236221_35236293_35236403_35240413_+\n",
      "done: 680 out of 1000\n",
      "done: 700 out of 1000\n",
      "done: 720 out of 1000\n",
      "done: 760 out of 1000\n",
      "no expression for: SHISAL2B chr5_63986641_63989914_63990009_63991334_+\n",
      "no expression for: SHISAL2B chr5_63986641_63989914_63990026_63991334_+\n",
      "done: 780 out of 1000\n",
      "no expression for: SLC66A2 chr18_77703462_77709519_77709655_77710724_-\n",
      "done: 800 out of 1000\n",
      "done: 840 out of 1000\n",
      "no expression for: TARS1 chr5_33445510_33446750_33446870_33448647_+\n",
      "done: 860 out of 1000\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "done: 940 out of 1000\n",
      "done: 960 out of 1000\n",
      "1000:2000 done\n",
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: ABAT chr16_8844396_8851052_8851167_8851614_+\n",
      "no expression for: AC008764.4 chr19_16689220_16757907_16758072_16770206_-\n",
      "no expression for: AC008770.2 chr19_12036088_12072476_12072573_12087853_+\n",
      "no expression for: AC023055.1 chr12_56197512_56214023_56214233_56215785_-\n",
      "no expression for: AC134669.1 chr17_29632648_29634905_29634977_29640997_-\n",
      "done: 20 out of 1000\n",
      "no expression for: AL139353.1 chr14_31869718_31872056_31872255_31897350_-\n",
      "done: 40 out of 1000\n",
      "no expression for: AP001781.2 chr11_111742213_111746422_111746828_111747205_-\n",
      "no expression for: APOC4-APOC2 chr19_45448147_45448397_45449332_45451723_+\n",
      "no expression for: ATP5F1A chr18_43671817_43673146_43673328_43675019_-\n",
      "no expression for: ATP5PD chr17_73035366_73035967_73036136_73038276_-\n",
      "no expression for: BMERB1 chr16_15528616_15541792_15541889_15609162_+\n",
      "no expression for: BMERB1 chr16_15609285_15626844_15626992_15661840_+\n",
      "no expression for: C2orf92 chr2_98286297_98287657_98287769_98292308_+\n",
      "no expression for: C2orf92 chr2_98291020_98292094_98292184_98292308_+\n",
      "no expression for: C2orf92 chr2_98306790_98310910_98311062_98315489_+\n",
      "no expression for: C2orf92 chr2_98317767_98318524_98318630_98319132_+\n",
      "done: 120 out of 1000\n",
      "done: 160 out of 1000\n",
      "no expression for: CHURC1-FNTB chr14_65392798_65398856_65399010_65470967_+\n",
      "no expression for: CIAO2A chr15_64381050_64382397_64382514_64385844_-\n",
      "no expression for: CIAO3 chr16_787329_787498_787695_789643_-\n",
      "no expression for: CIAO3 chr16_789738_790022_790103_790920_-\n",
      "done: 180 out of 1000\n",
      "done: 200 out of 1000\n",
      "no expression for: CORO7-PAM16 chr16_4411470_4411750_4411804_4411986_-\n",
      "done: 240 out of 1000\n",
      "done: 280 out of 1000\n",
      "done: 300 out of 1000\n",
      "done: 320 out of 1000\n",
      "done: 340 out of 1000\n",
      "done: 360 out of 1000\n",
      "no expression for: GRAMD2A chr15_72461337_72461758_72461904_72462188_-\n",
      "no expression for: HDGFL3 chr15_83832827_83861004_83861060_83876083_-\n",
      "no expression for: HEXD chr17_80382379_80385906_80386003_80386434_+\n",
      "no expression for: HEXD chr17_80397589_80398349_80398489_80398873_+\n",
      "no expression for: IRAG2 chr12_25254259_25254872_25254959_25255132_+\n",
      "no expression for: LTO1 chr11_69488089_69489319_69489534_69489958_-\n",
      "no expression for: MARS1 chr12_57909774_57909922_57909941_57910028_+\n",
      "no expression for: MEAK7 chr16_84529519_84531025_84531253_84531540_-\n",
      "no expression for: MEAK7 chr16_84531717_84537564_84537753_84538207_-\n",
      "no expression for: MINAR1 chr15_79755663_79756037_79756102_79760529_+\n",
      "done: 480 out of 1000\n",
      "done: 500 out of 1000\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "done: 600 out of 1000\n",
      "no expression for: PWWP3A chr19_1358463_1358587_1358692_1360135_+\n",
      "no expression for: PWWP3A chr19_1358463_1358587_1358699_1360135_+\n",
      "no expression for: RMC1 chr18_21088005_21088619_21088740_21089157_+\n",
      "no expression for: RPL17-C18orf32 chr18_47010141_47014870_47014936_47015729_-\n",
      "no expression for: RXYLT1 chr12_64174954_64176321_64176484_64178750_+\n",
      "no expression for: SEPTIN9 chr17_75277636_75290856_75290985_75303223_+\n",
      "no expression for: SEPTIN9 chr17_75489158_75492296_75492471_75493397_+\n",
      "no expression for: SINHCAF chr12_31448267_31449844_31449978_31451011_-\n",
      "no expression for: SLC66A2 chr18_77703462_77709519_77709636_77710724_-\n",
      "done: 740 out of 1000\n",
      "done: 780 out of 1000\n",
      "no expression for: TAFA2 chr12_62104198_62114468_62114705_62147403_-\n",
      "no expression for: TARS3 chr15_102226265_102239534_102239638_102241289_-\n",
      "no expression for: TEX49 chr12_49121342_49129871_49129955_49152992_+\n",
      "done: 820 out of 1000\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "no expression for: VPS35L chr16_19586444_19589626_19589680_19590374_+\n",
      "no expression for: WARS1 chr14_100828258_100832291_100832404_100835424_-\n",
      "done: 920 out of 1000\n",
      "done: 940 out of 1000\n",
      "no expression for: ZNF875 chr19_37826159_37837790_37837954_37838092_+\n",
      "2000:3000 done\n",
      "no expression for: AARS1 chr16_70311057_70312803_70312948_70316523_-\n",
      "LQ or UQ are zero length: ABCB11 chr2_169789043_169789291_169789450_169791694_-\n",
      "LQ or UQ are zero length: ABCB4 chr7_87076521_87077950_87078022_87079284_-\n",
      "LQ or UQ are zero length: ABCB7 chrX_74296489_74317291_74317373_74318777_-\n",
      "LQ or UQ are zero length: ABCC2 chr10_101606884_101609481_101609662_101610359_+\n",
      "LQ or UQ are zero length: ABCC4 chr13_95813589_95814019_95814046_95815376_-\n",
      "no expression for: AC009070.1 chr16_80864654_80874016_80874119_81030919_-\n",
      "no expression for: AC009070.1 chr16_81031034_81034853_81034939_81040339_-\n",
      "no expression for: AC107871.1 chr15_68489966_68510874_68510988_68521840_-\n",
      "no expression for: AC118553.2 chr1_100435718_100440527_100440625_100459093_+\n",
      "no expression for: AC118553.2 chr1_100480976_100483238_100483371_100515465_+\n",
      "no expression for: ADA2 chr22_17684663_17684744_17684779_17687961_-\n",
      "no expression for: AL109811.3 chr1_11085265_11089744_11089838_11090236_+\n",
      "no expression for: AL160272.2 chr9_120466843_120470841_120471007_120610343_+\n",
      "done: 40 out of 787\n",
      "no expression for: ARMH1 chr1_45169845_45189475_45189578_45189794_+\n",
      "no expression for: ATP5PO chr21_35276325_35278753_35278804_35279645_-\n",
      "no expression for: CCNQ chrX_152861645_152862880_152863164_152864420_-\n",
      "no expression for: CERT1 chr5_74696122_74698500_74698617_74698797_-\n",
      "no expression for: CFAP298-TCP10L chr21_33940389_33945594_33946986_33951004_-\n",
      "no expression for: CFAP298-TCP10L chr21_33954725_33956471_33956615_33975471_-\n",
      "no expression for: CFAP298-TCP10L chr21_33954725_33959623_33959787_33975471_-\n",
      "no expression for: COP1 chr1_176153828_176170397_176170481_176175708_-\n",
      "no expression for: ELP1 chr9_111637250_111637913_111638015_111640275_-\n",
      "no expression for: ELP1 chr9_111637250_111639893_111640080_111640275_-\n",
      "no expression for: ELP1 chr9_111644467_111647076_111647192_111651612_-\n",
      "no expression for: ELP1 chr9_111659565_111660416_111660514_111660782_-\n",
      "no expression for: ELP1 chr9_111681629_111682971_111683063_111685122_-\n",
      "done: 260 out of 787\n",
      "no expression for: FP565260.1 chr21_45546846_45547456_45547574_45547813_+\n",
      "no expression for: GARS1 chr7_30634759_30635440_30635549_30638412_+\n",
      "no expression for: GARS1 chr7_30638513_30639040_30639161_30639563_+\n",
      "no expression for: GARS1 chr7_30643188_30645942_30646013_30649201_+\n",
      "no expression for: GARS1 chr7_30643188_30648024_30648099_30649201_+\n",
      "no expression for: GARS1 chr7_30651861_30653061_30653213_30655512_+\n",
      "no expression for: GET1 chr21_40762789_40763153_40763402_40763695_+\n",
      "done: 320 out of 787\n",
      "no expression for: HARS1 chr5_140062804_140063131_140063249_140070440_-\n",
      "no expression for: HARS1 chr5_140062804_140067331_140067518_140070440_-\n",
      "done: 360 out of 787\n",
      "done: 380 out of 787\n",
      "no expression for: KIFBP chr10_70749014_70749473_70749615_70760180_+\n",
      "no expression for: KIFBP chr10_70749014_70759471_70759570_70760180_+\n",
      "no expression for: KIFBP chr10_70765671_70766598_70766685_70768597_+\n",
      "no expression for: KIFBP chr10_70765671_70767383_70767507_70768597_+\n",
      "no expression for: LARS1 chr5_145493874_145498572_145498637_145499934_-\n",
      "no expression for: LARS1 chr5_145515227_145519179_145519266_145519767_-\n",
      "done: 500 out of 787\n",
      "done: 520 out of 787\n",
      "done: 540 out of 787\n",
      "no expression for: PRKN chr6_161807909_161863926_161864011_161969886_-\n",
      "no expression for: QARS1 chr3_49133512_49134567_49134720_49135425_-\n",
      "done: 580 out of 787\n",
      "done: 680 out of 787\n",
      "done: 700 out of 787\n",
      "done: 740 out of 787\n",
      "no expression for: YARS1 chr1_33256855_33263193_33263239_33263364_-\n",
      "no expression for: YARS1 chr1_33263444_33270146_33270484_33272083_-\n",
      "3000:4000 done\n"
     ]
    }
   ],
   "source": [
    "NMD_cols = ['NMD_upst_junction','NMD_downst_junction']\n",
    "can_cols = ['coding_skip_junction']\n",
    "\n",
    "a = []\n",
    "s = 0\n",
    "while s<len(nmd_sample):\n",
    "    e=s+1000\n",
    "    temp_df = nmd_sample.iloc[s:min(e,len(nmd_sample))]\n",
    "    temp_summary = aaf.get_GTEX_summary_pooled(temp_df,NMD_cols,can_cols)\n",
    "    a = a+temp_summary\n",
    "    print(str(s)+':'+str(e)+' done')\n",
    "    s=s+1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df = pd.DataFrame(a,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df.to_csv('./novel_events_catalog/hg19_poison_exons_summary.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "# essential exons\n",
    "nmd_sample = pd.read_csv('./novel_events_catalog/hg19_essential_exons.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_sample['AS event position'] = nmd_sample['chr']+'_'+nmd_sample['upst_junction']+'_'+nmd_sample['downst_junction']+'_'+nmd_sample['str']\n",
    "nmd_sample['NMD_skip_junction'] = nmd_sample['chr']+'_'+nmd_sample['skip_junction']+'_'+nmd_sample['str']\n",
    "nmd_sample['coding_upst_junction'] = nmd_sample['chr']+'_'+nmd_sample['upst_junction']+'_'+nmd_sample['str']\n",
    "nmd_sample['coding_downst_junction'] = nmd_sample['chr']+'_'+nmd_sample['downst_junction']+'_'+nmd_sample['str']\n",
    "nmd_sample.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "nmd_sample = nmd_sample.drop_duplicates(['Target','AS event position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: ABHD4 chr14_23072667_23072830_23072984_23075328_+\n",
      "no expression for: ABRAXAS1 chr4_84391549_84393375_84393441_84397796_-\n",
      "no expression for: AC004997.1 chr22_30683549_30688762_30688840_30689640_-\n",
      "done: 20 out of 1000\n",
      "no expression for: AFG1L chr6_108687536_108723200_108723258_108768417_+\n",
      "done: 40 out of 1000\n",
      "done: 60 out of 1000\n",
      "no expression for: ATP5PO chr21_35279757_35281386_35281515_35284593_-\n",
      "done: 80 out of 1000\n",
      "no expression for: BUD23 chr7_73101230_73101329_73101425_73105246_+\n",
      "no expression for: BUD23 chr7_73101425_73105246_73105342_73106926_+\n",
      "done: 100 out of 1000\n",
      "no expression for: CASTOR1 chr22_30683549_30684695_30684765_30685374_-\n",
      "done: 120 out of 1000\n",
      "no expression for: CCDC196 chr14_66957537_66957739_66957822_66958344_+\n",
      "no expression for: CCN4 chr8_134203456_134225107_134225386_134237633_+\n",
      "done: 140 out of 1000\n",
      "no expression for: CFAP298 chr21_33976593_33979959_33980026_33982148_-\n",
      "done: 160 out of 1000\n",
      "no expression for: CFAP91 chr3_119422069_119425656_119425732_119426251_+\n",
      "no expression for: COP1 chr1_176133027_176145046_176145143_176153769_-\n",
      "no expression for: CPLANE1 chr5_37239971_37243115_37243221_37244477_-\n",
      "done: 260 out of 1000\n",
      "done: 280 out of 1000\n",
      "no expression for: EIPR1 chr2_3218019_3261070_3261226_3341788_-\n",
      "no expression for: EPRS1 chr1_220198600_220203728_220203822_220205729_-\n",
      "no expression for: GCOM1 chr15_57910386_57913806_57913898_57917977_+\n",
      "no expression for: GCOM1 chr15_57931743_57953648_57953731_57967166_+\n",
      "no expression for: GCOM1 chr15_57953731_57967166_57967266_58004182_+\n",
      "done: 360 out of 1000\n",
      "done: 380 out of 1000\n",
      "done: 400 out of 1000\n",
      "no expression for: IRAG2 chr12_25256992_25257295_25257396_25259877_+\n",
      "done: 420 out of 1000\n",
      "no expression for: ITPRID2 chr2_182757387_182760757_182760806_182761635_+\n",
      "done: 440 out of 1000\n",
      "done: 460 out of 1000\n",
      "no expression for: LTO1 chr11_69482780_69486517_69486587_69487984_-\n",
      "no expression for: MAP11 chr7_99755082_99755256_99755385_99755466_-\n",
      "no expression for: MARCHF6 chr5_10378944_10387106_10387178_10390444_+\n",
      "done: 480 out of 1000\n",
      "no expression for: METTL27 chr7_73254450_73254744_73254879_73255400_-\n",
      "done: 500 out of 1000\n",
      "done: 520 out of 1000\n",
      "done: 540 out of 1000\n",
      "no expression for: PALM2AKAP2 chr9_112811135_112898407_112900819_112918599_+\n",
      "done: 620 out of 1000\n",
      "no expression for: PRKN chr6_162475206_162622163_162622284_162683557_-\n",
      "no expression for: PRKN chr6_162622284_162683557_162683797_162864342_-\n",
      "no expression for: PRKN chr6_162683797_162864342_162864505_163148694_-\n",
      "done: 660 out of 1000\n",
      "no expression for: RETREG2 chr2_220043355_220043644_220043750_220044455_+\n",
      "done: 700 out of 1000\n",
      "done: 720 out of 1000\n",
      "done: 740 out of 1000\n",
      "no expression for: SEPTIN5 chr22_19708392_19709163_19709259_19709345_+\n",
      "done: 760 out of 1000\n",
      "done: 780 out of 1000\n",
      "no expression for: SLC49A4 chr3_122514382_122525704_122525797_122545647_+\n",
      "no expression for: SMIM29 chr6_34214926_34215244_34215269_34215458_-\n",
      "done: 820 out of 1000\n",
      "done: 840 out of 1000\n",
      "done: 860 out of 1000\n",
      "done: 880 out of 1000\n",
      "no expression for: TMEM131L chr4_154388433_154394997_154395040_154471225_+\n",
      "done: 900 out of 1000\n",
      "done: 920 out of 1000\n",
      "no expression for: TUT4 chr1_52961187_52962678_52962855_52975268_-\n",
      "no expression for: TVP23C-CDRT4 chr17_15450462_15456999_15457143_15458596_-\n",
      "0:1000 done\n",
      "done: 0 out of 1000\n",
      "LQ or UQ are zero length: AASDH chr4_57237809_57244314_57244630_57248643_-\n",
      "LQ or UQ are zero length: ABAT chr16_8829666_8839858_8839955_8841965_+\n",
      "no expression for: AC009779.3 chr12_56109980_56110717_56110789_56112875_+\n",
      "done: 20 out of 1000\n",
      "done: 40 out of 1000\n",
      "done: 80 out of 1000\n",
      "done: 100 out of 1000\n",
      "done: 140 out of 1000\n",
      "no expression for: CEP20 chr16_15967484_15973661_15973745_15977865_-\n",
      "done: 180 out of 1000\n",
      "no expression for: CHURC1-FNTB chr14_65390844_65392728_65392798_65470967_+\n",
      "no expression for: CIAO3 chr16_784871_786266_786398_787186_-\n",
      "done: 200 out of 1000\n",
      "no expression for: COA8 chr14_104038157_104040444_104040507_104045708_+\n",
      "done: 220 out of 1000\n",
      "done: 240 out of 1000\n",
      "done: 260 out of 1000\n",
      "no expression for: DNAAF4 chr15_55742565_55759128_55759359_55783321_-\n",
      "no expression for: ELAPOR1 chr1_109656958_109704516_109704636_109707121_+\n",
      "done: 300 out of 1000\n",
      "no expression for: FAM47E-STBD1 chr4_77172973_77177331_77177676_77184857_+\n",
      "done: 360 out of 1000\n",
      "no expression for: GOLM2 chr15_44620985_44624186_44624276_44629961_+\n",
      "done: 380 out of 1000\n",
      "done: 400 out of 1000\n",
      "no expression for: IFTAP chr11_36654988_36657601_36657667_36669566_+\n",
      "done: 420 out of 1000\n",
      "no expression for: KARS1 chr16_75670445_75674082_75674247_75675462_-\n",
      "done: 480 out of 1000\n",
      "no expression for: LTO1 chr11_69486587_69487984_69488089_69489958_-\n",
      "no expression for: MARF1 chr16_15706634_15709687_15709862_15710909_-\n",
      "no expression for: MARS1 chr12_57882892_57883050_57883128_57883990_+\n",
      "no expression for: MARS1 chr12_57883341_57883679_57883754_57883990_+\n",
      "done: 520 out of 1000\n",
      "done: 540 out of 1000\n",
      "done: 560 out of 1000\n",
      "no expression for: NARS1 chr18_55276675_55278869_55278939_55280834_-\n",
      "done: 580 out of 1000\n",
      "done: 600 out of 1000\n",
      "done: 640 out of 1000\n",
      "no expression for: PIP4P2 chr8_92008972_92020971_92021023_92030618_-\n",
      "done: 660 out of 1000\n",
      "done: 680 out of 1000\n",
      "done: 700 out of 1000\n",
      "no expression for: RARS1 chr5_167919852_167920899_167921007_167921555_+\n",
      "done: 740 out of 1000\n",
      "done: 760 out of 1000\n",
      "no expression for: RTRAF chr14_52458159_52460441_52460540_52465212_+\n",
      "no expression for: RUSF1 chr16_31512052_31519085_31519199_31519395_-\n",
      "no expression for: SEPTIN1 chr16_30392794_30393081_30393204_30393414_-\n",
      "done: 780 out of 1000\n",
      "done: 800 out of 1000\n",
      "done: 840 out of 1000\n",
      "done: 860 out of 1000\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "done: 920 out of 1000\n",
      "done: 940 out of 1000\n",
      "no expression for: VIRMA chr8_95502310_95503806_95504137_95504880_-\n",
      "no expression for: VIRMA chr8_95505046_95507088_95507184_95507959_-\n",
      "no expression for: VIRMA chr8_95508708_95511602_95511734_95518728_-\n",
      "done: 960 out of 1000\n",
      "1000:2000 done\n",
      "no expression for: AARS1 chr16_70289739_70291936_70292120_70292883_-\n",
      "no expression for: AARS1 chr16_70289739_70291936_70292147_70292883_-\n",
      "no expression for: AARS1 chr16_70295060_70296249_70296427_70298861_-\n",
      "no expression for: AARS1 chr16_70295060_70296270_70296427_70298861_-\n",
      "no expression for: AARS1 chr16_70295060_70296276_70296427_70298861_-\n",
      "LQ or UQ are zero length: ABCC8 chr11_17453791_17464267_17464429_17464725_-\n",
      "LQ or UQ are zero length: ABCC8 chr11_17485151_17491648_17491769_17496433_-\n",
      "LQ or UQ are zero length: ABHD12 chr20_25284264_25287469_25287551_25288602_-\n",
      "LQ or UQ are zero length: ABHD12 chr20_25289130_25290082_25290211_25295561_-\n",
      "no expression for: ABHD14A-ACY1 chr3_52012390_52019223_52019287_52019377_+\n",
      "no expression for: AC099489.1 chr16_11578929_11581118_11581242_11582565_-\n",
      "no expression for: AC118553.2 chr1_100464971_100472590_100472712_100476921_+\n",
      "no expression for: AC118553.2 chr1_100477089_100480858_100480976_100515465_+\n",
      "no expression for: AC245033.1 chr15_83224872_83226510_83226736_83240094_-\n",
      "no expression for: ADA2 chr22_17669337_17670832_17670922_17672573_-\n",
      "done: 60 out of 673\n",
      "done: 80 out of 673\n",
      "no expression for: CYBC1 chr17_80403836_80404499_80404572_80405456_-\n",
      "done: 180 out of 673\n",
      "no expression for: DMAC2 chr19_41939578_41944123_41944319_41945379_-\n",
      "no expression for: ELP1 chr9_111641030_111641726_111641837_111642332_-\n",
      "done: 240 out of 673\n",
      "done: 260 out of 673\n",
      "no expression for: GAB4 chr22_17449273_17450503_17451083_17468850_-\n",
      "no expression for: GAB4 chr22_17449273_17450833_17451083_17468850_-\n",
      "no expression for: GARS1 chr7_30662078_30665850_30665935_30668176_+\n",
      "no expression for: GATD3A chr21_45553720_45553984_45554036_45555942_+\n",
      "no expression for: GATD3B chr21_45553720_45553984_45554036_45555942_+\n",
      "done: 280 out of 673\n",
      "no expression for: HARS1 chr5_140056701_140056912_140057005_140057240_-\n",
      "done: 300 out of 673\n",
      "no expression for: IARS1 chr9_95050192_95050408_95050564_95051583_-\n",
      "no expression for: LARS1 chr5_145509721_145510582_145510672_145512461_-\n",
      "no expression for: QARS1 chr3_49135550_49135625_49135691_49135786_-\n",
      "no expression for: QARS1 chr3_49141405_49141757_49141904_49142050_-\n",
      "no expression for: QARS1 chr3_49141405_49141790_49141904_49142050_-\n",
      "no expression for: RETREG3 chr17_40738155_40738783_40738909_40739852_-\n",
      "no expression for: RETREG3 chr17_40738909_40739852_40739882_40744074_-\n",
      "no expression for: REX1BD chr19_18699887_18700223_18700493_18701664_+\n",
      "no expression for: REX1BD chr19_18699887_18700289_18700493_18701664_+\n",
      "no expression for: RMC1 chr18_21084411_21086934_21087018_21087949_+\n",
      "done: 520 out of 673\n",
      "no expression for: SYNJ2BP-COX16 chr14_70842488_70855187_70855323_70883617_-\n",
      "done: 600 out of 673\n",
      "no expression for: YARS1 chr1_33263444_33272083_33272212_33276192_-\n",
      "no expression for: YARS1 chr1_33272212_33276192_33276367_33282789_-\n",
      "2000:3000 done\n"
     ]
    }
   ],
   "source": [
    "NMD_cols = ['NMD_skip_junction']\n",
    "can_cols = ['coding_upst_junction','coding_downst_junction']\n",
    "\n",
    "a = []\n",
    "s = 0\n",
    "while s<len(nmd_sample):\n",
    "    e=s+1000\n",
    "    temp_df = nmd_sample.iloc[s:min(e,len(nmd_sample))]\n",
    "    temp_summary = aaf.get_GTEX_summary_pooled(temp_df,NMD_cols,can_cols)\n",
    "    a = a+temp_summary\n",
    "    print(str(s)+':'+str(e)+' done')\n",
    "    s=s+1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df = pd.DataFrame(a,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df.to_csv('./novel_events_catalog/hg19_essential_exons_summary.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {},
   "outputs": [],
   "source": [
    "# alt starts\n",
    "nmd_sample = pd.read_csv('./novel_events_catalog/hg19_alt_starts.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_sample['AS event position'] = nmd_sample['chr']+'_'+nmd_sample['flank_exon_start'].astype('str')+'_'+nmd_sample['flank_exon_end'].astype('str')+\\\n",
    "nmd_sample['exon_start'].astype('str')+'_'+nmd_sample['exon_start_coding'].astype('str')+'_'+nmd_sample['exon_end'].astype('str')+'_'+nmd_sample['str']\n",
    "\n",
    "nmd_sample['NMD_junction'] = nmd_sample['chr']+'_'+nmd_sample['flank_exon_end'].astype('str')+'_'+nmd_sample['exon_start'].astype('str')+'_'+nmd_sample['str']\n",
    "nmd_sample['coding_junction'] = nmd_sample['chr']+'_'+nmd_sample['flank_exon_end'].astype('str')+'_'+nmd_sample['exon_start_coding'].astype('str')+'_'+nmd_sample['str']\n",
    "nmd_sample.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "nmd_sample = nmd_sample.drop_duplicates(['Target','AS event position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LQ or UQ are zero length: AANAT chr17_74464754_7446499174465199_74465255_74465409_+\n",
      "no expression for: AARS1 chr16_70288524_7028863770289623_70289631_70289739_-\n",
      "no expression for: AARS1 chr16_70289631_7028973970291956_70291936_70292120_-\n",
      "no expression for: AARS1 chr16_70316523_7031668770323433_70323324_70323458_-\n",
      "LQ or UQ are zero length: ABCA7 chr19_1054779_10548771055082_1055096_1055350_+\n",
      "LQ or UQ are zero length: ABCC12 chr16_48172139_4817228648173099_48173074_48173247_-\n",
      "LQ or UQ are zero length: ABHD11 chr7_73151892_7315206573152663_73152658_73152793_-\n",
      "LQ or UQ are zero length: ABHD5 chr3_43740768_4374085343743879_43743707_43744079_+\n",
      "done: 20 out of 1000\n",
      "no expression for: ADSS1 chr14_105205665_105205715105206104_105206087_105206153_+\n",
      "done: 40 out of 1000\n",
      "no expression for: AP000812.5 chr11_71816010_7181605871817102_71816982_71817256_+\n",
      "done: 60 out of 1000\n",
      "done: 80 out of 1000\n",
      "no expression for: CCDC196 chr14_66957739_6695782266958325_66958344_66958403_+\n",
      "no expression for: CCNQ chrX_152861462_152861645152864424_152864420_152864480_-\n",
      "no expression for: CEMIP2 chr9_74344964_7434516974345655_74345621_74345830_-\n",
      "no expression for: CERT1 chr5_74715166_7471524974721117_74721186_74721324_-\n",
      "no expression for: CFAP20DC chr3_58870271_5887043558899424_58899434_58899590_-\n",
      "done: 200 out of 1000\n",
      "no expression for: CYBC1 chr17_80405456_8040549780407050_80407046_80407168_-\n",
      "done: 220 out of 1000\n",
      "done: 260 out of 1000\n",
      "done: 280 out of 1000\n",
      "no expression for: EEF1AKMT2 chr10_126463244_126463351126477608_126477612_126477726_-\n",
      "no expression for: ELP1 chr9_111658776_111658924111659171_111659233_111659318_-\n",
      "done: 300 out of 1000\n",
      "done: 340 out of 1000\n",
      "no expression for: GARS1 chr7_30639563_3063966530640745_30640675_30640816_+\n",
      "no expression for: GARS1 chr7_30651712_3065186130655514_30655512_30655674_+\n",
      "no expression for: GATD1 chr11_772427_772521773428_773522_773629_-\n",
      "no expression for: GUCY1A1 chr4_156629388_156629446156631834_156631694_156632403_+\n",
      "no expression for: HEXD chr17_80377625_8037775980382295_80382270_80382379_+\n",
      "no expression for: IRAG2 chr12_25232332_2523239625232636_25232590_25232697_+\n",
      "no expression for: LARS1 chr5_145524187_145524272145529115_145529185_145529262_-\n",
      "no expression for: LARS1 chr5_145524187_145524272145529115_145529221_145529262_-\n",
      "done: 460 out of 1000\n",
      "no expression for: MACO1 chr1_25784882_2578538325810618_25810607_25810765_+\n",
      "no expression for: MAP11 chr7_99754717_9975483799754916_99754996_99755082_-\n",
      "no expression for: MARS1 chr12_57883050_5788312857883238_57883207_57883341_+\n",
      "no expression for: MARS1 chr12_57905791_5790588657906086_57906019_57906136_+\n",
      "done: 560 out of 1000\n",
      "done: 600 out of 1000\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "no expression for: QARS1 chr3_49141296_4914140549141806_49141757_49141904_-\n",
      "no expression for: QARS1 chr3_49141296_4914140549141806_49141790_49141904_-\n",
      "done: 700 out of 1000\n",
      "no expression for: RETREG2 chr2_220044455_220044485220044920_220044832_220044967_+\n",
      "no expression for: REXO5 chr16_20824512_2082462420826307_20826249_20826375_+\n",
      "no expression for: RIPOR1 chr16_67573631_6757365467573753_67573749_67573802_+\n",
      "done: 740 out of 1000\n",
      "no expression for: SEPTIN2 chr2_242274541_242274627242275374_242275390_242275513_+\n",
      "no expression for: SEPTIN4 chr17_56603581_5660367456604293_56604043_56604339_-\n",
      "done: 780 out of 1000\n",
      "no expression for: SERINC4 chr15_44089043_4408913844089339_44089343_44089554_-\n",
      "done: 820 out of 1000\n",
      "no expression for: SLC66A1 chr1_19651127_1965125619652020_19651989_19652076_+\n",
      "done: 860 out of 1000\n",
      "no expression for: TARS1 chr5_33448647_3344883733453399_33453394_33453517_+\n",
      "done: 880 out of 1000\n",
      "no expression for: TEX51 chr2_127657099_127657188127657375_127657412_127657429_+\n",
      "done: 900 out of 1000\n",
      "no expression for: TVP23C-CDRT4 chr17_15450373_1545046215457079_15456999_15457143_-\n",
      "done: 960 out of 1000\n",
      "0:1000 done\n",
      "done: 0 out of 463\n",
      "LQ or UQ are zero length: ACTG2 chr2_74129487_7412961574135417_74135800_74135910_+\n",
      "LQ or UQ are zero length: ADGRD1 chr12_131621518_131621559131622189_131622682_131622774_+\n",
      "LQ or UQ are zero length: ADGRL2 chr1_82402412_8240252182406877_82407718_82407729_+\n",
      "done: 20 out of 463\n",
      "no expression for: ASDURF chr2_190528627_190528680190530083_190530148_190530177_+\n",
      "no expression for: ATP5F1A chr18_43671648_4367181743674946_43675019_43675097_-\n",
      "done: 40 out of 463\n",
      "no expression for: CBLIF chr11_59611352_5961152859612734_59612848_59612972_-\n",
      "no expression for: CCDC198 chr14_57949783_5794986957957660_57957707_57957789_-\n",
      "no expression for: CERT1 chr5_74715166_7471524974721095_74721186_74721324_-\n",
      "no expression for: CIAO3 chr16_787186_787329789540_789643_789738_-\n",
      "done: 80 out of 463\n",
      "done: 100 out of 463\n",
      "done: 120 out of 463\n",
      "no expression for: FYB1 chr5_39124355_3912438039126074_39126100_39126237_-\n",
      "no expression for: HDGFL2 chr19_4496299_44964024496916_4497955_4498028_+\n",
      "done: 200 out of 463\n",
      "no expression for: LARS1 chr5_145519767_145519824145522419_145522435_145522647_-\n",
      "done: 260 out of 463\n",
      "no expression for: NIBAN3 chr19_17654333_1765444017656032_17657495_17657606_+\n",
      "done: 280 out of 463\n",
      "done: 300 out of 463\n",
      "no expression for: QARS1 chr3_49136543_4913668649136763_49136777_49136864_-\n",
      "no expression for: SEPTIN12 chr16_4829691_48297874833318_4833463_4833558_-\n",
      "no expression for: SEPTIN12 chr16_4835981_48361064837470_4837481_4837668_-\n",
      "done: 360 out of 463\n",
      "done: 400 out of 463\n",
      "no expression for: TMEM273 chr10_50373811_5037384150374490_50374914_50375054_-\n",
      "no expression for: VPS35L chr16_19576173_1957627219580566_19580746_19580913_+\n",
      "no expression for: VXN chr8_67408672_6740872767417457_67417610_67417763_+\n",
      "1000:2000 done\n"
     ]
    }
   ],
   "source": [
    "NMD_cols = ['NMD_junction']\n",
    "can_cols = ['coding_junction']\n",
    "\n",
    "a = []\n",
    "s = 0\n",
    "while s<len(nmd_sample):\n",
    "    e=s+1000\n",
    "    temp_df = nmd_sample.iloc[s:min(e,len(nmd_sample))]\n",
    "    temp_summary = aaf.get_GTEX_summary_pooled(temp_df,NMD_cols,can_cols)\n",
    "    a = a+temp_summary\n",
    "    print(str(s)+':'+str(e)+' done')\n",
    "    s=s+1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df = pd.DataFrame(a,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df.to_csv('./novel_events_catalog/hg19_alt_starts_summary.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [],
   "source": [
    "# alt ends\n",
    "nmd_sample = pd.read_csv('./novel_events_catalog/hg19_alt_ends.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_sample['AS event position'] = nmd_sample['chr']+'_'+nmd_sample['exon_start'].astype('str')+'_'+nmd_sample['exon_end'].astype('str')+'_'+\\\n",
    "nmd_sample['exon_end_coding'].astype('str')+'_'+nmd_sample['flank_exon_start'].astype('str')+'_'+nmd_sample['flank_exon_end'].astype('str')+'_'+nmd_sample['str']\n",
    "\n",
    "nmd_sample['NMD_junction'] = nmd_sample['chr']+'_'+nmd_sample['exon_end'].astype('str')+'_'+nmd_sample['flank_exon_start'].astype('str')+'_'+nmd_sample['str']\n",
    "nmd_sample['coding_junction'] = nmd_sample['chr']+'_'+nmd_sample['exon_end_coding'].astype('str')+'_'+nmd_sample['flank_exon_start'].astype('str')+'_'+nmd_sample['str']\n",
    "nmd_sample.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "nmd_sample = nmd_sample.drop_duplicates(['Target','AS event position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no expression for: AARS1 chr16_70294947_70295053_70295060_70296249_70296427_-\n",
      "no expression for: AARS1 chr16_70296249_70296387_70296427_70298861_70299005_-\n",
      "no expression for: AARS1 chr16_70303521_70303737_70303666_70304099_70304243_-\n",
      "LQ or UQ are zero length: ABCC8 chr11_17419886_17420010_17419988_17424208_17424300_-\n",
      "LQ or UQ are zero length: ABCC8 chr11_17449836_17449944_17449952_17450112_17450217_-\n",
      "done: 20 out of 1000\n",
      "no expression for: AC118553.2 chr1_100459093_100459301_100459297_100464817_100464971_+\n",
      "done: 40 out of 1000\n",
      "done: 60 out of 1000\n",
      "done: 80 out of 1000\n",
      "no expression for: ATP5F1A chr18_43675019_43675113_43675097_43678138_43678262_-\n",
      "done: 100 out of 1000\n",
      "no expression for: BUD23 chr7_73098097_73098231_73098134_73100966_73101061_+\n",
      "no expression for: CCDC197 chr14_94467491_94467665_94467669_94469569_94469700_+\n",
      "no expression for: CCNP chr19_40729295_40729549_40729453_40730395_40730550_-\n",
      "done: 160 out of 1000\n",
      "no expression for: CEP20 chr16_15976806_15976905_15976877_15977865_15978062_-\n",
      "no expression for: CERT1 chr5_74681727_74681882_74681859_74685417_74685512_-\n",
      "done: 180 out of 1000\n",
      "no expression for: CIP2A chr3_108300955_108300985_108301049_108301824_108301930_-\n",
      "no expression for: CLBA1 chr14_105455280_105455411_105455425_105457828_105457957_+\n",
      "done: 220 out of 1000\n",
      "no expression for: CYBC1 chr17_80403740_80403756_80403836_80404499_80404572_-\n",
      "done: 240 out of 1000\n",
      "done: 300 out of 1000\n",
      "no expression for: ELP1 chr9_111656223_111656368_111656346_111658776_111658924_-\n",
      "no expression for: ELP1 chr9_111659428_111659563_111659565_111660782_111660861_-\n",
      "no expression for: ELP1 chr9_111673290_111673491_111673460_111674544_111674774_-\n",
      "no expression for: ELP1 chr9_111688803_111688875_111688883_111689652_111689733_-\n",
      "no expression for: FP565260.3 chr21_45656750_45656948_45657100_45658331_45658371_-\n",
      "done: 360 out of 1000\n",
      "no expression for: GARS1 chr7_30640675_30640840_30640816_30642650_30642738_+\n",
      "no expression for: GET1 chr21_40762624_40762845_40762789_40763695_40763762_+\n",
      "no expression for: GET1 chr21_40762624_40763402_40762789_40763695_40763762_+\n",
      "done: 380 out of 1000\n",
      "no expression for: HARS1 chr5_140056912_140057042_140057005_140057240_140057338_-\n",
      "done: 420 out of 1000\n",
      "done: 440 out of 1000\n",
      "no expression for: JPT2 chr16_1733510_1733621_1733593_1735440_1735588_+\n",
      "no expression for: KIFBP chr10_70748559_70748968_70749014_70760180_70760278_+\n",
      "done: 500 out of 1000\n",
      "done: 540 out of 1000\n",
      "done: 560 out of 1000\n",
      "done: 580 out of 1000\n",
      "no expression for: NKAPD1 chr11_111946291_111946374_111946367_111948939_111949039_+\n",
      "no expression for: NKAPD1 chr11_111946291_111946374_111946408_111948939_111949039_+\n",
      "done: 600 out of 1000\n",
      "done: 620 out of 1000\n",
      "done: 640 out of 1000\n",
      "no expression for: PIMREG chr17_6348396_6348662_6348724_6350783_6351078_+\n",
      "done: 660 out of 1000\n",
      "no expression for: QARS1 chr3_49135425_49135525_49135550_49135625_49135691_-\n",
      "no expression for: QARS1 chr3_49140778_49140864_49140842_49141064_49141139_-\n",
      "no expression for: RARS1 chr5_167919664_167919796_167919852_167920899_167921007_+\n",
      "done: 720 out of 1000\n",
      "done: 740 out of 1000\n",
      "no expression for: RSKR chr17_26940458_26940703_26940706_26941105_26941203_-\n",
      "done: 780 out of 1000\n",
      "done: 860 out of 1000\n",
      "done: 880 out of 1000\n",
      "done: 900 out of 1000\n",
      "done: 940 out of 1000\n",
      "0:1000 done\n",
      "LQ or UQ are zero length: ABCA13 chr7_48273615_48273824_48273748_48278838_48279002_+\n",
      "LQ or UQ are zero length: ABCB7 chrX_74291344_74291539_74291518_74293524_74293611_-\n",
      "LQ or UQ are zero length: ABCC12 chr16_48130657_48130843_48130813_48134783_48134920_-\n",
      "LQ or UQ are zero length: ABCC3 chr17_48735796_48735907_48735857_48736598_48736729_+\n",
      "LQ or UQ are zero length: ABCC8 chr11_17416719_17417017_17416822_17417157_17417265_-\n",
      "no expression for: BOLA2-SMG1P6 chr16_29458123_29458347_29458243_29465337_29465434_-\n",
      "no expression for: BUD23 chr7_73098097_73098330_73098134_73100966_73101061_+\n",
      "no expression for: CARS1 chr11_3062126_3062261_3062214_3063395_3063486_-\n",
      "done: 60 out of 381\n",
      "no expression for: DGLUCY chr14_91681749_91682137_91681900_91691028_91691699_+\n",
      "done: 100 out of 381\n",
      "no expression for: DUS4L-BCAP29 chr7_107214149_107214328_107214266_107215633_107215755_+\n",
      "done: 120 out of 381\n",
      "no expression for: HEXD chr17_80397507_80398210_80397589_80398873_80398951_+\n",
      "done: 160 out of 381\n",
      "no expression for: MACROH2A1 chr5_134696187_134696470_134696294_134705096_134705293_-\n",
      "no expression for: MACROH2A1 chr5_134696187_134696470_134696297_134705096_134705293_-\n",
      "no expression for: MAP11 chr7_99753294_99753542_99753448_99754009_99754172_-\n",
      "no expression for: MARS1 chr12_57909703_57909941_57909774_57910028_57910120_+\n",
      "done: 220 out of 381\n",
      "done: 240 out of 381\n",
      "done: 360 out of 381\n",
      "done: 380 out of 381\n",
      "1000:2000 done\n"
     ]
    }
   ],
   "source": [
    "NMD_cols = ['NMD_junction']\n",
    "can_cols = ['coding_junction']\n",
    "\n",
    "a = []\n",
    "s = 0\n",
    "while s<len(nmd_sample):\n",
    "    e=s+1000\n",
    "    temp_df = nmd_sample.iloc[s:min(e,len(nmd_sample))]\n",
    "    temp_summary = aaf.get_GTEX_summary_pooled(temp_df,NMD_cols,can_cols)\n",
    "    a = a+temp_summary\n",
    "    print(str(s)+':'+str(e)+' done')\n",
    "    s=s+1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df = pd.DataFrame(a,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df.to_csv('./novel_events_catalog/hg19_alt_ends_summary.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "metadata": {},
   "outputs": [],
   "source": [
    "# intron retention\n",
    "nmd_sample = pd.read_csv('./novel_events_catalog/hg19_intron_retention.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_sample['AS event position'] = nmd_sample['chr']+'_'+nmd_sample['exon_start'].astype('str')+'_'+nmd_sample['exon_end_coding'].astype('str')+'_'+\\\n",
    "nmd_sample['downst_exon_start_coding'].astype('str')+'_'+nmd_sample['exon_end'].astype('str')+'_'+nmd_sample['str']\n",
    "\n",
    "nmd_sample['coding_junction'] = nmd_sample['chr']+'_'+nmd_sample['exon_end_coding'].astype('str')+'_'+nmd_sample['downst_exon_start_coding'].astype('str')+'_'+nmd_sample['str']\n",
    "nmd_sample['NMD_site_upst'] = nmd_sample['chr']+'_'+nmd_sample['exon_end_coding'].astype('str')+'_'+nmd_sample['str']\n",
    "nmd_sample['NMD_site_downst'] = nmd_sample['chr']+'_'+nmd_sample['downst_exon_start_coding'].astype('str')+'_'+nmd_sample['str']\n",
    "\n",
    "nmd_sample.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "nmd_sample = nmd_sample.drop_duplicates(['Target','AS event position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done: 0 out of 93\n",
      "LQ or UQ are zero length: APBB3 chr5_139941172_139941307_139941685_139941812_-\n",
      "done: 20 out of 93\n",
      "done: 40 out of 93\n",
      "no expression for: PLPBP chr8_37623044_37623151_37623229_37623264_+\n",
      "done: 60 out of 93\n",
      "done: 80 out of 93\n",
      "0:1000 done\n"
     ]
    }
   ],
   "source": [
    "NMD_cols = ['NMD_site_upst','NMD_site_downst']\n",
    "can_cols = ['coding_junction']\n",
    "\n",
    "a = []\n",
    "s = 0\n",
    "while s<len(nmd_sample):\n",
    "    e=s+1000\n",
    "    temp_df = nmd_sample.iloc[s:min(e,len(nmd_sample))]\n",
    "    temp_summary = aaf.get_GTEX_summary_pooled(temp_df,NMD_cols,can_cols)\n",
    "    a = a+temp_summary\n",
    "    print(str(s)+':'+str(e)+' done')\n",
    "    s=s+1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df = pd.DataFrame(a,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df.to_csv('./novel_events_catalog/hg19_intron_retention_summary.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "metadata": {},
   "outputs": [],
   "source": [
    "# intron detention\n",
    "nmd_sample = pd.read_csv('./novel_events_catalog/hg19_intron_detention.tsv',delimiter=\"\\t\",\n",
    "                                   index_col=None,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_sample['AS event position'] = nmd_sample['chr']+'_'+nmd_sample['upst_exon_start'].astype('str')+'_'+nmd_sample['upst_exon_end'].astype('str')+'_'+\\\n",
    "nmd_sample['downst_exon_start'].astype('str')+'_'+nmd_sample['downst_exon_end'].astype('str')+'_'+nmd_sample['str']\n",
    "\n",
    "nmd_sample['NMD_junction'] = nmd_sample['chr']+'_'+nmd_sample['upst_exon_end'].astype('str')+'_'+nmd_sample['downst_exon_start'].astype('str')+'_'+nmd_sample['str']\n",
    "nmd_sample['coding_site_upst'] = nmd_sample['chr']+'_'+nmd_sample['upst_exon_end'].astype('str')+'_'+nmd_sample['str']\n",
    "nmd_sample['coding_site_downst'] = nmd_sample['chr']+'_'+nmd_sample['downst_exon_start'].astype('str')+'_'+nmd_sample['str']\n",
    "\n",
    "nmd_sample.rename(columns={'gene_name':'Target'},inplace=True)\n",
    "nmd_sample = nmd_sample.drop_duplicates(['Target','AS event position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LQ or UQ are zero length: ABHD4 chr14_23072295_23072401_23072506_23072667_+\n",
      "LQ or UQ are zero length: ANGPTL4 chr19_8438589_8438839_8438900_8439254_+\n",
      "LQ or UQ are zero length: ANKRD13B chr17_27940372_27940737_27941281_27941779_+\n",
      "LQ or UQ are zero length: APOL3 chr22_36541521_36541547_36541625_36541647_-\n",
      "done: 20 out of 148\n",
      "done: 40 out of 148\n",
      "done: 60 out of 148\n",
      "done: 80 out of 148\n",
      "done: 100 out of 148\n",
      "no expression for: SHFL chr19_10199787_10199884_10199983_10200054_+\n",
      "no expression for: TMEM250 chr9_139006435_139006796_139007302_139008870_-\n",
      "done: 140 out of 148\n",
      "0:1000 done\n"
     ]
    }
   ],
   "source": [
    "NMD_cols = ['NMD_junction']\n",
    "can_cols = ['coding_site_upst','coding_site_downst']\n",
    "\n",
    "a = []\n",
    "s = 0\n",
    "while s<len(nmd_sample):\n",
    "    e=s+1000\n",
    "    temp_df = nmd_sample.iloc[s:min(e,len(nmd_sample))]\n",
    "    temp_summary = aaf.get_GTEX_summary_pooled(temp_df,NMD_cols,can_cols)\n",
    "    a = a+temp_summary\n",
    "    print(str(s)+':'+str(e)+' done')\n",
    "    s=s+1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df = pd.DataFrame(a,columns = ['gene_name','AS event position','psi_L','psi_H',\n",
    "                                            'log2FC_global','log2FC_local','zscore_global','zscore_local'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_df.to_csv('./novel_events_catalog/hg19_intron_detention_summary.tsv', sep=str('\\t'),header=True,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
